## Sequence Processing

### Sequencing Data Quality Assessment

#### Error Rate

#### QA Metrics

### Pipelines  
Sequence data was processed using a number of commonly used pipelines.
* The Mothur (version 1.37, http://www.mothur.org/) pipeline used was based on the MiSeq SOP [@schloss2009introducing]. 
Modifications to the SOP as a different 16S region was sequenced than the region the SOP was developed for, see the Makefile in the project github repository (__TODO__ add website) and supplemental material  for details __TODO__ add pipeline details to supplemental.
* Four different pipelines using QIIME were used to process the sequencing data.  The methods included open reference and _de novo_ clustering with and without chimera removal [@caporaso2010qiime]. 
* DADA2 an R native pipeline was also used to process the sequencing data [@callahan2016dada2].
* Pop (https://github.com/MihaiPop/GEMS-db), the pipeline uses Sickle for read trimming [@sickle], XZY for merging paired-end reads, DNAclust for OTU assignment [@ghodsi2011dnaclust], and BLAST for taxonomic annotation [@altschul1990basic].
