## Data Analysis
All data analysis was performed using the statistical programming language R (REF) and the RStudio IDE (REF). 
Initial quality assessment of the sequence files (fastq) was performed using the Bioconductor package Rqc (REF). 

## Measurement Assessment  
* Measurement assessment of the count table values sample level feature abundance  
* Assessment count values with and without normalization and transformation 
* Variance - variability in observed counts between the four PCR replicates 
      * Coefficient of variation was calculated $sd/mu$ for each feature, pipeline, and normalization method
      * __TODO__ Method used to compare CV values

* Bias - how well the observed count values agree with expected values
      * The expectd count values were calculated as follows, with $p = 2^-t$ and $t$ is the titration factor. $C_{exp}$ is the expected count value and $C_{post}$ is the observed count value for the unmixed post-treatment sample and $C_{pre}$ is the unmixed pre-treatment sample. To account for potential within and between plate effects the observed counts for the pre- and post-treatment replicate in the same plate and side of the replicate 96 well plate as the samples being assessed were used to calculate the expected count value.   

$$
C_{exp} = [C_{post}\times p] + [C_{pre} \times (1-p)]
$$
      * The following error metric was used to summarize measurement bias  
      
$$
E = \frac{\sqrt{1/n \sum_i^n(C_{obs} -C_{exp})^2}}{\sqrt{C_{exp}^2}}
$$
      
* Normalization Methods
      * None
      * TSS - total sum scaling
      * CSS - cumulative sum scaling
      * Senthil's method
      * rareify
      * square root transformation - AIST spike-ins
      * Holmes arcsin transformation

##  log-fold Change Assessment
* Differential Abundance Methods
      * metagenomeFeatures
      * DESeq
      * EdgeR
      * Limma?
* Variance estimates provided by differential abundance methods were used to compare the logFC variance between bioinformatic pipelines, normalization methods, and differential abundance methods (__Q__ Not sure can directly compare differential abundance methods as differences in model assumptions determine the variability)
* Bias 
      * expectation calculation 
      
