16S rRNA marker-gene surveys use targeted sequencing to characterize prokaryotic microbial communities. Analysis of these studies is confronted with numerous bioinformatic pipelines and downstream analysis methods, with limited guidance on how to decide between appropriate methods from simulation studies or limited complexity benchmark studies. Appropriate data sets and statistics for assessing these methods are needed. A mixture of environmental samples is one approach to generating data sets for assessment with the complexity of real data while providing an expected value. We developed a mixture dataset for assessing 16S rRNA bioinformatic pipelines and downstream analysis methods using samples collected from vaccine trial participants. Stool samples were collected before and after exposure to pathogenic _Escherichia coli_ ETEC. DNA from stool samples prior to _E. coli_ exposure was titrated into stools samples collected after exposure, in effect diluting the amount of _E. coli_ in the mixed sample. The sequencing data were processed using multiple bioinformatic pipelines. We performed a qualitative and quantitative assessment. The dataset developed for this study will serve as a valuable community resource for assessing 16S rRNA marker-gene survey bioinformatic methods.
