---
title: "Discussion"
author: "Nate Olson"
date: '`r Sys.Date()`'
output:
  bookdown::pdf_document2: 
    toc: FALSE
bibliography: [mgtst.bib, packages.bib]
---

We generated a 16S metagenomic benchmarking dataset with the diversity, relative abundance dynamic range, and sequencing artifacts of a real dataset using mixtures of environmental samples 
We used the dataset to assess the feature presence/ absence, relative abundance, and log fold-change for count tables generated using four different bioinformatic pipelines

We assessed the performance of three bioinformatic pipelines, open-clustering, _de-novo_ clustering, sequencing inference, and unclustered. 
Running these pipelines on our mixture dataset resulted in a range of total feature abundance and features per sample. 
Despite the wide range in number of features the pipelines all generated similarly sparse datasets. 
As the dataset is highly redundant, 180 samples derived from 10 environmental samples lower sparsity was expected. 
__Paulson metagenomeSeq paper supplemental figure 4 1-3% features detected per sample__  
The qualitative assessmen results, specifically the high proportion of titration and unmixed sample specific features that could not be explained by sampling alone, indicates that the sequence inference method had a high false negative feature detection rate 
This high false negative rate likey resulted in higher sparsity. 
For the other pipelines the high sparsity is attributed to features that were artifacts of the sequencing process, false positives.  
The high rate of false positive features have been observed in benchmarking studies using mock communities __REF__.
The 16S region sequenced in the study is larger than the region the de-novo and open clustering pipelines were initially developed for __REF-Schloss__. 
The larger region has a smaller overlap between the forward and reverse reads as a result in our study the merging of the forward and reverse reads did not allow for the sequence error correction that occurs when their is greater overlap. 

As the qualitative assessment results were pipeline dependent the implications for 16S metagenomic studies vary by pipeline. 
For de-novo and open-reference clustering methods any conclusions made based on low abundance features require additional justification. 
Specifically, how do you know whether the feature is a measurement artifact or represents a member of the microbial community. 
This is especially relevant for studies characterizing the rare biosphere. 
A study exploring the microbial ecology of __SOME BIRD__ used a hard filter for low abundance features, but also compared the results with and without the filter ensuring that any conclusions were not biases by using the artibrary filter or including the low abundance features that are likely predominantly measurement artifacts __REF__. 
For 16S metagenomic studies using DADA2, missing low abundance features are more likely to impact presence/absence diversity analysis. 
Though a user can be more confident that an observed feature represents a member of the microbial community and not a measurement artifact. 
It is unlikely that the number of features in a sample accurately reflects the true richness of a sample though how well the results datasets are able to detect real differences in richness between samples is unknown.

The quantiative assessment results, both relative abundance and log fold-change estimates were individual specific. The individual specific results are a limitation in inferring the proportion of prokaryotic DNA in a titration from post-exposure samples, $\theta$. 
Pipeline had minimal effect on the quantiative results when accounting for individual effects. 
However, large outliers were commonly observed. 
Visual exploration of the results indicates a feature-specific effect. 
Experimental factors such as feature-level relative abundance of the titrations or unmixed samples could not account for the outliers. 
Additionally, outliers were not restricted to specific taxonomic groups. 
What dependency means for 16S gene surveys, when does bioinformatic pipeline matter and when does it not matter? 
Unable to define a set of characteristics that define can be used to identify poor performing features. 
In general low abundance features are noisier but high abundance features are not necessarily accurate? 

__Outliers as a product of artificial merging or splitting of features__

Compositional nature of 16S metagenomic data is well known within the field. 
The proportion of prokaryotic DNA in a sample is rarely taken into consideration in 16S studies. 
Our error metrics decrease when using our inferred $\theta$ values to correct for difference in prokaryotic DNA proportions. 
The extent to which the impact of differences in the proportion of prokaryotic DNA between sets of samples compared in a 16S studies is not well known. 
However, our results indicate that accounting for differences in proportion of prokaryotic DNA can improve the accuracy of the relative abundance comparisons. 

# Conclusions  
This two-sample-titration dataset  can be used to evaluate and characterize bioinformatic pipelines and clustering methods. 
The sequence dataset presented in this study can be processed with any 16S bioinformatic pipeline to generate a count table. 
Our quantitative and qualitative assessment can then be performed on the count table and the results compared to those obtained using the pipelines included in this study. 
Based on the results of our assessment of four bioinformatic pipelines the pipelines generate sets of features with different characteristics interms of numbers of features. 
The objective of any pipeline is to differentiate true biological sequences in a dataset from artifacts of the measurement process. 
Users should consider whether a pipeline is minimizes false positives (DADA2) or false negatives (Mothur). 
Continued effort to improve the 16S metagenomic measurement process to better understand and account for differences in the proportion of prokaryotic DNA in the samples and sequence artifacts. 
Addressing both of these issues requires advances in both the molecular biology and computational components of the measurement process. 

```{r include = FALSE}
### -------------------------------- Old text that was cut ---------------------
# There were two primary limitations of the study that were a product of the experimental design. 
# Only features that were differentially abundant between the pre- and post-exposure were used in the assessment. 
# Using samples from the vaccine trial provided a specific features, _E. coli_ that could be used during method development. 
# However, only a limited number of features were differentially abundant between the pre- and post-exposure samples resulting in a smaller set of features that could be used in our assessment. 
# Generating mixtures of samples with less similarity would increase the number of features used in the assessment. 
# Additionally, using samples from other environments would increase the taxonomic diversity of features used in the assessment and potentially allowing for a more rigorus evaluation of the relationship between the assessment metrics and phylogeny. 
# The second limitation of the experimental design was the difference in the proportion of bacterial DNA between the pre- and post-exposure samples. 
# We were able to use an assay targeting the 16S rRNA gene to detect changes in the concentation of bacterial DNA across titration but we were unable to estimate the proportion of bacterial DNA in the unmixed samples using the qPCR data. 
# Using the 16S sequencing data we inferred the proportion of bacterial DNA from the post-exposure sample in each titration. 
# However, the uncertainty and accuracy of the inferrence method is not known resulting in an unaccounted for error source. 
# A better method for estimating the proportion of bacterial DNA in the unmixed samples would increase the accuracy of the error metrics. 
# Alternatively samples that are primarily prokaryotic DNA could be used
```

