---
title: "Discussion"
author: "Nate Olson"
date: '`r Sys.Date()`'
output:
  bookdown::pdf_document2: 
    toc: FALSE
bibliography: [mgtst.bib, packages.bib]
---

We generated a 16S rRNA marker-gene survey benchmarking dataset with the diversity, relative abundance dynamic range, and sequencing artifacts of a real dataset using mixtures of environmental samples 
We used the dataset to assess feature presence/ absence, relative abundance, and log fold-change for count tables generated using four different bioinformatic pipelines.

We assessed the performance of three bioinformatic pipelines, open-clustering, _de-novo_ clustering, sequencing inference, and unclustered. 
Running these pipelines on our mixture dataset resulted in a wide range of total feature abundance and features per sample. 
Despite the wide range in number of features the pipelines all generated similarly sparse datasets. 
The sparsity (percentage of zero-valued elements in a matrix) was comparable to previously published values, 1-3\% [@paulson2013differential].
As the dataset is highly redundant, 180 samples derived from 10 environmental samples, we expected lower sparsity. 
We attributed the higher than expected sparsity to false positive and false negative or true low abundance features. 
False positives are sequencing or PCR artifacts that are not appropriately filtered or assigned to an appropriate feature by the bioinformatic pipeline. 
The 16S rRNA region sequenced in the study is larger than the region the _de-novo_ and open clustering pipelines were initially developed for, potentially explaining the higher than expected sparsity [@kozich2013development] __QIIME V4 REF__. 
The larger region has a smaller overlap between the forward and reverse reads as a result in our study the merging of the forward and reverse reads did not allow for the sequence error correction that occurs when their is greater overlap. <!-- But it's illumina, I'm surprised if this is a true issue.  What level of sequencing correction is presumed/found typically? -->
However, both the _de-novo_ (Mothur) and open-reference clustering (QIIME) pipelines produced count tables with significantly more features than expected in evaluation studies using defined mixtures of cell and DNA, mock communities __REF__. 
The sequence inference method (DADA2) produced a count table with significantly fewer features compared to the _de-novo_ and open-reference clustering pipelines, yet had comparable sparsity. 
False negatives provide a potential explanation for the higher than expected sparsity. 
A false negative occurs when a lower abundance sequence representing an organism within the sample is clustered with a higher abundance sequence inflating the dataset sparsity. 

The qualitative assessment results, specifically the high proportion of titration and unmixed sample specific features that could not be explained by sampling alone, indicate that the sequence inference method had a high false negative feature detection rate. 
While microbial abundance distributions are known to have long tails, it is likely that the observed sparsity is an artifact of the 16S measurement process based on results from previously mentioned mock community benchmarking studies.
As the qualitative assessment results were pipeline dependent the implications for 16S metagenomic studies vary by pipeline. 
For _de-novo_ and open-reference clustering methods any conclusions made based on low abundance features require additional justification or validation. 
Specifically, determining whether the feature is a measurement artifact or represents a member of the microbial community. 
This is especially relevant for studies characterizing the rare biosphere.  
For example, a study exploring the microbial ecology of the Red-necked stint _Calidris ruficollis_, a migratory shorebird, used a hard filter for low abundance features, but also compared the results with and without the filter ensuring that the study conclusions were not biased by using the arbitrary filter or including the low abundant features, likely predominantly measurement artifacts [@risely2017gut]. 
In contrast, for 16S metagenomic studies using DADA2, missing low abundance features are more likely to impact presence/absence ecological diversity analysis. 
When a sequencing dataset is processed using DADA2, the user can be more confident that an observed feature represents a member of the microbial community and not a measurement artifact. 
At the same it is unlikely that the number of features in a sample accurately reflects the true richness of a sample though whether real differences in richness between samples are detectable when a dataset is processed using DADA2 is unknown.

The quantitative assessment results, both relative abundance and log fold-change estimates were individual specific.
The individual specific results are a limitation in inferring the proportion of prokaryotic DNA in a titration from post-exposure samples, $\theta$. 
We were able to use an assay targeting the 16S rRNA gene to detect changes in the concentration of bacterial DNA across titration but we were unable to absolutely quantify the proportion of bacterial DNA in the unmixed samples using qPCR data. 
Using the 16S sequencing data we inferred the proportion of bacterial DNA from the post-exposure sample in each titration. 
However, the uncertainty and accuracy of the inference method is not known resulting in an unaccounted for error source. 
A better method for estimating the proportion of bacterial DNA in the unmixed samples would increase the accuracy of the error metrics. Based on this observation, we recommend that studies based on stool samples seeking inferences in a longitudinal series across multiple subjects carefully estimate bacterial DNA proportions and adjust inferences accordingly.

While the relative abundance bias metric was significantly different between pipelines overall, pipeline choice had minimal impact on the quantitative assessment results when accounting for individual effects. 
However, large outliers were commonly observed. 
The outliers could not be attributed to the bioinformatic pipelines as the outliers were observed for the unclustered datasets as well. 
We believe the poor performance for outliers was attributed to the biases inherenet to the molecular biology. 
Upon visual exploration of the results we noticed a feature-specific effect. 
Mismatches in the primer binding regions impact PCR efficiency, a potential cause for poor feature-specific performance __REF__. 
We also evaluated the taxonomy and phylogenetic relationship for the relative abundance and log fold-change bias and variance metrics. 
While a phylogenetic signal was detected for all metrics, poor feature-level performance we were unable to attributed to any specific taxonomic group or phylogenetic clade. 
We were unable to define a set of characteristics (quantitate, taxonomic, or phylogenetic) that can be used to identify poor performing features. 

Based on the results of our quantitative assessment, community level analysis are likely more accurate than feature-level analyses. <!-- is this an argument for true metagenomic seq instead of 16S? see recent bracken paper by Salzberg et al-->
Therefore results from individual analysis like differential abundance, which rely on log-fold change estimates, are more susceptible to unknown biases and are potentially artifacts of the measurement process. 
Additional work is needed to further characterize these outliers to increase confidence in the results of feature-level analyses.  

# Conclusions  
This two-sample-titration dataset can be used to evaluate and characterize bioinformatic pipelines and clustering methods. 
The sequence dataset presented in this study can be processed with any 16S bioinformatic pipeline to generate a count table. 
Our quantitative and qualitative assessment can then be performed on the count table and the results compared to those obtained using the pipelines included in this study. 
Based on the results of our assessment of four bioinformatic pipelines the pipelines generate sets of features with different characteristics interms of total abundance, features per samples, and total features. 
The objective of any pipeline is to differentiate true biological sequences from artifacts of the measurement process. 
Users should consider whether a pipeline minimizes false positives (DADA2) or false negatives (Mothur) is more appropriate for their study objectives. 
Furthermore as the feature-specific quantitative assessment results could not be attributed to any feature quantitative, phylogenetic, or taxonomy, feature-level results for any 16S metagenomic study should be interpreted with care. 
Addressing both of these issues requires advances in both the molecular biology and computational components of the measurement process. 

```{r include = FALSE}
### -------------------------------- Old text that was cut ---------------------
# There were two primary limitations of the study that were a product of the experimental design. 
# Only features that were differentially abundant between the pre- and post-exposure were used in the assessment. 
# Using samples from the vaccine trial provided a specific features, _E. coli_ that could be used during method development. 
# However, only a limited number of features were differentially abundant between the pre- and post-exposure samples resulting in a smaller set of features that could be used in our assessment. 
# Generating mixtures of samples with less similarity would increase the number of features used in the assessment. 
# Additionally, using samples from other environments would increase the taxonomic diversity of features used in the assessment and potentially allowing for a more rigorus evaluation of the relationship between the assessment metrics and phylogeny. 
# The second limitation of the experimental design was the difference in the proportion of bacterial DNA between the pre- and post-exposure samples. 
# We were able to use an assay targeting the 16S rRNA gene to detect changes in the concentation of bacterial DNA across titration but we were unable to estimate the proportion of bacterial DNA in the unmixed samples using the qPCR data. 
# Using the 16S sequencing data we inferred the proportion of bacterial DNA from the post-exposure sample in each titration. 
# However, the uncertainty and accuracy of the inferrence method is not known resulting in an unaccounted for error source. 
# A better method for estimating the proportion of bacterial DNA in the unmixed samples would increase the accuracy of the error metrics. 
# Alternatively samples that are primarily prokaryotic DNA could be used
```

