---
title: "Microbiome-Scale Mixture Use Demonstration"
author: "Nate Olson"
date: '`r Sys.Date()`'
output:
  bookdown::pdf_document2: 
    toc: FALSE
bibliography: [mgtst.bib, packages.bib]
---

<!-- 
BMC Microbiome Submission https://microbiomejournal.biomedcentral.com/submission-guidelines/preparing-your-manuscript/methodology 
-->

# Introduction
```{r intro, child="Introduction.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
``` 


# Methods  
```{r methods, child="Methods.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
``` 

# Results
## Dataset characteristics  

```{r seqChar, child="seq_data_characteristics.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
```

```{r pipeChar, child="pipeline_characteristics.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
```

## Titration Series Validation 
In order to use information from the unmixed samples to obtain expected count values for the titrations we need to evaluate two assumptions about the mixed samples: 
1. The samples were mixed volumetrically in a $log_2$ dilution series.  
2. The unmixed pre and post exposure samples have the same proportion of bacterial DNA. 
Exogenous DNA was spiked into the unmixed samples prior to mixing and quantified using qPCR to validate the samples were volumetrically mixed according to expectations. 
Total bacterial DNA in the unmixed samples was quantified using a qPCR assay targeting the 16S rRNA gene. 


### Spike-in qPCR results   

```{r ercc, child="ercc_validation.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
```

### Bacterial DNA Concentration 
```{r bacCon, child="bac_con_validation.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
```

### Theta Estimates
```{r thetaEst, child="theta_estimate_results.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
```

## Measurement Assessment  
Next we assessed the 16S rRNA measurement process using our two-sample titration dataset. 
We assessed the qualitative and quanitative nature of the 16S metagenomics measurement process. 
For the qualitative assessment we looked the relative abundance of features only observed in the unmixed samples and titrations. 
For the quantitative assessment we looked the the relative abundance and differential abundance log fold-change estimates. 

### Qualitative Assessment

<!--
Proportion of titration and endpoint specific features that could not be explained by sampling alone. 
-->

```{r qualAnalysis, child="qualitative_assessment_results.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
```

### Quantitative Assessment

<!--
Count Error Rate Results
-->
```{r quantAnalysis, child="relative_abundance_assessment_results.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
```

<!--
logFC Error Rate Results
-->
```{r logFCanalysis, child="logFC_assessment_results.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
```

\pagebreak 

# Discussion

* Sample experimental design  
    - Limitation: number of features with differentially abundant between pre- and post-exposure  

* Bacterial DNA proportion  
    - Limitation: additional uncertainty in expected values  

* Pipeline characterization differences  
    - Number of features per sample and total abundance  
    - DADA2 has higher feature abundance due to fewer features and lower filter rate. Resulting in increased statistical power.

* Why qualitative analysis is pipeline dependent?  
    - What dependency means for 16S gene surveys  
    - DADA2 spurious OTUs a result of higher counts and therefore increased statistical power.  

* Why quantitative analysis is biological replicate dependent?  
    - What dependency means for 16S gene surveys, when does bioinformatic pipeline matter and when does it not matter?  
    - Differences in proportion of bacterial DNA between the pre- and post-exposure samples drives individual specific results.  
    - The proportion of non-prokaryotic DNA in a sample is not taken into considering for nearly all 16S studies.  
    - How do differences impact inferences drawn from statistical analyses?

* Relative abundance  
    - Outliers  
    - Need to summarise across replicates  
    - Noisy data  

* log fold-change - outlier features   
* Relationship between factors impacting quant and qual analysis  

# Conclusions  

* How this dataset can be used to evaluate and characterize bioinformatic pipelines and clustering methods.  
* Given study results  
    - How would you analyze 16S sequencing data assuming current methods?  
    - How would you like to analyze 16S sequencing data?  
    - What are the limitations of current methods?  
    - What would you like to see a clustering method/ pipeline be able to do?   
        + What should be improved?  


\pagebreak  

# Session information 

## Git repo commit information
```{r include = FALSE}
library(tidyverse)
library(git2r)
repo <- repository(path = ".")
last_commit <- commits(repo)[[1]]
```

The current git commit of this file is `r last_commit@sha`, which is on the `r branches(repo)[[1]]@name`  branch and was made by `r last_commit@committer@name` on `r when(last_commit)`. The current commit  message is `r last_commit@summary`. The repository is online at https://github.com/nate-d-olson/mgtst-pub  


## Platform Information  
```{r session_info, warning=FALSE, message=FALSE, echo = FALSE}
s_info <- devtools::session_info()
print(s_info$platform)
```


## Package Versions  
```{r warning=FALSE, message=FALSE, echo = FALSE}
s_info$packages %>% filter(`*` == "*") %>% 
      dplyr::select(-`*`) %>%
      knitr::kable(booktabs = TRUE)
```

\pagebreak

# References