---
title: "Relative Abundance Error Rate"
author: "Nate Olson"
date: '`r Sys.Date()`'
output:
  bookdown::pdf_document2: 
    toc: FALSE
---

```{r relAbuSetup, include=FALSE}
## TODO feature count table
library(ggtree)
library(nlme)
library(ape)
library(multcomp)
library(kableExtra)
library(tidyverse)
library(forcats)
library(stringr)
library(ggpubr)
nb_counts <- readRDS("~/Desktop/nb_counts_titrations.RDS")
pa_summary_anno_df <- readRDS("~/Desktop/pa_summary_anno_df.RDS")
theta_est <- readRDS("~/Desktop/bootstrap_theta_estimates.rds")
```

```{r relAbuMunge, echo = FALSE, message = FALSE, warning = FALSE}
pre_post_prop <- nb_counts %>% 
      ungroup() %>% 
      filter(t_fctr %in% c(0,20)) %>% 
      mutate(end_point = if_else(t_fctr == 0 , "post", "pre")) %>% 
      dplyr::select(-t_fctr) %>% 
      ## setting values to 0 when one or more of the PCR replicates are 0 for titration end-points
      spread(end_point,nb_prop, fill = 0)

prop_inferred <- theta_est %>% 
      filter(pipe == "unclustered") %>% 
      ungroup() %>%
      mutate(t_fctr = factor(t_fctr, levels = c(0:5, 10, 15, 20))) %>% 
      dplyr::select(biosample_id, theta_hat_mean, t_fctr) %>% 
      right_join(nb_counts) %>% right_join(pre_post_prop) %>% 
      filter(t_fctr %in% c(1:5,10,15)) %>% 
      ## Using inferred theta estimates to calculate expected values
      mutate(inferred_prop = post * theta_hat_mean + pre * (1-theta_hat_mean))

## Excluding mix and unmix specific features
## Only including features observed in all or none of the four pre- post- PCR replicates
## Features with relative abundance estimates and expected values less than 1e-5, these are features that we would not expect to consistently observe in a PCR replicate for the given sequencing depth, ~100k 
## Excluding titrations where the inferred theta values are less than 1
pa_filter <- pa_summary_anno_df %>% 
      filter(pa_specific == "unspecific") %>% 
      dplyr::select(biosample_id, pipe, feature_id, full_pre, T00, T20, pa_mixed) %>% 
      filter(T00 %in% c(0,4), T20 %in% c(0,4))

prop_inferred <- prop_inferred %>% 
      right_join(pa_filter) %>% 
      # filter(nb_prop > 1e-5, 
      #        inferred_prop > 1e-5,
      #        theta_hat_mean > 0)
      ## Filtering absed on 1/median library size
      filter(nb_prop > 1/73571,
             inferred_prop > 1/73571,
             theta_hat_mean > 0)


#### Error Rate Calculations
rel_abu_error <- prop_inferred %>% 
      mutate(t_fctr = factor(t_fctr, levels = c(1:5, 10, 15))) %>% 
      mutate(inferred_error = abs(nb_prop - inferred_prop),
             inferred_error_rate = inferred_error/inferred_prop) 

rel_abu_error_summary <-  rel_abu_error %>% 
      group_by(pipe, biosample_id, feature_id) %>% 
      summarise(median_rel_abu = median(nb_prop),
                median_error = median(inferred_error_rate),
                iqr_error = IQR(inferred_error_rate),
                rcov_error = iqr_error/median_error, 
                mean_error = mean(inferred_error_rate),
                var_error = var(inferred_error_rate),
                cov_error = var_error/mean_error) 

### Error rate boxplot and outlier annotation
error_boxplot <- rel_abu_error %>% group_by(pipe, biosample_id, feature_id) %>% 
      summarise(median_error = median(inferred_error_rate)) %>%
      ggplot() + 
      geom_boxplot(aes(x = pipe, y = median_error), outlier.shape = NA) + 
      facet_wrap(~biosample_id, nrow = 1) + 
      theme_bw() + theme(axis.text.x = element_text(angle = 90)) +
      labs(x = "Individual", y = "Median")

## Annotating features as outliers based on boxplot
error_plot_dat <- ggplot_build(error_boxplot)$data[[1]] %>% 
      mutate(pipe = fct_recode(factor(group), 
                               dada2 = "1", 
                               mothur = "2", 
                               qiime = "3",
                               unclustered = "4"),
             biosample_id = fct_recode(PANEL, 
                                       E01JH0004 = "1", 
                                       E01JH0011 = "2", 
                                       E01JH0016 = "3", 
                                       E01JH0017 = "4", 
                                       E01JH0038 = "5"))
outlier_error_dat <- error_plot_dat %>% 
      dplyr::select(ymin, ymax, pipe, biosample_id)

rel_error_outlier_cat <- rel_abu_error_summary %>% 
      left_join(outlier_error_dat) %>% 
      mutate(outlier_cat = if_else(median_error < ymin | median_error > ymax, "outlier","inlier")) 

## Robust COV Analysis
rcov_boxplot <- rel_abu_error_summary %>%
      ggplot() + geom_boxplot(aes(x = pipe, y = rcov_error), outlier.shape = NA) + 
      facet_wrap(~biosample_id, nrow = 1) + 
      theme_bw() + theme(axis.text.x = element_text(angle = 90)) +
      labs(x = "Individual", y = "RCOV") 


## Annotating features as outliers based on boxplot
rcov_plot_dat <- ggplot_build(rcov_boxplot)$data[[1]] %>% 
     mutate(pipe = fct_recode(factor(group), 
                               dada2 = "1", 
                               mothur = "2", 
                               qiime = "3",
                               unclustered = "4"),
             biosample_id = fct_recode(PANEL, 
                                       E01JH0004 = "1", 
                                       E01JH0011 = "2", 
                                       E01JH0016 = "3", 
                                       E01JH0017 = "4", 
                                       E01JH0038 = "5"))
outlier_rcov_dat <- rcov_plot_dat %>% 
      dplyr::select(ymin, ymax, pipe, biosample_id)

rcov_outlier_cat <- rel_abu_error_summary %>% 
      left_join(outlier_rcov_dat) %>% 
      mutate(outlier_cat = if_else(rcov_error < ymin | rcov_error > ymax, "outlier","inlier"))

## Feature-level error summary data frame
rel_error_summary <- rel_error_outlier_cat %>% 
      dplyr::rename(error_cat = outlier_cat) %>% 
      dplyr::select(-ymin, -ymax) %>% 
      left_join(rcov_outlier_cat) %>% 
      dplyr::rename(rcov_cat = outlier_cat)

### Plot Code -----------------------------------------------------------------
## Observed v. Expected Scatter plot
relAbuOvE <- prop_inferred %>% 
      ggplot() + 
      geom_point(aes(x = inferred_prop, y = nb_prop), alpha = 0.15) +
      geom_abline(aes(intercept = 0, slope = 1), color = "darkorange") +
      geom_smooth(aes(x = inferred_prop, y = nb_prop), method = "lm") +
      facet_grid(pipe~biosample_id) +
      scale_y_log10() + scale_x_log10() +
      theme_bw() +
      labs(x = "Expected", 
           y = "Observed")

# ggsave(relAbuOvE, "~/Desktop/quant_exp_vs_obs.png", dpi = 450)

## Median Error Pipeline Comparison
ymin <- ggplot_build(error_boxplot)$data[[1]]$ymin %>% min()
ymax <- ggplot_build(error_boxplot)$data[[1]]$ymax %>% max()
error_boxplot <- error_boxplot + coord_cartesian(ylim = c(ymin, ymax))

# ggsave(error_boxplot, "~/Desktop/quant_bias.png", dpi = 450) 

## RCOV Error Pipeline Comparison
ymin <- ggplot_build(rcov_boxplot)$data[[1]]$ymin %>% min()
ymax <- ggplot_build(rcov_boxplot)$data[[1]]$ymax %>% max()
rcov_boxplot <- rcov_boxplot + coord_cartesian(ylim = c(ymin, ymax))

# ggsave(rcov_boxplot, "~/Desktop/quant_variance.png", dpi = 450) 
```


```{r relAbuStats, echo = FALSE, message = FALSE, warning = FALSE}
## Bias - Error rate
error_fit_dat <- rel_error_outlier_cat %>% 
      ungroup() %>% 
      filter(outlier_cat == "inlier", pipe != "unclustered") %>% 
      mutate(pipe = factor(pipe))

# Fitting mixed effects model with individual as the fixed effect
error_fit <- nlme::lme(median_error ~ pipe, random =  ~ 1 | biosample_id, 
                 data = error_fit_dat)  

# Pipe error estimates
dada_error <- error_fit$coefficients$fixed['(Intercept)']
mothur_error <- dada_error + error_fit$coefficients$fixed['pipemothur']
qiime_error <- dada_error + error_fit$coefficients$fixed['pipeqiime']

# Post-hoc test to check for pipeline differences 
## based on fit pipeline estimates are all negative, using alternative greater to determine which pipelines are closer to zero
error_post_hoc <- glht(error_fit, linfct=mcp(pipe="Tukey"), alternative = "less") 

error_tukey_p <- summary(error_post_hoc)$test$pvalues 
error_tukey_t <- summary(error_post_hoc)$test$tstat

# Checking whether indiviudal or pipeline contributes more to the overal vaiance
error_var <- ape::varcomp(error_fit)

## Variance - RCOV
rcov_fit_dat <- rcov_outlier_cat %>% 
      ungroup() %>% 
      filter(outlier_cat == "inlier", pipe != "unclustered") %>% 
      mutate(pipe = factor(pipe))

# Fitting mixed effects model with individual as the fixed effect
rcov_fit <- nlme::lme(rcov_error ~ pipe, random =  ~ 1 | biosample_id, 
                 data = rcov_fit_dat)

# Pipe RCOV estimates
dada_rcov <- rcov_fit$coefficients$fixed['(Intercept)']
mothur_rcov <- dada_rcov + rcov_fit$coefficients$fixed['pipemothur']
qiime_rcov <- dada_rcov + rcov_fit$coefficients$fixed['pipeqiime']

# Checking whether indiviudal or pipeline contributes more to the overal vaiance
rcov_var <- ape::varcomp(rcov_fit)  
```

```{r errorMeanVarTidy, echo = FALSE, message = FALSE, warning = FALSE}
rel_abu_count_df <- readRDS("~/Desktop/raw_counts.RDS") %>%       
      group_by(pipe, id) %>% 
      mutate(total_count = sum(count)) %>% 
      right_join(rel_error_summary)

unmix_counts <- rel_abu_count_df %>% 
      filter(t_fctr %in% c(0, 20)) %>% 
      mutate(rel_abu = count/total_count) %>% 
      group_by(pipe, biosample_id, feature_id, t_fctr) %>% 
      summarise(mean_abu = mean(rel_abu),
                var_abu = var(rel_abu),
                range_abu = max(rel_abu) - min(rel_abu)) %>%
      mutate(var_mean = var_abu/mean_abu) %>% 
      left_join(rel_error_summary)
```

```{r treeFuns, echo = FALSE, message = FALSE, warning = FALSE}
get_fig_tree <- function(tree, features_to_compare, tax){
      # trimming tree
      tips_to_drop <- tree$tip.label[!(tree$tip.label %in% features_to_compare)] 
      trimmed_tree <- drop.tip(tree,tip = tips_to_drop) 
      
      ## Getting Taxonomic assignment data
      tree_tax <- tax %>% 
            as.data.frame() %>% 
            dplyr::rename(id = OTUname) %>% 
            filter(id %in% trimmed_tree$tip.label)  
      
      # Adding taxonomic groupings to tree 
      trimmed_tree <- groupOTU(trimmed_tree, 
                               split(tree_tax$id, tree_tax$Rank2), 
                               group_name = "Phylum")
      
      trimmed_tree

}

get_feats_to_compare <- function(error_df, pipe){
      metrics <- rel_error_summary %>% filter(pipe == pipe)

      ## Most of the features were used in this analysis for 1 of the five individuals. 
      ## To compare feature performance across individuals were only looked at features used for at least three of the individuals
     metrics %>% 
            group_by(feature_id) %>% 
            summarise(count = n()) %>% 
            filter(count > 2) %>% 
            .$feature_id
}

get_error_mat <- function(error_df, features_to_compare, tip_labels){
      ## Feature-level error data
      error <- error_df %>% 
            filter(feature_id %in% features_to_compare) %>% 
            ungroup() %>% 
           dplyr::select(biosample_id, feature_id, median_error) %>% 
            mutate(median_error = log2(abs(median_error)+ 1)) %>% 
            spread(biosample_id, median_error)
      
      
      ## Reordering and converting to a matrix
      error_mat <- error %>% 
            right_join(data.frame(feature_id = tip_labels)) %>% 
            as.data.frame() %>% 
            column_to_rownames(var = "feature_id") %>% 
            as.matrix()
      colnames(error_mat) <- str_replace(colnames(error_mat), "E01JH00","ERR\n")
      
      ## Feature-level RCOV data
      rcov <- error_df %>% 
            filter(feature_id %in% features_to_compare) %>% 
            ungroup() %>% 
           dplyr::select(biosample_id, feature_id, rcov_error) %>% 
            mutate(rcov_error = log10(abs(rcov_error)+ 1)) %>% 
            spread(biosample_id, rcov_error)
      
      ## Reordering cat data frame and converting to a matrix
      rcov_mat <- rcov %>% 
            right_join(data.frame(feature_id = tip_labels)) %>% 
            as.data.frame() %>% 
            column_to_rownames(var = "feature_id") %>% 
            as.matrix()
      colnames(rcov_mat) <- str_replace(colnames(rcov_mat), "E01JH00","RCOV\n")
      
      ## Generating a single matrix
      filler_mat <- matrix(NA, ncol = 1, nrow = nrow(rcov_mat))
      colnames(filler_mat) <- ""
      
      ## Return Matrix Data
      cbind(error_mat, filler_mat, rcov_mat)
}

get_fig_tree_dat <- function(tree, tax, error_df, pipe){
      ## Generating trimmed tree and matrix for plot
      features_to_compare <- get_feats_to_compare(error_df = error_df, 
                                            pipe = pipe) 
      
      trimmed_tree <- get_fig_tree(tree = tree, 
                                features_to_compare = features_to_compare,
                                tax = tax)
      
      mat <- get_error_mat(error_df = error_df, 
                                features_to_compare = features_to_compare,
                                tip_labels = trimmed_tree$tip.label)
      
      list(tree = trimmed_tree, dat = mat)
}
```

```{r relAbuErrorTrees, echo = FALSE, message = FALSE, warning = FALSE}
### DADA2 ----------------------------------------------------------------------
## Tree data
dada_tree <- readRDS("~/Projects/mgtst_pipelines/dada2/dada_tree_GTR.rds") %>% .$tree


## Getting Taxonomic assignment data
dada_tax <- readRDS("../mgtst_pipelines/dada2/dada_mrexp.rds") %>% 
      .@featureData %>% .@data

dada_tree_dat <- get_fig_tree_dat(tree = dada_tree, 
                                 tax = dada_tax, 
                                 error_df = rel_error_summary, 
                                 pipe = "dada2")

dada_fig_tree <- ggtree(dada_tree_dat$tree, aes(color = Phylum), ladderize = TRUE) %>%
      gheatmap(dada_tree_dat$dat, width=0.5, font.size=2) +
      scale_fill_gradient(low = "#56B1F7", high = "#132B43", na.value = "white") +
      theme(legend.position = "bottom")

### QIIME ----------------------------------------------------------------------
## Tree data
qiime_tree <- read.tree("~/Projects/mgtst_pipelines/qiime/otus_uc_fast/rep_set.tre")


## Getting Taxonomic assignment data
qiime_tax <- readRDS("../mgtst_pipelines/qiime/qiime_mrexp.rds") %>% 
      .@featureData %>% .@data %>%
      mutate(Rank2 = str_replace(Rank2, "p__",""))
      
      

qiime_tree_dat <- get_fig_tree_dat(tree = qiime_tree, 
                                 tax = qiime_tax, 
                                 error_df = rel_error_summary, 
                                 pipe = "qiime")

qiime_fig_tree <- ggtree(qiime_tree_dat$tree, aes(color = Phylum), ladderize = TRUE) %>%
      gheatmap(qiime_tree_dat$dat, width=0.5, font.size=2) +
      scale_fill_gradient(low = "#56B1F7", high = "#132B43", na.value = "white") +
      theme(legend.position = "bottom")

### MOTHUR ---------------------------------------------------------------------
## Tree
seq_id_df <- read_lines("~/Projects/mgtst_pipelines/mothur/mgtst.trim.contigs.good.unique.good.filter.unique.precluster.pick.opti_mcc.unique_list.0.03.rep.fasta") %>% grep(pattern = ">", value = TRUE) %>%
      str_replace("\\|.*","") %>%
      str_replace(">","") %>%
      enframe(name = "X", value = "seq_id") %>%
     dplyr::select(-X) %>%
      separate(seq_id, into = c("seq_id", "feature_id"),sep = "\t")


mothur_tree <- read.tree("~/Projects/mgtst_pipelines/mothur/mgtst.trim.contigs.good.unique.good.filter.unique.precluster.pick.opti_mcc.unique_list.0.03.rep.tre")

seq_tree_ids <- data_frame(seq_id = mothur_tree$tip.label) %>%
      left_join(seq_id_df)

mothur_tree$tip.label <- seq_tree_ids$feature_id


## Getting Taxonomic assignment data
mothur_tax <- readRDS("../mgtst_pipelines/mothur/mothur_mrexp.rds") %>% 
      .@featureData %>% .@data
      
      

mothur_tree_dat <- get_fig_tree_dat(tree = mothur_tree, 
                                 tax = mothur_tax, 
                                 error_df = rel_error_summary, 
                                 pipe = "qiime")

mothur_fig_tree <- ggtree(mothur_tree_dat$tree, aes(color = Phylum), ladderize = TRUE) %>%
      gheatmap(mothur_tree_dat$dat, width=0.5, font.size=2) +
      scale_fill_gradient(low = "#56B1F7", high = "#132B43", na.value = "white") +
      theme(legend.position = "bottom")
```

```{r relAbuError, fig.cap = "(A) Expected and observed count relationship. Orange line indicates expected 1-to-1 relationship. Blue line a smoothed regression line of the observed and expected value relationship. Features with observed and expected relative abundance values less than 1/median library size were excluded from the figure. Distribution of feature-level relative abundance (B) median error rates and (C) robust coefficient of variation (RCOV) by individual and pipeline.", echo = FALSE, warning=FALSE, message = FALSE, fig.height = 8}
ggarrange(relAbuOvE, 
          error_boxplot + rremove("x.text"), 
          rcov_boxplot, 
          labels = c("A", "B", "C"),
          align = "v", ncol = 1, nrow = 3, heights = c(0.45,0.2,0.35))
```

```{r relAbuErrorTbl, echo = FALSE, warning = FALSE, message = FALSE}
rel_abu_error_summary %>% 
      group_by(pipe, biosample_id) %>% 
      summarise(med_med_error = median(median_error), 
                min_med_error = min(median_error), 
                max_med_error = max(median_error),
                med_rcov_error = median(rcov_error), 
                min_rcov_error = min(rcov_error), 
                max_rcov_error = max(rcov_error)) %>% 
      ## Value ranges
      # mutate(Median = paste0(round(med_med_error,2), " (",
      #                        round(max_med_error,2), "-",
      #                        round(min_med_error,2),")"),
      #        RCOV = paste0(round(med_rcov_error,2), " (",
      #                      round(max_rcov_error,2), "-",
      #                      round(min_rcov_error,2),")")) %>% 
      # Max values only 
      mutate(Median = max_med_error, RCOV = max_rcov_error) %>% 
     dplyr::select(pipe, biosample_id, Median, RCOV) %>%
      dplyr::rename(Pipeline = pipe, Individual = biosample_id) %>%
      gather("Metric","value", -Pipeline, -Individual) %>% 
      spread(Individual, value) %>% 
      arrange(Metric) %>% 
     dplyr::select(Metric, Pipeline, E01JH0004, E01JH0011, E01JH0016, E01JH0017, E01JH0038) %>% 
      knitr::kable(booktabs = TRUE, caption = "Maximum feature-level error rate bias (median error rate) and variance (robust COV) by pipeline and individual.", digits = 2) %>% 
      collapse_rows(columns = 1)
```

```{r errorOut, warning = FALSE, message = FALSE, echo = FALSE, fig.cap = "Feature-level median relative abundance and variance-mean relationship for the unmixed sample relative abundance not correlated with feature-level error rate and RCOV. Relationship between median relative abundance to the feature-level  (A) median error rate and (C) error RCOV. Relationship between unmixed sample relative abundance variance/mean to the feature-level (B) median error rate and (D) error RCOV."}

### Median relative abundance
error_rel_abu <- rel_error_summary %>% 
      ggplot(aes(x = median_rel_abu, y = median_error)) + 
      geom_point(aes(fill = error_cat), shape = 21) + 
      stat_cor(method = "spearman", label.y = 22) + 
      theme_bw() + 
      scale_x_log10() +       
      labs(x = "Relative Abundance", y = "Error Rate", fill = "Boxplot Outlier")

## excluding features with rcov_error == 0 - artifacts and not real....
rcov_rel_abu <- rel_error_summary %>% 
      filter(rcov_error != 0) %>% 
      ggplot(aes(x = median_rel_abu, y = rcov_error)) + 
      geom_point(aes(fill = rcov_cat), shape = 21) + 
      stat_cor(method = "spearman", label.y = 4.5) + 
      theme_bw() + 
      scale_x_log10() + scale_y_log10() +    
      labs(x = "Relative Abundance", y = "Error RCOV", fill = "Boxplot Outlier")


### Variance/Mean relationship
unmix_filt <- unmix_counts %>% 
      filter(!is.na(rcov_cat), !is.na(var_mean), var_abu != 0, mean_abu != 0, rcov_error != 0)

error_var_mean <- ggplot(unmix_filt, aes(x = var_mean, y = median_error)) + 
      geom_point(aes(fill = error_cat), shape = 21) +
      stat_cor(method = "spearman", label.y = 22) + 
      theme_bw() + scale_x_log10() +       
      labs(x = "Variance/Mean", y = "Error Rate", fill = "Boxplot Outlier")

## excluding features with rcov_error == 0 - artifacts and not real....
rcov_var_mean <- ggplot(unmix_filt, aes(x = var_mean, y = rcov_error)) +
      geom_point(aes(fill = rcov_cat), shape = 21) +
      stat_cor(method = "spearman", label.y = 4.5) + 
      theme_bw() + scale_x_log10() + scale_y_log10() +    
      labs(x = "Variance/Mean", y = "Error RCOV", fill = "Boxplot Outlier")

ggarrange(error_rel_abu, 
          error_var_mean, 
          rcov_rel_abu, rcov_var_mean, 
          ncol = 2, nrow = 2, align = "hv",labels = "AUTO",
      common.legend = TRUE, legend = "bottom")
```

```{r relAbuTreeFig, fig.cap = "Phylogenetic analysis of feature-level relative abundance error metrics, as heatmaps across individuals. Subplots for individual pipelines (A) DADA2, (B) QIIME, (C) Mothur. Feature phylum assignment is indicated by branch color.", fig.height = 10, message = FALSE, warning = FALSE, echo = FALSE}
ggarrange(dada_fig_tree, qiime_fig_tree, mothur_fig_tree, 
          labels = "AUTO", ncol = 1, nrow = 3)
```

Overall aggreement between the inferred and observed relative abundance was high for all individuals and bioinformatic pipelines (Fig. \@ref(fig:relAbuError)A). 
The pre- and post-exposure estimated relative abundance and inferred $\theta$ values were used to calculate titration and feature level error rates. 
To prevent over-fitting, $\theta$ estimates for the unclustered pipeline were used to calculate the error rates for the other three pipelines.  
Only features observed in all pre- and post-exposure PCR replicates and pre- and post-exposure specific features were included in the analysis (Table \@ref(tab:relAbuFeatTbl)). 
Pre- and post-exposure specific features were defined as present in all four PCR replicates of the pre-exposure or post-exposure PCR replicates, respectively, but none of the PCR replicates for the other unmixed sample.
There is lower confidence in the relative abundance of a feature in the pre- or post-exposure unmixed samples when the feature is observed in some of the 4 PCR replicates, therefore these features were not included in the error analysis.
The deviation from the expected value on the low end varies by biological replicate and pipeline. 

Next we evaluated the feature-level quantitative accuracy of the relative abundance values by comparing the distribution of the median error and robust coefficient of variation ($RCOV=(IQR)/|median|$)for the relative abundance error rate (Fig. \@ref(fig:relAbuError)). 
Feature-level median error rates and RCOV were compared across pipelines and individuals using a mixed effects model. 
Large error rates and RCOV values were observed for all pipelines (Table  \@ref(tab:relAbuErrorTbl)). 
Features with large error rates, defined as $1.5\times IQR$ from the median, were excluded from the analysis to prevent outliers from biasing the comparison. 
The mean feature-level error rate by pipeline was negative for all three pipelines (DADA2 `r round(dada_error,2)`, Mothur `r round(mothur_error,2)`, and QIIME `r round(qiime_error,2)`). 
Multiple comparisons test (Tukey) was used to test for significant differences in feature-level error between pipelines. 
A one-sided alternative hypothesis to determine which pipelines had a smaller, closer to zero, feature-level error rate. 
The Mothur and DADA2 mean feature-level error rates were closer to zero and signficantly different from the QIIME pipeline, (Mothur v. QIIME t = `r  round(error_tukey_t['qiime - mothur'],2)`, p = `r round(error_tukey_p[error_tukey_t == error_tukey_t['qiime - mothur']],2)`; DADA2 v. QIIME t = `r  round(error_tukey_t['qiime - dada2'],2)`, p = `r round(error_tukey_p[error_tukey_t == error_tukey_t['qiime - dada2']],2)`). 
Though the Mothur and DADA2 mean feature-level error rates were not significantly different from eachother, (t = `r  round(error_tukey_t['mothur - dada2'],2)`, p = `r round(error_tukey_p[error_tukey_t == error_tukey_t['mothur - dada2']],2)`).
Unlike feature-level error rates, large RCOV was observed for all individuals and pipelines (Table \@ref(tab:relAbuErrorTbl)). 
Outlier values were also excluded from the RCOV analysis. 
The feature-level RCOV was not significantly different between pipelines, Mothur = `r round(mothur_rcov,2)`, QIIME = `r round(qiime_rcov,2) ` and DADA2 = `r round(dada_rcov)` (Fig. \@ref(fig:relAbuError)C). 

In an attempt to identify feature characteristics that could be attributed to poor performance the feature-level error-rate and RCOV were compared to the unmixed sample relative abundance and relative abundance across the titrations. 
Additionally, feature-level relative abundance error metrics were compared across individuals to try and identify any relationship between poor performance and phylogeny. 
Outlier feature-level error-rates had lower median relative abundance, but not RCOV (Fig. \@ref(fig:errorOut)). 
The median relative abundance was significantly lower for features identified as outliers based on the feature-level error rate, but only when considering features with positive error rates. 
This is likely due to the more extreme error rates all being positive. 
For the RCOV the feature-level median relative abundance values were not significantly different between the outlier and non-outlier features.  
The error rate is dependent on the accuracy of the relative abundance estimates for the unmixed pre- and post-exposure samples. 
The feature-level median error-rate and RCOV was compared to the the unmixed sample variance/mean relative abundance to determine if extreme error-rate and RCOV values could be attributed to variability in relative abundance between PCR replicates for the unmixed samples. 
The variance/mean for the unmixed samples was lower for the outliers compared to the non-outliers for both the feature-level error rate and RCOV. 
Investigation of relationship between phylogeny and feature-level relative abundance error metrics (Fig. \@ref(fig:relAbuTreeFig)). 
Only features included in the relative abundance error analysis for at least three of the five individuals were included in the figure. 
No clear relationship between feature-level error rate or RCOV and phylogeny of representative sequences. 
