---
title: "Sequencing Data Quality Assessment"
author: "Nate Olson"
date: '`r Sys.Date()`'
output:
  bookdown::pdf_document2: 
    toc: FALSE
---

```{r seq_setup, warning=FALSE, message=FALSE, echo = FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
library(Rqc)
library(ProjectTemplate)
load.project()

project_dir <- "../16S_etec_mix_study/"
pipeline_dir <- "../mgtst_pipelines/"
``` 




```{r fq_data_filesp, warning=FALSE, message=FALSE, echo = FALSE}
fastq_dir <- file.path(project_dir, "data/")
dat_files <- list.files(path = fastq_dir,pattern = "001.fastq.gz$",
                           recursive = TRUE, full.names = TRUE)
seq_ds_id <- dat_files %>% str_split("/") %>% flatten_chr() %>% 
    grep(pattern = "fastq.gz", .,value = TRUE) %>%
    str_replace(".fastq.gz","") %>% paste0("Fq_",.) %>% 
    str_replace("-",".")
names(dat_files) <- seq_ds_id
```

```{r loadQAp, warning=FALSE, message=FALSE, echo = FALSE}
# from script but not saved
read_groups <- rep(NA,length(dat_files))
read_groups[grepl("/1-.*_R1", dat_files)] <- "plate1_R1"
read_groups[grepl("/1-.*_R2", dat_files)] <- "plate1_R2"
read_groups[grepl("/2-.*_R1", dat_files)] <- "plate2_R1"
read_groups[grepl("/2-.*_R2", dat_files)] <- "plate2_R2"

qa_list <- readRDS("data/rqcQA_list.rds")
```

```{r metadat, warning=FALSE, message=FALSE, echo = FALSE}
# Tidy data

## Run 
grp_df <- data_frame(read_group = read_groups, 
                     seq_ds_id, 
                     filename = basename(dat_files)) %>%
    separate(read_group,c("plate","Read")) %>% 
    mutate(ill_id = str_replace(filename, "_.*",""))

## read count data 
qa_file_info <- perFileInformation(qa_list) %>% 
    select(-format,-path)

## study metadata
meta_df <- sampleSheet %>% 
    mutate(pos_ns = str_replace(pos, "_",""),
           ill_id = paste(pcr_16S_plate, pos_ns, sep = "-")) %>% 
    filter(seq_lab == "JHU", barcode_lab == "JHU") %>% 
    mutate(pcr_16S_plate = as.character(pcr_16S_plate)) %>% 
    left_join(grp_df) %>% left_join(qa_file_info)

## Library sizes
lib_size <- meta_df %>% filter(Read == "R1", biosample_id != "NTC", reads != 2700) %>% .$reads
```

Quality assessment of sequencing run summarizing number of reads per sample.
Two barcoded experimental samples have less than 35,000 reads (Fig. \@ref(fig:readCount)). 
The rest of the samples with less than 35,000 reads are no template PCR controls (NTC). 
Excluding the one failed reaction with 2,700 reads and the NTCs, the total range in the observed number of sequences per samples is `r min(lib_size)` to `r max(lib_size)` reads with a median library size of `r median(lib_size)`. 
For the expected overlap region, based on primer positions and read lengths (16S PCR fig), the forward read has consistently higher base quality scores relative to the reverse read with a narrow overlap region with high base quality scores for both forward and reverse reads (Fig. \@ref(fig:lowcycleq)).

```{r cycle_metrics, warning=FALSE, message=FALSE, echo = FALSE}
## amplicon position __Cycle Level Metrics__
amp_pos_df <- data_frame(cycle = rep(1:300, 2),
                         Read = rep(c("R1","R2"), each = 300),
                         amp_pos = c(1:300,c(460 - 1:300)))

qa_cycle_q_df <- qa_list %>% map_df(perCycleQuality) %>%
    as_data_frame() %>%
    mutate(cycle = as.numeric(as.character(cycle))) %>%
    filter(count != 0) %>% # not sure if this impacts the smoothing function ...
    left_join(meta_df) %>% left_join(amp_pos_df)
```

```{r tableCount, warning=FALSE, message=FALSE, echo = FALSE}
# meta_df %>% mutate(exp_ntc = ifelse(biosample_id == "NTC", "NTC","EXP")) %>% 
#     filter(Read == "R1", ill_id != "1-F9") %>% 
#     group_by(biosample_id) %>% 
#     summarise(median = median(reads),
#               min_lib_size = min(reads),
#               max_lib_size = max(reads)) %>% 
#     kable(caption = "Summary statistics for number of reads per sample by biological replicate.", booktab = TRUE)
```

```{r readCount, fig.cap="Distribution in the number of reads per barcoded sample (Library Size) by individual. Dashed horizontal line indicates 35,000 reads per barcoded sample.", fig.lp = "fig:", warning=FALSE, message=FALSE, echo = FALSE }
meta_df %>% filter(Read == "R1") %>% unique() %>% 
    ggplot() + 
        geom_boxplot(aes(x = biosample_id, y = reads)) + 
      scale_y_log10() + 
        geom_hline(aes(yintercept = 35000), color = "grey60", linetype = 2) +
        theme_bw() + labs(y = "Library Size", 
                          x = "Individual")
```



```{r lowcycleq, fig.cap = "Smoothing spline of the base quality score by sequencing cycle. Vertical lines indicate approximate overlap region between forward and reverse reads.", echo = FALSE, warning = FALSE, message = FALSE}
qa_cycle_q_df %>% 
      filter(biosample_id != "NTC", ill_id != "1-F9") %>%
      ggplot(aes(x = amp_pos, y = score)) +
    geom_vline(aes(xintercept = 300), color = "grey60") +
    geom_vline(aes(xintercept = 160), color = "grey60") +
    # geom_smooth(aes(weight = count, color = biosample_id, group = filename)) +
    geom_smooth(aes(weight = count,group = Read)) +
    # geom_vline(aes(xintercept = 280), color = "grey40", linetype = 2) +
    theme_bw() +
    scale_x_continuous(breaks = c(0,150,300, 450)) +
    labs(x = "Amplicon Position",
         y = "Base Quality Score", color = "Individual") +
    theme(legend.position = "bottom")
```
