---
title: "Sequencing Data Quality Assessment"
author: "Nate Olson"
date: '`r Sys.Date()`'
output:
  bookdown::pdf_document2: 
    toc: FALSE
---

```{r seq_setup, warning=FALSE, message=FALSE, echo = FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
library(Rqc)
library(ProjectTemplate)
load.project()

project_dir <- "../16S_etec_mix_study/"
pipeline_dir <- "../mgtst_pipelines/"
``` 




```{r fq_data_filesp, warning=FALSE, message=FALSE, echo = FALSE}
fastq_dir <- file.path(project_dir, "data/")
dat_files <- list.files(path = fastq_dir,pattern = "001.fastq.gz$",
                           recursive = TRUE, full.names = TRUE)
seq_ds_id <- dat_files %>% str_split("/") %>% flatten_chr() %>% 
    grep(pattern = "fastq.gz", .,value = TRUE) %>%
    str_replace(".fastq.gz","") %>% paste0("Fq_",.) %>% 
    str_replace("-",".")
names(dat_files) <- seq_ds_id
```

```{r loadQAp, warning=FALSE, message=FALSE, echo = FALSE}
# from script but not saved
read_groups <- rep(NA,length(dat_files))
read_groups[grepl("/1-.*_R1", dat_files)] <- "plate1_R1"
read_groups[grepl("/1-.*_R2", dat_files)] <- "plate1_R2"
read_groups[grepl("/2-.*_R1", dat_files)] <- "plate2_R1"
read_groups[grepl("/2-.*_R2", dat_files)] <- "plate2_R2"

qa_list <- readRDS("data/rqcQA_list.rds")
```

```{r metadat, warning=FALSE, message=FALSE, echo = FALSE}
# Tidy data

## Run 
grp_df <- data_frame(read_group = read_groups, 
                     seq_ds_id, 
                     filename = basename(dat_files)) %>%
    separate(read_group,c("plate","Read")) %>% 
    mutate(ill_id = str_replace(filename, "_.*",""))

## read count data 
qa_file_info <- perFileInformation(qa_list) %>% 
    select(-format,-path)

## study metadata
meta_df <- sampleSheet %>% 
    mutate(pos_ns = str_replace(pos, "_",""),
           ill_id = paste(pcr_16S_plate, pos_ns, sep = "-")) %>% 
    filter(seq_lab == "JHU", barcode_lab == "JHU") %>% 
    mutate(pcr_16S_plate = as.character(pcr_16S_plate)) %>% 
    left_join(grp_df) %>% left_join(qa_file_info)
```

Quality assessment of sequencing run summarizing number of reads per sample.
Two barcoded experimental samples have less than 50,000 reads \@ref(fig:readCount). 
The rest of the samples with less than 50,000 reads are negative PCR controls (NTC). 
Sample E01JH0016 titration 5 position F9 of plate 1 initial 16S PCR failed.  
Excluding the one failed reaction the total range in the observed number of sequences per sample is approximately 40,000 to 150,000 reads. 
The low variabiltiy in number of reads per sample suggests that normalization methods are less likely to impact the results compared other datasets with larger variability in the number of reads per sample.  

__TODO__ Figure out why E01JH0011 titration 3 position D2 plate 2 is also low, look at picogreen post normalization data. 


```{r tableCount, warning=FALSE, message=FALSE, echo = FALSE}
meta_df %>% mutate(exp_ntc = ifelse(biosample_id == "NTC", "NTC","EXP")) %>% 
    filter(Read == "R1", ill_id != "1-F9") %>% 
    group_by(biosample_id) %>% 
    summarise(median = median(reads),
              min_lib_size = min(reads),
              max_lib_size = max(reads)) %>% 
    kable(caption = "Summary statistics for number of reads per sample by biological replicate.", booktab = TRUE)
```


```{r readCount, fig.cap="Number of reads per barcoded sample (Library Size), by read direction (X-facet) and replicate 16S PCR plate (Y-facet). Vertical line indicates 50,000 reads per barcoded sample.", fig.lp = "fig:", warning=FALSE, message=FALSE, echo = FALSE }
meta_df %>% # filter(Read == "R1") %>%
    ggplot() + 
        geom_histogram(aes(x = reads), position = "dodge") + 
        facet_grid(plate~.) + scale_x_log10() + 
        geom_rug(aes(x = reads, color = biosample_id)) +
        geom_vline(aes(xintercept = 50000), color = "grey80") +
        theme_bw() + labs(x = "Library Size (log10)", 
                          y = "Count", color = "Sample ID") +
        theme(legend.position = "bottom")
```

