---
title: "Assessing 16S marker gene survey data analysis methods using mixtures of human stool sample DNA extracts."
shorttitle: "Abundance Assessment"
authorrunning: Olson et al
thanks: |
    Grants or other notes about the article that should go on the front
    page should be placed here. General acknowledgments should be placed at the
    end of the article.

authors:
- name: ND Olson
  email: nolson@nist.gov
- name: MS Kumar
  email: smuthiah@umiacs.umd.edu 
- name: Shan Li
  email: sli1@epi.umaryland.edu
- name: S Hao
  email: shao4@jhu.edu
- name: D Catoe
  email: david.catoe@nist.gov
- name: W Timp
  email: wtimp@jhu.edu
- name: ML Salit
  email: msalit@stanford.edu
- name: OC Stine
  email: CSTINE@som.umaryland.edu
- name: H Corrada Bravo
  email: hcorrada@umiacs.umd.edu

institute:
- auth: ND Olson \and ML Salit
  address: Material Measurement Laboratory National Institute for Standards and Technology
- auth: ND Olson \and MS Kumar \and H Corrada Bravo
  address: University of Maryland, College Park
- auth: S Hao \and W Timp
  address: Department of Biomedical Engineering, Johns Hopkins University
- auth: OC Stine
  address: School of Medicine, University of Maryland



keywords:
- 16S rRNA gene
- assessment
- bioinformatic pipeline
- normalization
- differential abundance

output:
      # bookdown::word_document2:
      bookdown::pdf_document2:
            # template: springer_template.tex
            toc: FALSE
            keep_tex: TRUE
            citation_package: 'biblatex'
bibliography: "abundanceAssessment.bib"
---
```{r figSetup, echo = FALSE}
## setting DPI to 300 for all figures
knitr::opts_chunk$set(dpi = 300) 
```


# Background {#intro}
```{r intro, child="docs/Introduction.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
```


# Results
## Two-Sample Titration Design
```{r mixDesign, child="results/mixture_design.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
```

## Dataset characteristics
```{r seqChar, child="results/seq_data_characteristics.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
```

## Titration Series Validation
To validate the two-sample titration dataset for use in abundance assessment we evaluated two assumptions about the titrations:
1. The samples were mixed volumetrically in a $log_2$ dilution series according to the mixture design.
2. The unmixed PRE and POST samples have the same proportion of prokaryotic DNA.
To validate the sample volumetric mixing exogenous DNA was spiked into the unmixed samples before mixing and quantified using qPCR .
To evaluate if the PRE and POST samples had the same proportion of prokaryotic DNA total prokaryotic DNA in the titrations samples was quantified using a qPCR assay targeting the 16S rRNA gene.


### Spike-in qPCR results
```{r ercc, child="results/ercc_validation.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
```

### Bacterial DNA Concentration
```{r bacCon, child="results/bac_con_validation.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
```

### Theta Estimates
```{r thetaEst, child="results/theta_estimate_results.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
```

## Measurement Assessment
Next, we assessed the qualitative and quantitative nature of 16S rRNA measurement process using our two-sample titration dataset.
For the qualitative assessment, we analyzed the relative abundance of features only observed in the unmixed samples or titrations which are not expected given the titration experimental design.
The quantitative assessment evaluated relative and differential abundance estimates.

### Qualitative Assessment

<!--
Proportion of titration and endpoint specific features that could not be explained by sampling alone.
-->

```{r qualAnalysis, child="results/qualitative_assessment_results.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
```

### Quantitative Assessment
<!-- __Quant Assessment Intro__ -->
<!--
Count Error Rate Results
-->
```{r quantAnalysis, child="results/relative_abundance_assessment_results.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
```

<!--
logFC Error Rate Results
-->
```{r logFCanalysis, child="results/logFC_assessment_results.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
```

# Discussion
```{r discussion, child="docs/Discussion.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
```

# Conclusions
```{r conclusions, child="docs/Conclusions.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
```


# Methods {#methods}
```{r methods, child="docs/Methods.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
```

# Acknowledgements  
This work was partially supported by National Institutes of Health (NIH) [NIH R01HG005220 to H.C.B.]
The authors would like to thank Mihai Pop, Scott Pine, Justin Zook, and Prachi Kulkarni for feedback on manuscript drafts. 
Opinions expressed in this paper are the authors and do not necessarily reflect the policies and views of NIST, or affiliated venues. Certain commercial equipment, instruments, or materials are identified in this paper in order to specify the experimental procedure adequately. Such identification is not intended to imply recommendations or endorsement by NIST, nor is it intended to imply that the materials or equipment identified are necessarily the best available for the purpose. Official contribution of NIST; not subject to copyrights in USA.

# References
