---
title: "Assessing 16S marker gene survey data analysis methods using mixtures of human stool sample DNA extracts."
shorttitle: "Abundance Assessment"
authorrunning: Olson et al
thanks: |
    Grants or other notes about the article that should go on the front
    page should be placed here. General acknowledgments should be placed at the
    end of the article.

authors:
- name: ND Olson
  email: nolson@nist.gov
- name: MS Kumar
- name: S Hao
- name: W Timp
- name: ML Salit
- name: OC Stine
- name: H Corrada Bravo

institute:
- auth: ND Olson \and ML Salit
  address: Material Measurement Laboratory National Institute for Standards and Technology
- auth: ND Olson \and MS Kumar \and H Corrada Bravo
  address: University of Maryland, College Park
- auth: S Hao \and W Timp
  address: Department of Biomedical Engineering, Johns Hopkins University
- auth: OC Stine
  address: School of Medicine, University of Maryland



keywords:
- 16S rRNA
- assessment
- bioinformatic pipeline
- normalization
- differential abundance


abstract: |
      We compared 16S rRNA bioinformatic pipeliens and downstream analysis methods using a novel dataset derived from mixing human stool samples. Analysis of 16S rRNA marker-gene surveys, used to characterize prokaryotic microbial communities, may be performed by numerous bioinformatic pipelines and downstream analysis methods. However, there is only limited guidance on how to decide between appropriate methods from simulation studies or benchmark studies with limited microbial community complexity. Appropriate data sets and statistics for assessing these methods are needed. We developed a mixture dataset that provides real data complexity and an expected value for assessing 16S rRNA bioinformatic pipelines and downstream analysis methods. We used samples collected from participants in an Enterotoxigenic _Escherichia coli_ (ETEC) vaccine trial participants. A two-sample titration mixture design was used where DNA from stool samples prior to ETEC exposure was titrated into stools samples collected after exposure, in effect diluting the amount of ETEC in the mixed sample. The sequencing data were processed using multiple bioinformatic pipelines, i) DADA2 a sequence inference method, ii) Mothur a _de novo_ clustering method, and iii) QIIME with open-reference clustering. The pipelines varied in the number of features and proportion of reads passing quality control but had similar levels of sparsity. The mixture dataset was used to qualitatively and quantitatively assess the count tables generated using the pipelines. Statistical tests were used to determine if features only present in unmixed samples and titrations, _unmixed_- and _titration_-specific features which should not be found given our experimental design, had abundances that could be explained by sampling effects alone. For Mothur and QIIME less than 5\% of  _unmixed_- and _titration_-specific feature abundance could not be explained by sampling alone whereas for DADA2 greater than 50\% of _unmixed_-specific features and 10\% of _titration_- specific features could not be explained by sampling alone. The quantitative assessment evaluated pipeline performance by comparing observed to expected relative and differential abundance values. Expected relative abundance and differential abundance values were calculated using information from the unmixed samples and mixture design. Overall the observed relative abundance and differential abundance values were consistent with the expected values. We developed feature-level bias and variance metric to further characterize relative abundance and differential abundance quantitative performance. Relative abundance feature-level bias metric was significantly different across the three platforms with DADA2 having the lowest bias, followed by Mothur, and QIIME. The relative abundance feature-level variance metric and both the differential abundance feature-level bias and variance metrics did not differ significantly across the three pipelines. The dataset and methods developed for this study will serve as a valuable community resource for assessing 16S rRNA marker-gene survey bioinformatic methods.

output:
      bookdown::word_document2:
      # bookdown::pdf_document2:
            # template: springer_template.tex
            toc: FALSE
            # keep_tex: TRUE
            # citation_package: 'none'
bibliography: abundance_assessment.bib
---

# Introduction {#intro}
```{r intro, child="docs/Introduction.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
```

# Methods {#methods}
```{r methods, child="docs/Methods.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
```

# Results
## Dataset characteristics
```{r seqChar, child="results/seq_data_characteristics.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
```

## Titration Series Validation
To validate the two-sample titration dataset for use in abundance assessment we evaluated two assumptions about the titrations:
1. The samples were mixed volumetrically in a $log_2$ dilution series according to the mixture design.
2. The unmixed PRE and POST samples have the same proportion of prokaryotic DNA.
To validate the sample volumetric mixing exogenous DNA was spiked into the unmixed samples before mixing and quantified using qPCR .
To evaluate if the PRE and POST samples had the same proportion of prokaryotic DNA total prokaryotic DNA in the titrations samples was quantified using a qPCR assay targeting the 16S rRNA gene.


### Spike-in qPCR results
```{r ercc, child="results/ercc_validation.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
```

### Bacterial DNA Concentration
```{r bacCon, child="results/bac_con_validation.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
```

### Theta Estimates
```{r thetaEst, child="results/theta_estimate_results.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
```

## Measurement Assessment
Next, we assessed the qualitative and quantitative nature of 16S rRNA measurement process using our two-sample titration dataset.
For the qualitative assessment, we analyzed the relative abundance of features only observed in the unmixed samples or titrations which are not expected given the titration experimental design.
The quantitative assessment evaluated relative and differential abundance estimates.

### Qualitative Assessment

<!--
Proportion of titration and endpoint specific features that could not be explained by sampling alone.
-->

```{r qualAnalysis, child="results/qualitative_assessment_results.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
```

### Quantitative Assessment
<!-- __Quant Assessment Intro__ -->
<!--
Count Error Rate Results
-->
```{r quantAnalysis, child="results/relative_abundance_assessment_results.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
```

<!--
logFC Error Rate Results
-->
```{r logFCanalysis, child="results/logFC_assessment_results.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
```

# Discussion
```{r discussion, child="docs/Discussion.Rmd", warning=FALSE, message=FALSE, echo = FALSE}
```


# References
