---
title: "Supplemental"
author: "Nate Olson"
date: '`r Sys.Date()`'
output:
  bookdown::pdf_document2: 
    toc: FALSE
bibliography: [mgtst.bib, packages.bib]
---
```{r supplementalSetup, include=FALSE}
library(ggtree)
library(tidyverse)
library(forcats)
library(stringr)
library(ggridges)
library(ggpubr)
```

## theta estimate feature table 
```{r thetaEst}
## Raw counts 
raw_counts <- readRDS("~/Desktop/raw_counts.RDS")

## Negative Binomial Summaries
nb_counts <- readRDS("~/Desktop/nb_counts_titrations.RDS") 

## Pre-Post Expected Counts
pre_post <- nb_counts %>% 
      filter(t_fctr %in% c(0,20)) %>% 
      mutate(t_fctr = if_else(t_fctr == 0, "post", "pre")) %>% 
      spread(t_fctr, nb_prop)

total_count_df <- raw_counts %>% 
      group_by(pipe, biosample_id, id) %>% 
      summarise(total_count = sum(count))

pre_post_exp_count <- pre_post %>% 
      left_join(total_count_df) %>% 
      mutate(post_count = post * total_count,
             pre_count = pre * total_count)

####### Feature Filters
## PA - excluding features observed in less than 14 mixed sample PCR replicates  
pa_summary_anno_df <- readRDS("~/Desktop/pa_summary_anno_df.RDS")

pa_good <- pa_summary_anno_df %>% 
      ## Features present in all 4 PCR replicates for unmixed pre and/or post
      filter(full_pre + full_post != 0) %>% 
      ## Observed in more than half of the titration PCRs
      filter(pa_mixed >= 14)


## logFC > 1 between pre and post with a p-value < 0.05
pre_post_logFC_df <- readRDS("~/Desktop/logFC_edgeR_df.rds") %>% 
      ## Go back and change when logFC generated
     dplyr::rename(feature_id = OTUname) %>% 
     dplyr::select(pipe, biosample_id, T1, T2, feature_id, 
             logFC, logCPM, PValue, FDR) %>% 
      filter(T1 == 0, T2 == 20)

logFC_good <- pre_post_logFC_df %>% 
      filter(abs(logFC) > 1, FDR < 0.05) 


good_features <- pa_good %>% inner_join(logFC_good)

count_good <- raw_counts %>% 
      right_join(good_features) %>% ## feature filtering 
      left_join(pre_post_exp_count)

theta_est_df <- count_good %>% 
      filter(t_fctr %in% c(1:5,10,15)) %>% 
     dplyr::select(pipe, feature_id, id, pcr_rep, biosample_id, 
             t_fctr, count, post_count, pre_count, total_count) 
```

```{r thetaFeaturesMA, echo = FALSE, fig.cap = "MA plot (x-axis mean abundance, y-axis log2 fold change between unmixed pre- and post-exposure samples. Orange points indicate features used to estimate theta. Dashed lines indicate the +/- 1 log2 fold change filter used when to pick features."}
# pre_post_logFC_df %>% ggplot() +
#       geom_point(aes(x = logCPM, y = logFC), color = "grey60", alpha = 0.25) +
#       geom_hline(aes(yintercept = -1), linetype = 2) +
#       geom_hline(aes(yintercept = 1), linetype = 2) +
#       geom_point(data = good_features, 
#                  aes(x = logCPM, y = logFC),
#                  color = "darkorange",alpha = 0.5) +
#       facet_grid(pipe~biosample_id) +
#       # scale_x_log10() + 
#       theme_bw() +
#       labs(y = "log Fold-Change", x = "log Mean Abundance")
```

```{r thetaFeatures, warning=FALSE, message=FALSE, echo = FALSE}
 theta_est_df %>% ungroup() %>% 
     dplyr::select(pipe, biosample_id, feature_id) %>% 
      unique() %>% 
      group_by(pipe, biosample_id) %>% 
      summarise(count = n()) %>% 
      spread(biosample_id, count) %>% 
      knitr::kable(booktab = TRUE, caption = "Number of features used to estimate theta by biological replicate and pipeline.")
```

## relative abundance assessment 

```{r}
nb_counts <- readRDS("~/Desktop/nb_counts_titrations.RDS")
pa_summary_anno_df <- readRDS("~/Desktop/pa_summary_anno_df.RDS")
theta_est <- readRDS("~/Desktop/bootstrap_theta_estimates.rds")

### TODO - move to separate Rmd and generate data_frame
pre_post_prop <- nb_counts %>% 
    ungroup() %>% 
    filter(t_fctr %in% c(0,20)) %>% 
    mutate(end_point = if_else(t_fctr == 0 , "post", "pre")) %>% 
    dplyr::select(-t_fctr) %>% 
    ## setting values to 0 when one or more of the PCR replicates are 0 for titration end-points
    spread(end_point,nb_prop, fill = 0)

prop_inferred <- theta_est %>% 
    filter(pipe == "unclustered") %>% 
    ungroup() %>%
    mutate(t_fctr = factor(t_fctr, levels = c(0:5, 10, 15, 20))) %>% 
    dplyr::select(biosample_id, theta_hat_mean, t_fctr) %>% 
    right_join(nb_counts) %>% right_join(pre_post_prop) %>% 
    filter(t_fctr %in% c(1:5,10,15)) %>% 
    ## Using inferred theta estimates to calculate expected values
    mutate(inferred_prop = post * theta_hat_mean + pre * (1 - theta_hat_mean))

## Excluding mix and unmix specific features
## Only including features observed in all or none of the four pre- post- PCR replicates
## Features with relative abundance estimates and expected values less than 1e-5, these are features that we would not expect to consistently observe in a PCR replicate for the given sequencing depth, ~100k 
## Excluding titrations where the inferred theta values are less than 1
pa_filter <- pa_summary_anno_df %>% 
    filter(pa_specific == "unspecific") %>% 
    dplyr::select(biosample_id, pipe, feature_id, full_pre, T00, T20, pa_mixed) %>% 
    filter(T00 %in% c(0,4), T20 %in% c(0,4))

prop_inferred <- prop_inferred %>% 
    right_join(pa_filter) %>% 
    # filter(nb_prop > 1e-5, 
    #        inferred_prop > 1e-5,
    #        theta_hat_mean > 0)
    ## Filtering absed on 1/median library size
    filter(nb_prop > 1/73571,
           inferred_prop > 1/73571,
           theta_hat_mean > 0)


#### Error Rate Calculations
rel_abu_error <- prop_inferred %>% 
    mutate(t_fctr = factor(t_fctr, levels = c(1:5, 10, 15))) %>% 
    mutate(inferred_error = abs(nb_prop - inferred_prop),
           inferred_error_rate = inferred_error/inferred_prop)  
#### Error Rate Calculations
rel_abu_error <- prop_inferred %>% 
    mutate(t_fctr = factor(t_fctr, levels = c(1:5, 10, 15))) %>% 
    mutate(inferred_error = abs(nb_prop - inferred_prop),
           inferred_error_rate = inferred_error/inferred_prop) 

rel_abu_error_summary <-  rel_abu_error %>% 
    group_by(pipe, biosample_id, feature_id) %>% 
    summarise(median_rel_abu = median(nb_prop),
              median_error = median(inferred_error_rate),
              iqr_error = IQR(inferred_error_rate),
              rcov_error = iqr_error/median_error, 
              mean_error = mean(inferred_error_rate),
              var_error = var(inferred_error_rate),
              cov_error = var_error/mean_error) 

### Error rate boxplot and outlier annotation
error_boxplot <- rel_abu_error %>% group_by(pipe, biosample_id, feature_id) %>% 
    summarise(median_error = median(inferred_error_rate)) %>%
    ggplot() + 
    geom_boxplot(aes(x = pipe, y = median_error, color = pipe), outlier.shape = NA) + 
    facet_wrap(~biosample_id, ncol = 1) + 
    theme_bw() + theme(axis.text.x = element_text(angle = 90)) +
    labs(x = "Individual", y = "Median", color = "Pipeline")

## Annotating features as outliers based on boxplot
error_plot_dat <- ggplot_build(error_boxplot)$data[[1]] %>% 
    mutate(pipe = fct_recode(factor(group), 
                             dada2 = "1", 
                             mothur = "2", 
                             qiime = "3",
                             unclustered = "4"),
           biosample_id = fct_recode(PANEL, 
                                     E01JH0004 = "1", 
                                     E01JH0011 = "2", 
                                     E01JH0016 = "3", 
                                     E01JH0017 = "4", 
                                     E01JH0038 = "5"))
outlier_error_dat <- error_plot_dat %>% 
    dplyr::select(ymin, ymax, pipe, biosample_id)

rel_error_outlier_cat <- rel_abu_error_summary %>% 
    left_join(outlier_error_dat) %>% 
    mutate(outlier_cat = if_else(median_error < ymin | median_error > ymax, 
                                 "outlier","inlier")) 

## Robust COV Analysis
rcov_boxplot <- rel_abu_error_summary %>%
    ggplot() + 
    geom_boxplot(aes(x = pipe, y = rcov_error, color = pipe), outlier.shape = NA) + 
    facet_wrap(~biosample_id, ncol = 1) + 
    theme_bw() + theme(axis.text.x = element_text(angle = 90)) +
    labs(x = "Individual", y = "RCOV", color = "Pipeline") 


## Annotating features as outliers based on boxplot
rcov_plot_dat <- ggplot_build(rcov_boxplot)$data[[1]] %>% 
    mutate(pipe = fct_recode(factor(group), 
                             dada2 = "1", 
                             mothur = "2", 
                             qiime = "3",
                             unclustered = "4"),
           biosample_id = fct_recode(PANEL, 
                                     E01JH0004 = "1", 
                                     E01JH0011 = "2", 
                                     E01JH0016 = "3", 
                                     E01JH0017 = "4", 
                                     E01JH0038 = "5"))
outlier_rcov_dat <- rcov_plot_dat %>% 
    dplyr::select(ymin, ymax, pipe, biosample_id)

rcov_outlier_cat <- rel_abu_error_summary %>% 
    left_join(outlier_rcov_dat) %>% 
    mutate(outlier_cat = if_else(rcov_error < ymin | rcov_error > ymax, 
                                 "outlier","inlier"))

## Feature-level error summary data frame
rel_error_summary <- rel_error_outlier_cat %>% 
   dplyr::rename(error_cat = outlier_cat) %>% 
    dplyr::select(-ymin, -ymax) %>% 
    left_join(rcov_outlier_cat) %>% 
   dplyr::rename(rcov_cat = outlier_cat)
```

* feature table
```{r}
prop_inferred %>% 
   dplyr::select(pipe, biosample_id, feature_id) %>% distinct() %>% 
    group_by(pipe, biosample_id) %>% 
    summarise(count = n()) %>% 
    spread(biosample_id, count) %>% 
    knitr::kable(booktabs = TRUE, caption = "Number of features by pipeline and individual used in the relative abundance error rate analysis.")
```


```{r}
prop_inferred %>%
      ggplot() +
      geom_point(aes(x = inferred_prop, y = nb_prop), alpha = 0.15) +
      geom_abline(aes(intercept = 0, slope = 1), color = "darkorange") +
      geom_smooth(aes(x = inferred_prop, y = nb_prop), method = "lm") +
      facet_grid(pipe~biosample_id) +
      scale_y_log10() + scale_x_log10() +
      theme_bw() +
      labs(x = "Expected",
           y = "Observed")
```
    

* relative abundance feature characterization  
```{r errorMeanVarTidy, echo = FALSE, message = FALSE, warning = FALSE}
rel_abu_count_df <- readRDS("~/Desktop/raw_counts.RDS") %>%       
      group_by(pipe, id) %>% 
      mutate(total_count = sum(count)) %>% 
      right_join(rel_error_summary)

unmix_counts <- rel_abu_count_df %>% 
      filter(t_fctr %in% c(0, 20)) %>% 
      mutate(rel_abu = count/total_count) %>% 
      group_by(pipe, biosample_id, feature_id, t_fctr) %>% 
      summarise(mean_abu = mean(rel_abu),
                var_abu = var(rel_abu),
                range_abu = max(rel_abu) - min(rel_abu)) %>%
      mutate(var_mean = var_abu/mean_abu) %>% 
      left_join(rel_error_summary)
```

```{r treeFuns, echo = FALSE, message = FALSE, warning = FALSE}
get_fig_tree <- function(tree, features_to_compare, tax){
      # trimming tree
      tips_to_drop <- tree$tip.label[!(tree$tip.label %in% features_to_compare)] 
      trimmed_tree <- drop.tip(tree,tip = tips_to_drop) 
      
      ## Getting Taxonomic assignment data
      tree_tax <- tax %>% 
            as.data.frame() %>% 
            dplyr::rename(id = OTUname) %>% 
            filter(id %in% trimmed_tree$tip.label)  
      
      # Adding taxonomic groupings to tree 
      trimmed_tree <- groupOTU(trimmed_tree, 
                               split(tree_tax$id, tree_tax$Rank2), 
                               group_name = "Phylum")
      
      trimmed_tree

}

get_feats_to_compare <- function(error_df, pipe){
      metrics <- rel_error_summary %>% filter(pipe == pipe)

      ## Most of the features were used in this analysis for 1 of the five individuals. 
      ## To compare feature performance across individuals were only looked at features used for at least three of the individuals
     metrics %>% 
            group_by(feature_id) %>% 
            summarise(count = n()) %>% 
            filter(count > 2) %>% 
            .$feature_id
}

get_error_mat <- function(error_df, features_to_compare, tip_labels){
      ## Feature-level error data
      error <- error_df %>% 
            filter(feature_id %in% features_to_compare) %>% 
            ungroup() %>% 
           dplyr::select(biosample_id, feature_id, median_error) %>% 
            mutate(median_error = log2(abs(median_error) + 1)) %>% 
            spread(biosample_id, median_error)
      
      
      ## Reordering and converting to a matrix
      error_mat <- error %>% 
            right_join(data.frame(feature_id = tip_labels)) %>% 
            as.data.frame() %>% 
            column_to_rownames(var = "feature_id") %>% 
            as.matrix()
      colnames(error_mat) <- str_replace(colnames(error_mat), "E01JH00","ERR\n")
      
      ## Feature-level RCOV data
      rcov <- error_df %>% 
            filter(feature_id %in% features_to_compare) %>% 
            ungroup() %>% 
           dplyr::select(biosample_id, feature_id, rcov_error) %>% 
            mutate(rcov_error = log10(abs(rcov_error) + 1)) %>% 
            spread(biosample_id, rcov_error)
      
      ## Reordering cat data frame and converting to a matrix
      rcov_mat <- rcov %>% 
            right_join(data.frame(feature_id = tip_labels)) %>% 
            as.data.frame() %>% 
            column_to_rownames(var = "feature_id") %>% 
            as.matrix()
      colnames(rcov_mat) <- str_replace(colnames(rcov_mat), "E01JH00","RCOV\n")
      
      ## Generating a single matrix
      filler_mat <- matrix(NA, ncol = 1, nrow = nrow(rcov_mat))
      colnames(filler_mat) <- ""
      
      ## Return Matrix Data
      cbind(error_mat, filler_mat, rcov_mat)
}

get_fig_tree_dat <- function(tree, tax, error_df, pipe){
      ## Generating trimmed tree and matrix for plot
      features_to_compare <- get_feats_to_compare(error_df = error_df, 
                                            pipe = pipe) 
      
      trimmed_tree <- get_fig_tree(tree = tree, 
                                features_to_compare = features_to_compare,
                                tax = tax)
      
      mat <- get_error_mat(error_df = error_df, 
                                features_to_compare = features_to_compare,
                                tip_labels = trimmed_tree$tip.label)
      
      list(tree = trimmed_tree, dat = mat)
}
```

```{r relAbuErrorTrees, echo = FALSE, message = FALSE, warning = FALSE}
### DADA2 ----------------------------------------------------------------------
## Tree data
dada_tree <- readRDS("~/Projects/mgtst_pipelines/dada2/dada_tree_GTR.rds") %>% .$tree


## Getting Taxonomic assignment data
dada_tax <- readRDS("../mgtst_pipelines/dada2/dada_mrexp.rds") %>% 
      .@featureData %>% .@data

dada_tree_dat <- get_fig_tree_dat(tree = dada_tree, 
                                 tax = dada_tax, 
                                 error_df = rel_error_summary, 
                                 pipe = "dada2")

dada_fig_tree <- ggtree(dada_tree_dat$tree, aes(color = Phylum), ladderize = TRUE) %>%
      gheatmap(dada_tree_dat$dat, width = 0.5, font.size = 2) +
      scale_fill_gradient(low = "#56B1F7", high = "#132B43", na.value = "white") +
      theme(legend.position = "bottom")

### QIIME ----------------------------------------------------------------------
## Tree data
qiime_tree <- read.tree("~/Projects/mgtst_pipelines/qiime/otus_uc_fast/rep_set.tre")


## Getting Taxonomic assignment data
qiime_tax <- readRDS("../mgtst_pipelines/qiime/qiime_mrexp.rds") %>% 
      .@featureData %>% .@data %>%
      mutate(Rank2 = str_replace(Rank2, "p__",""))
      
      

qiime_tree_dat <- get_fig_tree_dat(tree = qiime_tree, 
                                 tax = qiime_tax, 
                                 error_df = rel_error_summary, 
                                 pipe = "qiime")

qiime_fig_tree <- ggtree(qiime_tree_dat$tree, aes(color = Phylum), ladderize = TRUE) %>%
      gheatmap(qiime_tree_dat$dat, width = 0.5, font.size = 2) +
      scale_fill_gradient(low = "#56B1F7", high = "#132B43", na.value = "white") +
      theme(legend.position = "bottom")

### MOTHUR ---------------------------------------------------------------------
## Tree
seq_id_df <- read_lines("~/Projects/mgtst_pipelines/mothur/mgtst.trim.contigs.good.unique.good.filter.unique.precluster.pick.opti_mcc.unique_list.0.03.rep.fasta") %>% grep(pattern = ">", value = TRUE) %>%
      str_replace("\\|.*","") %>%
      str_replace(">","") %>%
      enframe(name = "X", value = "seq_id") %>%
     dplyr::select(-X) %>%
      separate(seq_id, into = c("seq_id", "feature_id"),sep = "\t")


mothur_tree <- read.tree("~/Projects/mgtst_pipelines/mothur/mgtst.trim.contigs.good.unique.good.filter.unique.precluster.pick.opti_mcc.unique_list.0.03.rep.tre")

seq_tree_ids <- data_frame(seq_id = mothur_tree$tip.label) %>%
      left_join(seq_id_df)

mothur_tree$tip.label <- seq_tree_ids$feature_id


## Getting Taxonomic assignment data
mothur_tax <- readRDS("../mgtst_pipelines/mothur/mothur_mrexp.rds") %>% 
      .@featureData %>% .@data
      
      

mothur_tree_dat <- get_fig_tree_dat(tree = mothur_tree, 
                                 tax = mothur_tax, 
                                 error_df = rel_error_summary, 
                                 pipe = "qiime")

mothur_fig_tree <- ggtree(mothur_tree_dat$tree, aes(color = Phylum), ladderize = TRUE) %>%
      gheatmap(mothur_tree_dat$dat, width = 0.5, font.size = 2) +
      scale_fill_gradient(low = "#56B1F7", high = "#132B43", na.value = "white") +
      theme(legend.position = "bottom")
``` 

```{r errorOut, warning = FALSE, message = FALSE, echo = FALSE, fig.cap = "Feature-level median relative abundance and variance-mean relationship for the unmixed sample relative abundance not correlated with feature-level error rate and RCOV. Relationship between median relative abundance to the feature-level  (A) median error rate and (C) error RCOV. Relationship between unmixed sample relative abundance variance/mean to the feature-level (B) median error rate and (D) error RCOV."}

### Median relative abundance
error_rel_abu <- rel_error_summary %>% 
      ggplot(aes(x = median_rel_abu, y = median_error)) + 
      geom_point(aes(fill = error_cat), shape = 21) + 
      stat_cor(method = "spearman", label.y = 22) + 
      theme_bw() + 
      scale_x_log10() +       
      labs(x = "Relative Abundance", y = "Error Rate", fill = "Boxplot Outlier")

## excluding features with rcov_error == 0 - artifacts and not real....
rcov_rel_abu <- rel_error_summary %>% 
      filter(rcov_error != 0) %>% 
      ggplot(aes(x = median_rel_abu, y = rcov_error)) + 
      geom_point(aes(fill = rcov_cat), shape = 21) + 
      stat_cor(method = "spearman", label.y = 4.5) + 
      theme_bw() + 
      scale_x_log10() + scale_y_log10() +    
      labs(x = "Relative Abundance", y = "Error RCOV", fill = "Boxplot Outlier")


### Variance/Mean relationship
unmix_filt <- unmix_counts %>% 
      filter(!is.na(rcov_cat), !is.na(var_mean), var_abu != 0, mean_abu != 0, rcov_error != 0)

error_var_mean <- ggplot(unmix_filt, aes(x = var_mean, y = median_error)) + 
      geom_point(aes(fill = error_cat), shape = 21) +
      stat_cor(method = "spearman", label.y = 22) + 
      theme_bw() + scale_x_log10() +       
      labs(x = "Variance/Mean", y = "Error Rate", fill = "Boxplot Outlier")

## excluding features with rcov_error == 0 - artifacts and not real....
rcov_var_mean <- ggplot(unmix_filt, aes(x = var_mean, y = rcov_error)) +
      geom_point(aes(fill = rcov_cat), shape = 21) +
      stat_cor(method = "spearman", label.y = 4.5) + 
      theme_bw() + scale_x_log10() + scale_y_log10() +    
      labs(x = "Variance/Mean", y = "Error RCOV", fill = "Boxplot Outlier")

ggarrange(error_rel_abu, 
          error_var_mean, 
          rcov_rel_abu, rcov_var_mean, 
          ncol = 2, nrow = 2, align = "hv",labels = "AUTO",
      common.legend = TRUE, legend = "bottom")
```

```{r relAbuTreeFig, fig.cap = "Phylogenetic analysis of feature-level relative abundance error metrics, as heatmaps across individuals. Subplots for individual pipelines (A) DADA2, (B) QIIME, (C) Mothur. Feature phylum assignment is indicated by branch color.", fig.height = 10, message = FALSE, warning = FALSE, echo = FALSE}
ggarrange(dada_fig_tree, qiime_fig_tree, mothur_fig_tree, 
          labels = "AUTO", ncol = 1, nrow = 3)
```


## log fold-change assessment

```{r}
logFC_pre <- readRDS("~/Desktop/logFC_pre.RDS")
```


```{r preCountTbl, echo = FALSE}
logFC_pre %>% filter(T1 == 0, T2 == 20) %>%
      mutate(Type = if_else(T00 == 0, "specific", "dominant")) %>% 
      group_by(pipe, biosample_id, Type) %>% 
      summarise(count = n()) %>% 
      spread(pipe, count,fill = 0) %>% 
      dplyr::rename(Individual = biosample_id) %>% 
      knitr::kable(caption = "Number of pre-specific and pre-dominant features by individual and pipeline", booktabs = TRUE)
```

```{r logFCestVexp, echo = FALSE, warning = FALSE, message = FALSE}
logFC_pre %>%
      filter(T1 != 0, T2 != 20) %>%
      ggplot() +
      geom_point(aes(x = exp_logFC, y = logFC), alpha = 0.15) +
      geom_abline(aes(intercept = 0, slope = 1), color = "darkorange") +
      facet_grid(pipe~biosample_id, scales = "free_y") + theme_bw()  +
      labs(x = "Expected", y = "Estimate")
```

## log fold-change normalization comparison

```{r normLogFCmunge, include = FALSE}
### Loading Data ---------------------------------------------------------------
### RDS file generated in 2017-10-19_logFC-error-metrics.RDS 
logFC_pre <- readRDS("~/Desktop/norm_logFC_pre.RDS")

logFC_pre$norm_method <- logFC_pre$norm 
logFC_pre$norm <- NULL

logFC_feature_summary <- readRDS("~/Desktop/norm_logFC_feature_summary.RDS") %>% 
      mutate(slope_error = 1 - slope)

logFC_feature_summary$norm_method <- logFC_pre$norm 
logFC_feature_summary$norm <- NULL

## Calculating error rate
logFC_error <- logFC_pre %>% 
      filter(T1 != 0, T2 != 20) %>% 
      mutate(abs_error = abs(logFC - exp_logFC),
             error = logFC - exp_logFC)
```


```{r normCountTbl, echo = FALSE}
logFC_pre %>% 
      filter(T1 == 0, T2 == 20) %>%
      mutate(Type = if_else(T00 == 0, "specific", "dominant")) %>% 
      group_by(norm_method, biosample_id, Type) %>% 
      summarise(count = n()) %>% 
      spread(norm_method, count, fill = 0) %>% 
      dplyr::rename(Individual = biosample_id) %>% 
      knitr::kable(caption = "Number of pre-specific and pre-dominant features by individual and normalization method for Mothur", booktabs = TRUE)
```


```{r logfcNormComp, fig.cap = "Impact of normalization methods on the agreement between log fold-change estimates and expected values for pre-specific and pre-dominant features. (A) Linear model relating the log fold-change estimates with the expected values by individual and normalization method. (B) Distribution of log fold-change (B) absolute error and (C) error by normalization method and individual.", echo=FALSE, message = FALSE, fig.height = 8}
logFCestVexp <- logFC_error %>% 
      filter(T1 != 0, T2 != 20) %>% 
      ggplot() + 
      geom_smooth(aes(x = exp_logFC, y = logFC, color = norm_method), method = "lm") +
      geom_abline(aes(intercept = 0, slope = 1), color = "grey20", linetype = 2) + 
      facet_wrap(~biosample_id, ncol = 1, scales = "free_y") + theme_bw()  +
      labs(x = "Expected", y = "Estimate", color = "Normalization") 

logFC_abs_error_dist <- logFC_error %>% 
      group_by(biosample_id, norm_method) %>% 
      mutate(med_error = median(abs_error)) %>% 
      ggplot() + 
      geom_density_ridges(aes(x = abs_error, y = norm_method, color = norm_method), 
                          alpha = 0.5, stat = "binline", bins = 30, draw_baseline = FALSE)  + 
      geom_text(aes( x = -1, y = norm_method, label = round(med_error,2))) + 
      facet_wrap(~biosample_id, ncol = 1) + theme_bw() + 
      labs(x = "|Error|", y = "Normalization", color = "Normalization") +
      theme(legend.position = "none")

logFC_error_dist <- logFC_error %>% 
      group_by(biosample_id, norm_method) %>% 
      mutate(med_error = median(error)) %>% 
      ggplot() + 
      geom_density_ridges(aes(x = error, y = norm_method, color = norm_method), 
                          alpha = 0.5, stat = "binline", bins = 30, draw_baseline = FALSE)  + 
      geom_text(aes( x = -10, y = norm_method, label = round(med_error,2))) + 
      facet_wrap(~biosample_id, ncol = 1) + theme_bw() + 
      labs(x = "Error", y = "Normalization", color = "Normalization") +
      theme(legend.position = "none")

ggarrange(logFCestVexp, 
          logFC_error_dist, 
          logFC_abs_error_dist, 
          labels = "AUTO",
          align = "v", 
          ncol = 3, nrow = 1,
          common.legend = TRUE,
          legend = "bottom")
```

## log fold-change estimator comparison 

## log fold-change feature characterization  
* metric comparison plot  
* phylosignal plot


\pagebreak  


# Session information 

## Git repo commit information
```{r include = FALSE}
library(tidyverse)
library(git2r)
repo <- repository(path = ".")
last_commit <- commits(repo)[[1]]
```

The current git commit of this file is `r last_commit@sha`, which is on the `r branches(repo)[[1]]@name`  branch and was made by `r last_commit@committer@name` on `r when(last_commit)`. The current commit  message is `r last_commit@summary`. The repository is online at https://github.com/nate-d-olson/mgtst-pub  


## Platform Information  
```{r session_info, warning=FALSE, message=FALSE, echo = FALSE}
s_info <- devtools::session_info()
print(s_info$platform)
```


## Package Versions  
```{r warning=FALSE, message=FALSE, echo = FALSE}
s_info$packages %>% filter(`*` == "*") %>% 
      dplyr::select(-`*`) %>%
      knitr::kable(booktabs = TRUE, longtable=TRUE)
```
