---
title: "EO - Metric Categorization"
author: "Nate Olson"
date: '`r Sys.Date()`'
always_allow_html: yes
output:
  pdf_document: default
  html_document: default
---
```{r setup, warning=FALSE, message=FALSE, echo = FALSE}
library(git2r)
library(tidyverse)
```


```{r}
feature_exp_df <- readRDS("../data/nb_expected_eo_metric_feature_df.rds") 

genus_exp_df <- readRDS("../data/nb_expected_eo_metric_genus_df.rds") 

exp_df <- bind_rows(feature = feature_exp_df, genus = genus_exp_df, .id = "fLvl")
```

## Objective
Qualitative analysis of feature abundance for features with EO metrics of -1 and 1. 

## Approach 
For EO metrics of -1, for a PCR replicate-features determine if random sampling can account for zero observed count values. 
This will be achieved assuming a negative binomial sampling distribution, where the expected proportion is the probability of success and the total abundance is the sample size. 
For features where sampling cannot account for the observed 0 counts look at relative frequency by PCR replicate, biosample, pipeline, and taxonomy. 

## EO -1
The expected count values tend to be higher for DADA2 compared to the other pipelines for all biological replicates. 
At the feature level (OTU and SV) the median expected count is less than 1 for all five biological replicates when data is processed using the QIIME pipeline and two of the five biological replicates for the Mothur pipeline. 
The median expected count value is consistently higher for DADA2 compared to the other pipelines at both the feature and genus aggregated count level. 
Overall count values are higher for DADA2 and the feature inference method does not allow for the generation of singletons. 

```{r fig.cap = "Distribution of expected count values for features with a PCR replicate EO metric of -1. The horizontal black line indicates a expected count value of 1."}
exp_df %>% filter(eo_metric == -1) %>% 
      ggplot() +
      geom_boxplot(aes(x = pipe, y = exp_count + 1, fill = biosample_id), alpha = 0.25) + 
      geom_hline(aes(yintercept = 1 + 1)) + 
      scale_y_log10() + theme_bw() + 
      labs(x = "Pipeline", y = "Expected Count Values", fill = "Biological Replicate") +
      facet_wrap(~fLvl)
```

```{r}
exp_df %>% filter(eo_metric == -1) %>% 
      group_by(fLvl, biosample_id, pipe) %>% 
      summarise(med_exp_count = median(exp_count)) %>% 
      spread(biosample_id, med_exp_count) %>% 
      knitr::kable()
```

Attempted to test for differences in the expected count values but unable to fit a model to the data. See appendix for additional information. 

### Probability of 0 observed counts 

Assuming a negative binomial distribution for the counts able to calculate the probability of no observed counts based on the expected count proportion and sample total abundance.  

```{r}
eo_neg1 <- exp_df %>%
      filter(eo_metric == -1) %>% 
      mutate(p0 = pbinom(q = 1, size = total_abu, prob = exp_prop))
```

Example calculation

Look into using the binomial test to determine the p-value for the expected proportion to be less than that calculated based on the observed pre and post counts. 
```{r}
b_test <- binom.test(x = 1, n = 10, p = 0.001, alternative="greater")
```

Need to make sure this is testing the correct thing. 
Want to test whether the true proportion is less than the estimated expected proportion
```{r}
eo_binom_test <- eo_neg1 %>% 
      group_by(fLvl, pipe, biosample_id, id, feature_id, t_fctr) %>% 
      nest() %>% 
      mutate(binom_test = map(data, ~with(., binom.test(x = 0, n = total_abu, exp_prop, alternative = "less"))),
             binom_tidy = map(binom_test, broom::tidy))
```

```{r}
eo_binom_tidy <- eo_binom_test %>% 
      select(fLvl, pipe, biosample_id, id, feature_id, t_fctr, binom_tidy) %>% 
      unnest()
```

```{r}
eo_binom_tidy %>% filter(fLvl == "feature") %>% ggplot() + 
      geom_density(aes(x = p.value, color = pipe)) + theme_bw() 
```


__Distribution of probabilities__  
For most of the features the probability of 0 observed counts is > 0.01 indicating that sampling depth alone accounts for most of the features with EO values of 1. 
Choosing 0.01 as a cutoff to explore features where sampling alone does not account for no observed counts. 
This value is arbitry and should be revised. 
Also while not a hypothesis test, may want to correct for multiple comparisons.

```{r}
eo_neg1 %>% 
      mutate(high_prob = if_else(p0 > 0.01, "> 0.01", "<= 0.01")) %>% 
      ggplot() +
      geom_bar(aes(x = biosample_id, fill = high_prob), position = "fill") +
      labs(fill = "", x = "Biological Replicate", y = "Proportion") + 
      facet_grid(pipe~fLvl) +
      theme_bw() 
```

## Frequency of Features with a -1 EO by sample
Comparing the total number of features per sample (PCR replicate) with a -1 EO metric. 
Will want to figure out how to normalize for differences in the number of features per pipeline, biological replicate, and sample to appropriately compare.

```{r}
eo_neg1 %>% filter(fLvl == "feature") %>% 
      group_by(fLvl, pipe, biosample_id, id) %>% 
      summarise(count = n()) %>% 
      ggplot() + geom_density(aes(x = count, fill = biosample_id), alpha = 0.25) +
      facet_grid(pipe~fLvl, scales = "free") + theme_bw()
```

However some samples consistently have high number of features with -1 EO metrics across pipelines and feature level. 

```{r}
top_fs <- eo_neg1 %>% 
      group_by(fLvl, pipe, biosample_id, id) %>% 
      summarise(count = n()) %>% 
      group_by(fLvl, pipe) %>% select(-biosample_id) %>% 
      top_n(2, count) %>% ungroup() %>% select(id) %>% unique()

eo_neg1 %>% 
      group_by(fLvl, pipe, biosample_id, id) %>% 
      summarise(count = n()) %>% 
      group_by(fLvl, pipe) %>% select(-biosample_id) %>% 
      right_join(top_fs) %>% spread(id, count)
```

Samples 1-E4 and 1-F9 have low counts which 
```{r}
eo_neg1 %>% 
      group_by(fLvl, pipe, biosample_id, id) %>% 
      summarise(count = n()) %>% 
      group_by(fLvl, pipe) %>% select(-biosample_id) %>% 
      top_n(3, count) %>% select(id) %>% left_join(exp_df) %>% 
      ungroup() %>% 
      select(id, total_abu) %>% unique() %>% knitr::kable()
```



## -1 EO Metric and Sample Total Abundance
Aside from samples with total abundance less than 50,000 reads, the number of features with a EO metric of -1 is consistent within biological replicates and pipelines. 

```{r}
exp_df %>% filter(eo_metric == -1) %>% 
      group_by(fLvl, pipe, biosample_id, id, total_abu) %>% 
      summarise(count = n()) %>%
      ggplot() +
      geom_point(aes(x = total_abu, y = count)) + 
      geom_smooth(aes(x = total_abu, y = count, color = biosample_id)) +
      facet_grid(fLvl~pipe, scales = "free_y") + theme_bw()
```


### Number PCR Replicates with -1 E0 metrics for a sample 

__NOTES__  
- Features with count of 1, where only 1 of 4 PCR replicates have EO metric values of -1  
- Features with only one PCR replicate with a EO metric of -1, indicates that the one PCR is potentially an outlier. 
- Features with -1 EO metric for all four PCR replicates indicates a potential error or bias in sequencing or bioinformatic pipeline. 


```{r}
exp_df %>% filter(eo_metric == -1) %>% 
      group_by(fLvl, pipe, biosample_id, feature_id, t_fctr) %>% 
      summarise(count = n()) %>% 
      ggplot() +
      geom_bar(aes(x = pipe, fill = as.factor(count)), position = "dodge") + 
      facet_grid(fLvl~biosample_id, scale = "free_y") + theme_bw()
```

## E0 1
__NOTES__  
- The observed count values for PCR replicate level EO metrics is 1 or less for most biological replicates and pipelines for Mothur and QIIME at both the feature level and genus aggregated level.  
- For the DADA2 pipeline the observed count values are higher, with a median value around 10 for all biological replicates at both the feature and genus level.  
- The consistently higher observed count levels indicate a potential error with the feature inference method. Though the higher than observed counts could be the result of a potential artifact of the sequencing/ PCR procedure where a sequence fails to amplify in the presence of other sequences in the sample due to competing for primer binding sites. The fact that the higer observd count values are only observed for one of the three pipelines indicates that it is more likely, though not guaranteed  the result of an issue with the feature inference procedure and not sequencing.  

```{r}
exp_df %>% filter(eo_metric == 1) %>% 
      ggplot() +
      geom_boxplot(aes(x = biosample_id, y = count, fill = pipe), alpha = 0.25) + 
      geom_hline(aes(yintercept = 1 + 1)) + 
      scale_y_log10() + theme_bw() + 
      labs(x = "Biological Replicate", y = "Observed Count Values", fill = "Pipeline") +
      facet_wrap(~fLvl) + theme(axis.text.x = element_text(angle = 90))
```

__Next Steps__
- Look into differences in sampling depth between unmixed and mixed samples.  
- Is there a higher frequency of EO metric values of 1 for some PCR replicates. 
      - Correlate number of features with a metric value of 1 with `total_abu` 
```{r}
exp_df %>% filter(eo_metric == 1, pipe == "dada2") %>% 
      ggplot() + geom_bar(aes(x = id)) + 
      facet_wrap(~fLvl, scale = "free_y", ncol = 1) +
      theme(axis.text.x = element_text(angle = 90))
```

What is the one outlier sample? 
```{r}
exp_df %>% filter(eo_metric == 1, pipe == "mothur") %>% 
      ggplot() + geom_bar(aes(x = id)) + 
      facet_wrap(~fLvl, scale = "free_y", ncol = 1) +
      theme(axis.text.x = element_text(angle = 90))
```

```{r}
exp_df %>% filter(eo_metric == 1, pipe == "qiime") %>% 
      ggplot() + geom_bar(aes(x = id)) + 
      facet_wrap(~fLvl, scale = "free_y", ncol = 1) +
      theme(axis.text.x = element_text(angle = 90))
```

- The number of features with EO metrics of 1 for a PCR replicate is correlated with the total abundance, excluding the one outlier for Mothur (not included in plot).  
- This correlation indicates that for at least some of the features with EO metric values of 1 can be explained by sampling. Will want to compare sampling depth for the titration to the unmixed samples.  

```{r}
exp_df %>% filter(eo_metric == 1) %>% 
      group_by(fLvl, pipe, biosample_id, id, total_abu) %>% 
      summarise(count = n()) %>% filter(count < 1000) %>% 
      ggplot() +
      geom_point(aes(x = total_abu, y = count)) + 
      geom_smooth(aes(x = total_abu, y = count, color = biosample_id)) +
      facet_grid(fLvl~pipe, scales = "free_y")
```

```{r}
exp_df %>% filter(eo_metric == 1) %>% 
      group_by(fLvl, pipe, biosample_id, id, total_abu) %>% 
      summarise(count = n()) %>% 
      filter(count > 1000)
```


## Appendix
### EO -1 Expected Count Value Regression 
Best fit for regression model, other models, family poisson and lm with and without log transformation (2 and 10) tended to have NA or Inf AIC values
```{r}
fit <- glm(log2(exp_count + 1) ~ log2(total_abu) + pipe*biosample_id, 
           data = filter(exp_df, eo_metric == -1, fLvl == "feature"), 
           family=gaussian(link="log"))
summary(fit)
```

```{r}
library(ggfortify)
autoplot(fit)
```

# Session information 

## Git repo commit information
```{r}
library(git2r)
repo <- repository(path = "../")
last_commit <- commits(repo)[[1]]
```

The current git commit of this file is `r last_commit@sha`, which is on the `r branches(repo)[[1]]@name` branch and was made by `r last_commit@committer@name` on `r when(last_commit)`. The current commit message is `r last_commit@summary`. The repository is online at https://github.com/nate-d-olson/mgtst-pub  


## Platform Information
```{r}
s_info <- devtools::session_info()
print(s_info$platform)
```


## Package Versions
```{r}
s_info$packages %>% filter(`*` == "*") %>% select(-`*`) %>% 
      knitr::kable()
```