---
title: "Identification of Informative and Uninformative Features"
author: "Nate Olson"
date: '`r Sys.Date()`'
always_allow_html: yes
output:
  pdf_document: default
  html_document: default
---

```{r setup, warning=FALSE, message=FALSE, echo = FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
library(ProjectTemplate)
cwd <- getwd()
setwd("../")
load.project()
setwd(cwd)
```

```{r echo = FALSE}
pipeline_dir <- "../../mgtst_pipelines"
mrexp <- get_mrexp(pipeline_dir)
```

## Objective 
Identification of informative and uninformative features. 
Informative features defined as features with observed counts in all pre-treatment samples and all titration samples and all or none of the post-treatment replicates. 


## Feature Categories
* Null - features not present in more than one PCR replicate for any sample of a biological replicate, and pipeline.  

__Informative__  

* Full - features present all PCR replicates for all samples (unmixed and titrations) for a biological replicate.  
* Pre - present in all PCR replicates for unmixed pre-treatment samples, not observed in any PCR replicates of the unmixed post treatment samples, and present in all titration PCR replicates.  
* Post - present in all PCR replicates for the unmixed post-treatment samples, not observed in any PCR replicates of the unmixed pre-treatment samples, and present all titration PCR replicates.   


__Characterization of low abundance features__
* Mix - features present in mixed samples but not observed in any of the unmixed sample PCR replicates. 



```{r}
## Extracting a tidy dataframe with count values from MRexpiment objects
get_count_df <- function(mrobj, agg_genus = FALSE){
      if(agg_genus){
            mrobj <- aggregateByTaxonomy(mrobj, lvl = "Rank6", 
                                         norm = FALSE, log = FALSE, sl = 1)
      }
      
      mrobj <- cumNorm(mrobj, p = 0.75)
      mrobj %>%
            # not sure whether or not to normalize counts prior to analysis
            MRcounts(norm = TRUE, log = FALSE, sl = 1000) %>%  
            as.data.frame() %>% 
            rownames_to_column(var = "feature_id") %>% 
            gather("id","count", -feature_id)
}


get_rep_info <- function(count_df){
      count_replicate_df <- count_df %>% 
            mutate(detect = if_else(count > 1, 1, 0)) %>%
            group_by(pipe, biosample_id, titration, t_fctr, feature_id) %>% 
            summarise(total_detect = sum(detect),
                      n_replicates = n(),
                      avg_non0_count = sum(count)/total_detect) %>%
            mutate(detect_prop = total_detect/n_replicates) %>% 
            select(-total_detect) 
      
      count_replicate_df %>% ungroup() %>% 
      mutate(t_fctr = paste0("T",t_fctr)) %>% 
      select(pipe, biosample_id, feature_id, t_fctr, detect_prop) 
}


assign_cat <- function(rep_info){
      prop_summary <- rep_info %>% 
            group_by(pipe, biosample_id, feature_id) %>% 
            summarise(prop_max = max(detect_prop), 
                      prop_min = min(detect_prop), 
                      prop_sum = sum(detect_prop)) 
      
      unmix_prop <- rep_info %>% 
            filter(t_fctr %in% c("T0","T20")) %>% 
            spread(t_fctr, detect_prop)
      
      left_join(prop_summary, unmix_prop) %>% 
            mutate(cat_null = if_else(prop_max == 0, 1, 0),
                   cat_full = if_else(prop_min == 1, 1, 0),
                   cat_near_full = if_else(prop_min == 0.75, 1, 0),
                   cat_mix  = if_else(prop_max == 1 & T0 == 0 & T20 == 0, 1, 0),
                   ## Post prop 5 - expected at least three replicates for titrations 4, 5, 10, and 15 
                   ## Pre prop 3 - expected at least three replicates for titrations 1, 2, 3, and 4 
                   ## titration 4, is ~94% post
                   ## titration 4, is ~94% post
                   cat_pre  = if_else(T20 == 1 & T0 == 0 & prop_sum == 8, 1, 0),
                   cat_post = if_else(T0 == 1 & T20 == 0 & prop_sum == 8, 1, 0),
                   cat_none = if_else(cat_null + cat_full + cat_near_full + cat_mix + cat_pre + cat_post == 0, 1, 0))
}
```

### Feature Level Category Assignments
```{r}
count_df <- mrexp %>% map_df(get_count_df, .id = "pipe") %>% 
      left_join(pData(mrexp$dada2)) #%>% 
      # filter(biosample_id != "NTC") 

#count_df
rep_info <- get_rep_info(count_df) 

#rep_info
feature_info <- assign_cat(rep_info)

feature_cat <- feature_info %>% 
      select(-prop_max, -prop_min, -prop_sum, -T0, -T20) %>% 
      gather(cat, value, -pipe, -biosample_id, -feature_id) %>% 
      filter(value == 1) %>% select(-value) 
    
feature_cat %>% saveRDS("../data/feature_categories_df.rds")
```




#### Category Sanity Check
```{r}
cat_check <- feature_cat %>% 
      group_by(pipe, biosample_id, feature_id) %>% 
      summarise(n_cat = n())
cat_check %>% filter(n_cat != 1)
```


```{r}
# cat_check <- feature_info %>%
#       select(-prop_max, -prop_min, -prop_sum, -T0, -T20) %>% 
#       gather(cat, value, -pipe, -biosample_id, -feature_id) %>%
#       group_by(pipe, biosample_id, feature_id) %>%
#       mutate(n_cat = sum(value)) %>%
#       filter(n_cat != 1, value != 0)
# cat_check %>% arrange(feature_id)
```


```{r}
      prop_summary <- rep_info %>% 
            group_by(pipe, biosample_id, feature_id) %>% 
            summarise(prop_max = max(detect_prop), 
                      prop_min = min(detect_prop), 
                      prop_sum = sum(detect_prop))
```

## Informative Features 
```{r}
feature_cat %>% filter(cat != "cat_null", cat != "cat_none", cat != "cat_mix") %>% 
      ggplot() + geom_bar(aes(x = cat)) + 
      facet_grid(pipe ~ biosample_id, scales = "free_y") + 
      theme_bw() + theme(axis.text.x = element_text(angle = 90))
```



### Recovering Semi-Informative Features
```{r}
cat_none_df <- feature_cat %>% left_join(rep_info) %>% filter(cat == "cat_none")
total_prop_cat_none <- cat_none_df %>% group_by(pipe, biosample_id, feature_id) %>% 
    summarise(total_prop = sum(detect_prop))
```

Most of the uncategorized features were observed in less than 4 PCR replicates
```{r}
total_prop_cat_none %>% mutate(total_prop = floor(total_prop)) %>% 
    group_by(pipe, total_prop) %>% summarise(count = n()) %>% 
    spread(pipe, count) %>% knitr::kable()
```

Assuming 4 PCR replicates for all samples. 
Using proportions if samples are excluded from analysis, for example samples with few reads compared to the rest of the samples.  

```{r}
cat_none_df %>% filter(!(detect_prop %in% c(0,0.25,0.5,0.75,1)))
```

What is the detect_proportions for samples with total prop < 2. 
```{r}
cat_none_low <- cat_none_df %>% group_by(pipe, biosample_id, feature_id) %>% 
    mutate(total_prop = sum(detect_prop)) %>% filter(total_prop > 0, total_prop < 2)
```

Most of the low total detect proportion features were only observed in one PCR replicate. 
```{r}
cat_none_low %>% 
    ggplot() + geom_bar(aes(x = total_prop, fill = biosample_id), 
                        position = "dodge") + 
    facet_wrap(~pipe, scales = "free_y") + theme_bw()
```

```{r}
# cat_none_low %>% filter(total_prop > 0.5) %>% spread(t_fctr, detect_prop) %>% arrange(total_prop)
```

Most of the low detect features are only present in 1 sample
```{r}
cat_none_low %>% mutate(pa = if_else(detect_prop == 0, 0, 1)) %>% summarise(total_presence = sum(pa)) %>% 
    ggplot() + geom_bar(aes(x = total_presence, fill = biosample_id), 
                        position = "dodge") + 
    facet_wrap(~pipe, scales = "free_y") + theme_bw()
```


__Open Question__ Are low detect features due to random sampling or bioinformatic/ experimental artifact. 
* Approach - average counts by number of PCR replicates and number of samples with observed counts.

```{r}
count_replicate_df <- count_df %>% 
            mutate(detect = if_else(count > 1, 1, 0)) %>%
            group_by(pipe, biosample_id, titration, t_fctr, feature_id) %>% 
            mutate(total_detect = sum(detect),
                      n_replicates = n(),
                      avg_non0_count = sum(count)/total_detect) %>%
            mutate(detect_prop = total_detect/n_replicates) %>% 
            select(-total_detect)
count_rep_pa <- count_replicate_df %>% 
    group_by(pipe, biosample_id, feature_id, t_fctr) %>% 
    summarise(pa = if_else(sum(detect) != 0, 1, 0)) %>% 
    group_by(pipe, biosample_id, feature_id) %>% 
    mutate(sample_detect = sum(pa))
count_replicate_pa_df <- count_rep_pa %>% filter(sample_detect != 0) %>% left_join(count_replicate_df)
```

X-axis: Number of samples out of the 2 unmixed and 7 titrations with at least one of the four PCR replicates with observed counts
Y-axis: Mean counts for PCR replicates with non-zero count values 
Color: Of the sample (either unmixed or titration) proportion of PCR replicates with observed counts
```{r}
count_replicate_pa_df %>% ggplot() + 
    geom_boxplot(aes(y = avg_non0_count + 1, x = factor(sample_detect), color = factor(detect_prop))) + 
    scale_y_log10() + facet_wrap(~pipe, ncol = 1) + theme_bw()
```

## Extracting additional informative features
```{r}
pre_post_titrate <- cat_none_df %>% spread(t_fctr, detect_prop) %>% 
    mutate(pre_titration = if_else(T20 != 0 & T15 != 0 & 
                                       T20 >= T15 & T15 >= T10 & T10 >= T5 & 
                                       T5 >= T4 & T4 >= T3 & T3 >= T2  & 
                                       T2 >= T1 & T1 >= T0, 1, 0),
           post_titration = if_else(T0 != 0 & T1 != 0 & 
                                        T20 <= T15 & T15 <= T10 & T10 <= T5 & 
                                       T5 <= T4 & T4 <= T3 & T3 <= T2  & 
                                       T2 <= T1 & T1 <= T0, 1, 0)) %>% 
    filter(pre_titration == 1 | post_titration == 1) 

```

Feature assigned to both groups if present in the same number of PCR replicates for all samples
```{r}
pre_post_titrate %>% mutate(t_sum = pre_titration + post_titration) %>% filter(t_sum > 1)
```


```{r}
pre_post_titrate %>% mutate(t_sum = pre_titration + post_titration) %>% filter(t_sum == 1) %>% 
    mutate(titrate = if_else(pre_titration == 1, "pre","post")) %>% 
    ggplot() + geom_bar(aes(x = titrate, fill = biosample_id)) + facet_wrap(~pipe) + theme_bw()
```



```{r}
total_prop_titrate <- pre_post_titrate %>% mutate(t_sum = pre_titration + post_titration) %>% filter(t_sum == 1) %>% 
    mutate(titrate = if_else(pre_titration == 1, "pre","post")) %>% 
    select(pipe, biosample_id, feature_id, titrate) %>% left_join(cat_none_df) %>% 
    group_by(pipe, biosample_id, feature_id,titrate) %>% mutate(total_prop = sum(detect_prop))
```

```{r}
total_prop_titrate %>% ggplot() + geom_bar(aes(x = total_prop)) + facet_grid(titrate~pipe)
```

##### Summary Figures
```{r}
feature_cat %>% filter(cat != "cat_null", cat != "cat_none") %>% 
      ggplot() + geom_bar(aes(x = cat)) + 
      facet_grid(pipe ~ biosample_id, scales = "free_y") + 
      theme_bw() + theme(axis.text.x = element_text(angle = 90))
```





### Genus Level Category Assignments
```{r}
count_df <- mrexp %>% map_df(get_count_df,agg_genus = TRUE, .id = "pipe") %>% 
      left_join(pData(mrexp$dada2)) %>% 
      filter(biosample_id != "NTC") 

rep_info <- get_rep_info(count_df) 
rep_info %>% saveRDS("../data/genus_rep_info_df.rds") 

feature_info <- assign_cat(rep_info) 
feature_info %>% saveRDS("../data/genus_info_df.rds") 

feature_cat <- feature_info %>% 
      select(pipe, biosample_id, feature_id, 
             cat_null, cat_full, cat_mix, cat_pre, cat_post, cat_none) %>% 
      gather(cat, value, -pipe, -biosample_id, -feature_id) %>% 
      filter(value == 1) %>% select(-value)

feature_cat %>% saveRDS("../data/genus_categories_df.rds") 
```

#### Category Sanity Check
```{r}
cat_check <- feature_cat %>% 
      group_by(pipe, biosample_id, feature_id) %>% 
      summarise(n_cat = n())
cat_check %>% filter(n_cat != 1)
```


```{r}
# cat_check <- feature_categories %>% 
#       select(pipe, biosample_id, feature_id, 
#              cat_null, cat_full, cat_mix, cat_pre, cat_post, cat_none) %>% 
#       gather(cat, value, -pipe, -biosample_id, -feature_id) %>% 
#       group_by(pipe, biosample_id, feature_id) %>% 
#       mutate(n_cat = sum(value)) %>% 
#       filter(n_cat != 1, value != 0)
# cat_check %>% arrange(feature_id)
```


##### Summary Figures 
Larger proportion of full category features and fewer mix specific features when aggregating to the genus level compared to unaggregated features. 
```{r}
feature_cat %>% filter(cat != "cat_null") %>% 
      ggplot() + geom_bar(aes(x = cat)) + 
      facet_grid(pipe ~ biosample_id, scales = "free_y") + 
      theme_bw() + theme(axis.text.x = element_text(angle = 90))
```

While there are a large number of unclassified features, few are potentially informative. 
Ones that stand out are features detected in 4 T0 (pre-treatment features) with prop sum value close to 8. 

```{r}
feature_info %>% filter(cat_none == 1, prop_sum > 2, T0 > 0.5 | T20 > 0.5) %>%
      ggplot() + geom_histogram(aes( x= prop_sum)) + facet_grid(T0 ~ T20) + theme_bw()
```

## Alternative Characterization 
```{r}
detect_idx <- rep_info %>% mutate(detect = detect_prop * 4, 
                    t_fctr = fct_relevel(t_fctr, paste0("T",c(0:5,10,15,20))),
                    idx_buffer = 10^(as.numeric(t_fctr)-1),
                    idx = detect * idx_buffer) %>% 
      group_by(pipe,biosample_id,feature_id) %>% mutate(idx = sum(idx)) %>% 
      filter(idx != 0)  

detect_idx %>% 
      group_by(idx, pipe, biosample_id, feature_id) %>% 
      mutate(total_detect = sum(detect)) %>% 
      filter(total_detect > 9) %>%
      group_by(idx, pipe) %>% 
      summarise(count = n()) %>% 
       arrange(desc(count)) 
```

## Appendix
Table with the total proportion of PCR replicates by biosample and pipeline. 

```{r}
total_prop_cat_none %>% mutate(total_prop = floor(total_prop)) %>% 
    group_by(pipe, biosample_id, total_prop) %>% summarise(count = n()) %>% 
    spread(biosample_id, count) %>% knitr::kable()
```

# Session information
```{r}
s_info <- devtools::session_info()
print(s_info$platform)
s_info$packages %>% filter(`*` == "*") %>% select(-`*`) %>% 
      knitr::kable()
```