---
title: "Theta Estimate Features log fold change"
author: "Nate Olson"
date: '`r Sys.Date()`'
output:
  bookdown::pdf_document2: 
    toc: FALSE
---

__Look into edgeR for logFC calc__  
__Calc logFC and take ratio - directly calculate logFC__
__Change order for filtering features__ 


```{r theta_logFC_setup, warning=FALSE, message=FALSE, echo = FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
library(metagenomeSeq)
library(tidyverse)
library(stringr)
library(modelr)
count_nb <- readRDS("../data/nb_expected_eo_metric_feature_df.rds")
mothur_mrexp <- readRDS("../../mgtst_pipelines/mothur/mothur_mrexp.rds")
```

Issue logFC results are not close to expectation with difference if 1 between subsequent titrtations. 
* Two approaches to validate methods 
1. Generate simulated count data to validate code 
2. Look at results for a subset of samples - one biological replicate and two titrations.

### logFC for Mothur  Biosample ID E01JH0011
E01JH0011 had the theta estimates that were in the best agreement with the expected theta values. 

Looking at logFC between the unmixed post-exposure sample and the first titration. 
```{r}
## Only looking at 
mothur_11 <- mothur_mrexp[,which(pData(mothur_mrexp)$biosample_id == "E01JH0011")]
mothur_11 <- mothur_11[,which(pData(mothur_11)$titration %in% c(1,2))]
mothur_11 <- mothur_11[rowSums(mothur_11) > 0,]
mothur_11 <- filterData(mothur_11, present = 8, depth = 1)
mothur_11 <- cumNorm(mothur_11, p = 0.75)
pd <- pData(mothur_11)
mod <- model.matrix(~1 + t_fctr, data = pd)
mothur_res1 <- fitFeatureModel(mothur_11, mod)
mothur_logFC <- MRcoefs(mothur_res1, number = 195) %>% as.data.frame() %>% 
      rownames_to_column(var = "feature_id")
mothur_prepost <- count_nb %>% ungroup() %>% 
      filter(pipe == "mothur", biosample_id == "E01JH0011") %>%
      select(feature_id, pre, post) %>% unique()
mothur_logFC <- mothur_logFC %>% 
      left_join(mothur_prepost) %>% #filter(!is.na(post)) %>% 
      mutate(feat_spec = case_when(.$pre != 0 & .$post != 0 ~ "non",
                                   .$pre == 0 & .$post != 0 ~ "post",
                                   .$post ==0 & .$pre != 0 ~ "pre"))
```

```{r}
mothur_logFC %>% ggplot() + 
      geom_histogram(aes(x = logFC, fill = feat_spec)) + 
      geom_vline(aes(xintercept = 1)) +
      geom_vline(aes(xintercept = -1)) +
      theme_bw()
```

Looking for a feature abundance effect
```{r}
mothur_logFC %>% 
      ggplot() + 
      geom_density2d(aes(x = post*100000, y = logFC, color = feat_spec)) +
      geom_point(aes(x = post*100000, y = logFC, color = feat_spec)) + 
      scale_x_log10() + facet_wrap(~feat_spec)
``` 

```{r}
mothur_11 <- mothur_mrexp[,which(pData(mothur_mrexp)$biosample_id == "E01JH0011")]
mothur_11
```


```{r}
mothur_11 <- mothur_11[,which(pData(mothur_11)$titration %in% c(0,1,2,20))]
mothur_11
```


```{r}
mothur_11 <- mothur_11[rowSums(mothur_11) > 0,]
mothur_11
```

```{r}
mothur_11_counts <- MRcounts(mothur_11)
```

Resulting count matrix sparsity
```{r}
calc_sparsity <- function(mat){
      nentry <- length(mat)
      nzero <- sum(mat == 0)
      ## calculate sparsity 
      nzero/nentry
}
calc_sparsity(mothur_11_counts)
```


```{r}
norm_fact <- colSums(mothur_11_counts)#edgeR::calcNormFactors(mothur_11_counts)
norm_counts <- sweep(mothur_11_counts,2,as.vector(unlist(norm_fact)),"/")
```

```{r}
norm_fact
```
```{r}
pData(mothur_11)
```

```{r}
head(norm_counts)
```

```{r}
count_df <- norm_counts %>% as.data.frame() %>% 
      rownames_to_column(var = "feature_id") %>% 
      gather("id","norm_count", -feature_id) %>% 
      left_join(pData(mothur_11)) 
```



```{r}
wide_count_df <-count_df %>% 
      group_by(feature_id, biosample_id, t_fctr) %>% 
      summarise(mean_count = median(norm_count)) %>% 
      mutate(t_fctr = paste0("T",t_fctr)) %>% 
      spread(t_fctr, mean_count) %>% 
      ungroup() %>% 
      mutate(feat_spec = case_when(.$T0 != 0 & .$T20 != 0 ~ "non",
                                   .$T20 == 0 & .$T0 != 0 ~ "post",
                                   .$T0 ==0 & .$T20 != 0 ~ "pre"))
```

```{r}
count_df %>% filter(norm_count != 0) %>% .$norm_count %>% summary()
```


```{r}
logFC_df <- wide_count_df %>% 
      filter(T2 != 0) %>% 
      mutate(logFC = log2(T1/T2))
```

```{r}
logFC_df %>% ggplot() + geom_histogram(aes(x = logFC)) + 
      geom_vline(aes(xintercept = 1)) +
      geom_vline(aes(xintercept = -1)) +
      facet_wrap(~feat_spec)
```

Next steps - calculate logFC using edgeR

```{r}
raw_count_df <- mothur_11_counts %>% as.data.frame() %>% 
      rownames_to_column(var = "feature_id") %>% 
      gather("id","count", -feature_id) %>% 
      left_join(pData(mothur_11)) 
raw_logFC <- raw_count_df %>% filter(id %in% c("1-B2", "1-C8")) %>% 
      select(-pcr_16S_plate, -pos, -pcr_half, -pcr_rep, -titration, -t_fctr) %>% 
      spread(id, count) %>% 
      mutate(logFC = log2(`1-B2`/`1-C8`))
```


```{r}
feat_anno <- wide_count_df %>% select(feature_id, biosample_id, feat_spec)
```

```{r}
raw_logFC %>% left_join(feat_anno) %>% 
      ggplot() + geom_histogram(aes(x=logFC)) + facet_wrap(~feat_spec)
```



```{r}
total_abu <- mothur_mrexp %>% colSums()
```

```{r}
mothur_pdata <- pData(mothur_mrexp) %>% add_column(total_abu)
```

```{r}
mothur_pdata %>% nrow()
```


```{r}
mothur_pdata %>% filter(total_abu < 55000, total_abu > 45000) %>% group_by(biosample_id, t_fctr) %>% summarise(count = n()) %>% 
      spread(t_fctr, count)
```


Using non-normalized counts to look at logFC between sequential titrations
```{r}
mothur_sub_pdat <- mothur_pdata %>% filter(total_abu < 55000, total_abu > 45000) %>% 
      group_by(biosample_id, t_fctr) %>% filter(biosample_id == "E01JH0016", t_fctr %in% c(1,2))
```

```{r}
count_mat <- mothur_mrexp@assayData$counts
```

```{r}
count_df <- count_mat %>% as.data.frame() %>% rownames_to_column(var = "feature_id") %>% 
      gather("id","count", -feature_id)
```


```{r}
sub_count <- mothur_sub_pdat %>% left_join(count_df)
```

```{r}
sub_logFC <- sub_count %>% 
      group_by(biosample_id, feature_id, t_fctr) %>% 
      summarise(med_count = sum(count)) %>% 
      ungroup() %>%
      mutate(t_fctr = paste0("T",t_fctr)) %>% 
      spread(t_fctr, med_count) %>% 
      mutate(logFC = if_else(T2 != 0, log2(T1/T2), NaN))
```

```{r}
sub_logFC %>% ggplot() + geom_histogram(aes(x = logFC))
```

```{r}
unmix_count <- mothur_pdata %>% 
      filter(biosample_id == "E01JH0016", t_fctr %in% c(0,20)) %>% 
      left_join(count_df) %>% 
      mutate(t_fctr = paste0("T",t_fctr)) %>% 
      group_by(biosample_id, t_fctr, feature_id) %>% 
      summarise(total_count = sum(count)) %>% 
      spread(t_fctr, total_count) %>% 
      mutate(feat_spec = case_when(.$T0 != 0 & .$T20 != 0 ~ "non",
                             .$T20 == 0 & .$T0 != 0 ~ "post",
                             .$T0 ==0 & .$T20 != 0 ~ "pre"))
```

Total counts for unmixed pre and post 
```{r}
unmix_count %>% ggplot() + 
      geom_point(aes(x = T0 + 1, y = T20 + 1, color = feat_spec)) + 
      scale_y_log10() + scale_x_log10()
```

```{r}
sub_logFC %>% left_join(unmix_count) %>% 
      ggplot() + geom_point(aes(x = feat_spec, y = logFC))
```

```{r}
sub_logFC %>% left_join(unmix_count) %>% 
      filter(feat_spec == "post") %>% 
      ggplot() + geom_point(aes(x = T0, y = logFC))
```

```{r}
sub_logFC %>% left_join(unmix_count) %>% 
      filter(T0 > 4) %>% # arrange(T20, -T0)
      arrange(-T1)
```


```{r}
sub_logFC %>% left_join(unmix_count) %>% 
      gather("t_fctr","total_count", -biosample_id, -feature_id, -logFC, -feat_spec) %>% 
     ungroup() %>%
     ggplot() + geom_point(aes(x = factor(t_fctr), y = total_count, color = feat_spec)) +
     geom_line(aes(x = factor(t_fctr) %>% as.numeric(), y = total_count, group = feature_id)) +
     facet_wrap(~feat_spec) + scale_y_log10()
 ```


```{r}
sub_logFC %>% left_join(unmix_count) %>% 
      filter(T0 > 1000, T20 < 100) %>% 
      ggplot() + geom_point(aes(x = T0, y = logFC))
```

```{r}

```

### DADA2
```{r}
total_abu <- dada_mrexp %>% colSums()
```

```{r}
hist(total_abu)
```

```{r}
dada_pdata <- pData(dada_mrexp) %>% add_column(total_abu)
```

```{r}
dada_pdata %>% filter(biosample_id == "E01JH0011", t_fctr %in% c(2,3)) %>% arrange(total_abu)
```

```{r}
dada_pdata %>% filter(total_abu < 65000, total_abu > 55000) %>% group_by(biosample_id, t_fctr) %>% summarise(count = n()) %>% 
      spread(t_fctr, count)
```

```{r}
dada_pdata %>% filter(total_abu < 65000, total_abu > 55000)  %>% filter()
```


```{r}
count_mat <- dada_mrexp@assayData$counts
count_df <- count_mat %>% as.data.frame() %>% rownames_to_column(var = "feature_id") %>% 
      gather("id","count", -feature_id)

dada_count <- dada_pdata %>% 
      left_join(count_df) %>% 
      mutate(t_fctr = paste0("T",t_fctr)) 
```
```{r}
dada_sub <- dada_count %>% filter(total_abu < 600000, total_abu > 58000, 
                                  t_fctr %in% c("T0","T1","T2","T20"), 
                                  biosample_id == "E01JH0004") %>% 
      group_by(biosample_id, t_fctr, feature_id) %>% 
      summarise(med_count = median(count)) %>% 
      spread(t_fctr, med_count) %>% 
      filter(T0 != 0 | T20 != 0) %>% 
      mutate(logFC = if_else(T2 != 0, log2(T1/T2), NaN)) %>% 
      mutate(feat_spec = case_when(.$T0 != 0 & .$T20 != 0 ~ "non",
                       .$T20 == 0 & .$T0 != 0 ~ "post",
                       .$T0 ==0 & .$T20 != 0 ~ "pre"))
```

```{r}
dada_sub %>% ggplot(aes(x = feat_spec, y = logFC)) + geom_boxplot() + geom_point()
```

```{r}
dada_sub %>% ggplot() + geom_point(aes(x = T20 + 1, y = logFC)) + 
      facet_wrap(~feat_spec) + scale_x_log10()
```

Subset based on pair of theta estimates that are closest to expectation.
```{r}
dada_sub <- dada_count %>% filter(#total_abu < 67000, total_abu > 57000, 
                                  t_fctr %in% c("T0","T2","T3","T20"), 
                                  biosample_id == "E01JH0004") %>% 
      group_by(biosample_id, t_fctr, feature_id) %>% 
      summarise(med_count = median(count))# %>% 
      # spread(t_fctr, med_count) #%>% 
     # filter(T0 != 0 | T20 != 0) #%>% 
      # mutate(logFC = if_else(T2 != 0, log2(T2/T3), NaN)) %>% 
      # mutate(feat_spec = case_when(.$T0 != 0 & .$T20 != 0 ~ "non",
                       .$T20 == 0 & .$T0 != 0 ~ "post",
                       .$T0 ==0 & .$T20 != 0 ~ "pre"))
```

```{r}
dada_count %>% filter(total_abu < 67000, total_abu > 57000, 
                                  t_fctr %in% c("T0","T2","T3","T20"), 
                                  biosample_id == "E01JH0011") %>% 
      select(id, t_fctr, total_abu) %>% 
      unique() %>% arrange(t_fctr)
```

```{r}
dada_pdata %>% filter(biosample_id == "E01JH0011") %>% arrange(-total_abu) %>% ggplot() + geom_histogram(aes(x = total_abu))
```
```{r}
dada_pdata %>% filter(biosample_id == "E01JH0011") %>% arrange(-total_abu) %>% ggplot() + geom_histogram(aes(x = total_abu)) + facet_wrap(~t_fctr)
```


```{r}
dada_sub$t_fctr %>% unique() %>% ggplot(aes(x = feat_spec, y = logFC)) + geom_boxplot() + geom_point()
```


```{r}
dada_sub_count <- dada_count %>% filter(total_abu < 67000 & total_abu > 57000, 
                                  t_fctr %in% c("T2","T3"),
                                  biosample_id == "E01JH0011")
```


```{r}
dada_sub <- dada_count %>% filter(total_abu < 67000, total_abu > 57000, 
                                  t_fctr %in% c("T2","T3"), 
                                  biosample_id == "E01JH0011") %>% 
      group_by(biosample_id, t_fctr, feature_id) %>% 
      summarise(med_count = median(count)) %>% 
      spread(t_fctr, med_count) %>%
      mutate(logFC = if_else(T2 != 0, log2(T2/T3), NaN))
```

```{r}
dada_endpoint <- dada_count %>% filter(t_fctr %in% c("T0","T20"), 
                                  biosample_id == "E01JH0011") %>% 
      group_by(biosample_id, t_fctr, feature_id) %>% 
      summarise(med_count = median(count)) %>% 
      spread(t_fctr, med_count) %>%
      mutate(feat_spec = case_when(.$T0 != 0 & .$T20 != 0 ~ "non",
                                   .$T20 == 0 & .$T0 != 0 ~ "post",
                                   .$T0 ==0 & .$T20 != 0 ~ "pre"))
```

```{r}
dada_sub %>% left_join(dada_endpoint) %>% 
      ggplot() + geom_point(aes(x = feat_spec, y = logFC))
```


```{r}
dada_sub %>% left_join(dada_endpoint) %>% filter(!is.nan(logFC)) %>% arrange(feat_spec)
```

```{r}
dada_sub %>% left_join(dada_endpoint) %>% filter(!is.nan(logFC)) %>% arrange(feat_spec) %>% filter(T0 > 1000, T20 < 100)
```


Using edgeR to calculate logFC without normalization
```{r}
subset_ids <- dada_count %>% filter(total_abu < 670000, total_abu > 57000, 
                                    t_fctr %in% c("T2","T3"), 
                                    biosample_id == "E01JH0004") %>% 
      select(id, t_fctr) %>% unique()

sub_count_mat <- count_mat[,subset_ids$id]
library(edgeR)

y <- DGEList(sub_count_mat, group = subset_ids$t_fctr)
y
```

```{r}
y <- estimateDisp(y)
y <- estimateCommonDisp(y)
y <- estimateTagwiseDisp(y)
```

```{r}
et <- exactTest(y)
topTags(et)
```

```{r}
et$table %>% as.data.frame()
```

```{r}
end_point_counts <-
      dada_count %>% filter(t_fctr %in% c("T0","T20"), 
                       biosample_id == "E01JH0004") %>% 
      group_by(feature_id, t_fctr) %>% 
      summarise(med_count = median(count)) %>% 
      ungroup() %>% 
      spread(t_fctr, med_count) %>% 
      mutate(feat_spec = case_when(.$T0 != 0 & .$T20 != 0 ~ "non",
                       .$T20 == 0 & .$T0 != 0 ~ "post",
                       .$T0 ==0 & .$T20 != 0 ~ "pre"))
```

```{r}
et$table %>% as.data.frame() %>% rownames_to_column(var = "feature_id") %>% left_join(end_point_counts) %>% 
      ggplot() + geom_point(aes(x = feat_spec, y = logFC))
```


Simplying dataset has not lead to cleaner results. 
Potential sources of error
1. normalization - 
      1. test - using samples with similar total abundance
1. sample mixtures - titrations with theta estimates near expectation
1. feature inference
      1. Potential tests - looking at dereplicated sequence set, looking at taxonomic lineage e.g. genus...
2. PCR count values are too noisy to obtain accurate logFC estimates - look at count estimate variance



```{r}
dada_pdata %>% ggplot() + geom_point(aes(x = t_fctr, y = total_abu)) + facet_wrap(~biosample_id)
```

```{r}
dada_pdata %>% 
      filter(biosample_id != "NTC") %>% 
      mutate(total_abu_bins = cut_width(total_abu, width = 1000)) %>% 
      group_by(biosample_id, t_fctr, total_abu_bins) %>% 
      summarise(n_samples = n()) %>% 
      ggplot() + geom_raster(aes(x = t_fctr, y = total_abu_bins, fill = n_samples)) + facet_wrap(~biosample_id)
```


```{r}
dada_count %>% filter(feature_id == "SV1", biosample_id != "NTC") %>%
      mutate(t_fctr = factor(t_fctr, levels = paste0("T",c(0:5, 10, 15, 20)))) %>% 
      ggplot() + 
      geom_point(aes(x = t_fctr, y = count)) + 
      geom_smooth(aes(x = as.numeric(t_fctr), y = count)) +
      facet_wrap(~biosample_id, nrow = 1) + theme_bw()
```

