---
title: "logFC data munging NIST2"
author: "Nate Olson"
date: '`r Sys.Date()`'
always_allow_html: yes
output:
  pdf_document: default
  html_document: default
---

```{r message = FALSE}
library(ProjectTemplate)
cwd <- getwd()
setwd("../")
load.project()
setwd(cwd)
pipeline_dir <- "../../mgtst_runs/nist2" 

get_mrexp <- function(pipelines_dir){
      mrexp_files <- list(
            dada2 = file.path(pipeline_dir, "dada2/dada_mrexp.rds"),
            mothur =  file.path(pipeline_dir, "mothur/mothur_mrexp.rds"),
            qiime =  file.path(pipeline_dir, "qiime/qiime_mrexp.rds")
      )
      
      mrexp_files %>% map(readRDS)
}
mrexp <- get_mrexp(pipeline_dir)
```

## Objective
Workout code for generating data frame with estimated logFC and variance between titrations for the different pipelines and differential abundance methods. 
Currently using the default normalization for each differential abundance detection method.

Pipelines: 
1. DADA2  
2. Mothur   
3. QIIME  

Differential Abundance Methods:  
1. metagenomeSeq - fitFeatureModel  
2. DESeq2  
3. EdgeR  

## Approach 
Use a data_frame object and list columns to apply differential abundance methods to different datasets  

## Outline
1. Define parameter fields for titration comparisons 
3. Develop function for each differential abundance method 
4. Save data frames with log fold-change calculations for each differential abundance detection method. 

## Titration Comparisons

```{r}
pdat <- mrexp$dada2 %>% pData %>% 
      mutate(t_fctr = factor(t_fctr, level = c(0:5,10,15,20)))
t1_pdat <- pdat %>% select(biosample_id, id, t_fctr) %>% 
      rename(T1 = t_fctr, T1_id = id) 
t2_pdat <- pdat %>% select(biosample_id, id, t_fctr) %>% 
      rename(T2 = t_fctr, T2_id = id) 
titration_comp_df <- left_join(t1_pdat, t2_pdat) %>% 
      filter(as.numeric(T1) < as.numeric(T2)) %>% 
      group_by(biosample_id, T1, T2) %>% 
      summarise(sam_names = c(T1_id,T2_id) %>% unique() %>% list(),
                n_sams = c(T1_id,T2_id) %>% unique() %>% length())
```


Comparison for all titration combinations for each biological replicate. 
Total of `r nrow(titration_comp_df)` sets of logFC calculations.

Pairwise comparisons sanity check, expectation T1 always less than T2 with 8 samples for each pairwise comparison
```{r}
titration_comp_df %>% ggplot() + 
      geom_raster(aes(x = T1, y = T2, fill = n_sams)) + 
      facet_wrap(~biosample_id)
```

## Subset MRexperiments
For each set of logFC calculations subset the MRexperiment objects 
```{r}
make_titration_comp_subset_df <- function(mrexp_obj, titration_comp_df){

      ## Subsetting mrexp for titrations
      subset_mrexp <- function(sam_names){
            mrexp_obj %>% {.[,which(colnames(.) %in% sam_names)]}
      }
      
      ## Dataframe with list of subsetted mrexp
     titration_comp_df %>% 
            mutate(mrexp_sub = map(sam_names, subset_mrexp))
}

### DADA2
titration_comp_dada2_df <- mrexp$dada2 %>% 
  make_titration_comp_subset_df(titration_comp_df)

### Mothur
titration_comp_mothur_df <- mrexp$mothur %>%
      make_titration_comp_subset_df(titration_comp_df)

### QIIME
titration_comp_qiime_df <- mrexp$qiime %>%
      make_titration_comp_subset_df(titration_comp_df)
```


## Estimate logFC 
### MetagenomeSeq - fitFeatureModel  

Functions for calculating logFC using fitFeatureModel
```{r}
fit_model <- function(mrexp_sub, T1, T2, css_p = 0.75, present = 8, depth = 1){
      ## Normalize count data
      T1 <- as.numeric(as.character(T1))
      T2 <- as.numeric(as.character(T2))
      mrexp_sub <-  cumNorm(mrexp_sub, p = css_p)

      mrexp_sub2 <- filterData(mrexp_sub, present = present, depth = depth)

      pd <- pData(mrexp_sub2) 

      ## Fitting Model
      mod <- model.matrix(~1 + t_fctr, data = pd)
      fitFeatureModel(mrexp_sub2, mod)
}

fit_safely <- safely(fit_model)
mrcoefs_safely <- safely(MRcoefs)

```

#### DADA2
```{r message = FALSE, warning = FALSE}
logFC_MgSeq_dada2_df <- titration_comp_dada2_df %>% 
      mutate(fit = pmap(list(mrexp_sub, T1, T2), fit_safely)) 

logFC_MgSeq_dada2_coefs_df <- logFC_MgSeq_dada2_df %>% 
      mutate(fit_coefs = map(fit, ~mrcoefs_safely(.$result, number = Inf))) 

logFC_MgSeq_dada2_coefs_tidy <- logFC_MgSeq_dada2_coefs_df %>% 
      select(biosample_id, T1, T2, fit_coefs) %>% 
      mutate(fit_coefs = map(fit_coefs, ~.$result),
             fit_coefs_class = map_chr(fit_coefs, class)) %>% 
      filter(fit_coefs_class == "data.frame") %>%
      select(-fit_coefs_class) %>%
      mutate(fit_coefs = map(fit_coefs, rownames_to_column, var = "featureNames")) %>%
      unnest()
```


#### Mothur
```{r message = FALSE, warning = FALSE}
logFC_MgSeq_mothur_df <- titration_comp_mothur_df %>% 
      mutate(fit = pmap(list(mrexp_sub, T1, T2), fit_safely)) 

logFC_MgSeq_mothur_coefs_df <- logFC_MgSeq_mothur_df %>% 
      mutate(fit_coefs = map(fit, ~mrcoefs_safely(.$result, number = Inf))) 

logFC_MgSeq_mothur_coefs_tidy <- logFC_MgSeq_mothur_coefs_df %>% 
      select(biosample_id, T1, T2, fit_coefs) %>% 
      ## filtering titration comparisons where the model failed
      mutate(fit_coefs = map(fit_coefs, ~.$result),
             fit_coefs_class = map_chr(fit_coefs, class)) %>% 
      filter(fit_coefs_class == "data.frame") %>% select(-fit_coefs_class) %>% 
      mutate(fit_coefs = map(fit_coefs, rownames_to_column, var = "featureNames")) %>% 
      unnest()
```

#### QIIME
```{r message = FALSE, warning = FALSE}
logFC_MgSeq_qiime_df <- titration_comp_qiime_df %>% 
      mutate(fit = pmap(list(mrexp_sub, T1, T2), fit_safely)) 

logFC_MgSeq_qiime_coefs_df <- logFC_MgSeq_qiime_df %>% 
      mutate(fit_coefs = map(fit, ~mrcoefs_safely(.$result, number = Inf))) 

logFC_MgSeq_qiime_coefs_tidy <- logFC_MgSeq_qiime_coefs_df %>% 
      select(biosample_id, T1, T2, fit_coefs) %>% 
      ## excluding titration comparisons where the model failed
      mutate(fit_coefs = map(fit_coefs, ~.$result),
             fit_coefs_class = map_chr(fit_coefs, class)) %>% 
      filter(fit_coefs_class == "data.frame") %>% select(-fit_coefs_class) %>% 
      mutate(fit_coefs = map(fit_coefs, rownames_to_column, var = "featureNames")) %>% 
      unnest()
```

Cleanup fitfeatureModel
```{r}
rm(logFC_MgSeq_dada2_df)
rm(logFC_MgSeq_dada2_coefs_df)
rm(logFC_MgSeq_mothur_df)
rm(logFC_MgSeq_mothur_coefs_df)
rm(logFC_MgSeq_qiime_df)
rm(logFC_MgSeq_qiime_coefs_df)
```

### EdgeR
Function for running EdgeR
```{r}
## Function based on phyloseq_to_edgeR code 
## http://joey711.github.io/phyloseq-extensions/edgeR.html
mrexp_to_edgeR <- function(mrexp_obj, group, method = "RLE", ...){
      require(edgeR)
      ## Extracting count data - no scaling or transformation
      x <- mrexp_obj %>% MRcounts(sl = 1) %>% as.matrix()
      x <- x + 1 # add 1 to prevent log(0) issues

      ## Check `group` argument
      if(identical(all.equal(length(group), 1), TRUE) & ncol(mrexp_obj)> 1){
            ## Assumes grouop is a categorical sample variable name
            group <- pData(mrexp_obj) %>% .[,group]
      }

      ## Use taxonomy information at gene annotations 
      ## - Where OTUname is incorporated into the results
      taxonomy <- fData(mrexp_obj)
      if(!is.null(taxonomy)){
            taxonomy <- taxonomy %>% as.matrix() %>% data.frame()
      }


      ## Convert into a DGEList
      y <- DGEList(counts = x, group = group, genes = taxonomy,
                   remove.zeros = TRUE, ...)


      ## Calc normalization factors
      z <- edgeR::calcNormFactors(y, method = method)

      ## Check for division by zero inside `calcNormFactors`
      if( !all(is.finite(z$amples$norm.factors))){
            stop("Something wrong with edgeR::calcNormFactors on this data,
                 non-finite $norm.factors, consider changing `method` argument.")
      }

      ## Estimate dispersions
      z %>% estimateCommonDisp() %>% estimateTagwiseDisp()
}
```

#### DADA2
```{r message = FALSE}
require(edgeR)
logFC_edgeR_dada2_df <- titration_comp_dada2_df %>% 
      mutate(fit = map(mrexp_sub, mrexp_to_edgeR, group = "t_fctr"),
             fit = map(fit, exactTest)) 

logFC_edgeR_dada2_coefs_df <- logFC_edgeR_dada2_df %>% 
      mutate(fit_coefs = map(fit, topTags, n = Inf, adjust.method = "BH")) 

logFC_edgeR_dada2_coefs_tidy <- logFC_edgeR_dada2_coefs_df %>% 
      select(biosample_id, T1, T2, fit_coefs) %>% 
      mutate(fit_coefs = map(fit_coefs, ~.@.Data[[1]])) %>% 
      unnest()
```

#### Mothur
```{r message = FALSE}
logFC_edgeR_mothur_df <- titration_comp_mothur_df %>% 
 mutate(fit = map(mrexp_sub, mrexp_to_edgeR, group = "t_fctr"),
             fit = map(fit, exactTest)) 

logFC_edgeR_mothur_coefs_df <- logFC_edgeR_mothur_df %>% 
      mutate(fit_coefs = map(fit, topTags, n = Inf, adjust.method = "BH")) 

logFC_edgeR_mothur_coefs_tidy <- logFC_edgeR_mothur_coefs_df %>% 
      select(biosample_id, T1, T2, fit_coefs) %>% 
      mutate(fit_coefs = map(fit_coefs, ~.@.Data[[1]])) %>% 
      unnest()
```

#### QIIME
```{r message = FALSE}
logFC_edgeR_qiime_df <- titration_comp_qiime_df %>% 
      mutate(fit = map(mrexp_sub, mrexp_to_edgeR, group = "t_fctr"),
             fit = map(fit, exactTest)) 

logFC_edgeR_qiime_coefs_df <- logFC_edgeR_qiime_df %>% 
      mutate(fit_coefs = map(fit, topTags, n = Inf, adjust.method = "BH")) 

logFC_edgeR_qiime_coefs_tidy <- logFC_edgeR_qiime_coefs_df %>% 
      select(biosample_id, T1, T2, fit_coefs) %>% 
      mutate(fit_coefs = map(fit_coefs, ~.@.Data[[1]])) %>% 
      unnest()
```


## Save logFC tables 
Combine tables for the different pipelines into individual data frames for each differential abundance methods and save as RDS files.  
```{r}
list(dada2 = logFC_MgSeq_dada2_coefs_tidy,
     mothur = logFC_MgSeq_mothur_coefs_tidy,
     qiime = logFC_MgSeq_qiime_coefs_tidy) %>% 
      bind_rows(.id = "pipe") %>% saveRDS("~/Desktop/nist2_logFC_MgSeq_df.rds") 

list(dada2 = logFC_edgeR_dada2_coefs_tidy,
     mothur = logFC_edgeR_mothur_coefs_tidy,
     qiime = logFC_edgeR_qiime_coefs_tidy) %>%
      bind_rows(.id = "pipe") %>% saveRDS("~/Desktop/nist2_logFC_edgeR_df.rds")
```



# Session information 

## Git repo commit information
```{r}
library(git2r)
repo <- repository(path = "../")
last_commit <- commits(repo)[[1]]
```

The current git commit of this file is `r last_commit@sha`, which is on the `r branches(repo)[[1]]@name` branch and was made by `r last_commit@committer@name` on `r when(last_commit)`. The current commit message is `r last_commit@summary`. The repository is online at https://github.com/nate-d-olson/mgtst-pub  


## Platform Information
```{r}
s_info <- devtools::session_info()
print(s_info$platform)
```


## Package Versions
```{r}
s_info$packages %>% filter(`*` == "*") %>% select(-`*`) %>% 
      knitr::kable()
```
