---
title: "logFC data munging"
author: "Nate Olson"
date: '`r Sys.Date()`'
output: html_document
---

```{r message = FALSE}
library(ProjectTemplate)
cwd <- getwd()
setwd("../")
load.project()
setwd(cwd)
```

## Objective
Workout code for generating data frame with estimated logFC and variance between titrations for the different pipelines and differential abundance methods. 

Pipelines: 
1. DADA2  
2. Mothur   
3. QIIME  

Normalization Methods: 
1. TSS  
2. CSS  
3. UQ  
4. DESeq2 method 
5. EdgeR method

Differential Abundance Methods:  
1. metagenomeSeq - fitFeatureModel  
2. DESeq2  
3. EdgeR  
4. Permutation  

## Approach 
Use a data_frame object and list columns to apply differential abundance methods to different datasets  

## Outline
1. Load pipeline data
2. Define parameter fields for titration comparisons 
3. Develop function for each differential abundance method 

## Loading pipeline data
```{r}
mrexp <- list(dada2  = "../data/mrexp_dada2.RDS",
              mothur = "../data/mrexp_mothur.RDS",
              qiime  = "../data/mrexp_qiime_refclus_nochimera.RDS") %>% 
      map(readRDS)
```

Adding `sam_names` to `pData` for DADA2 and QIIME MRexperiments, and converting dilutions to factors.
```{r}
pdat <- pData(mrexp$dada2)
pdat$sam_names <- rownames(pdat)
pdat$dilution <- pdat$dilution %>% factor()
pData(mrexp$dada2) <- pdat 

pdat <- pData(mrexp$mothur)
pdat$dilution <- pdat$dilution %>% factor()
pData(mrexp$mothur) <- pdat 

pdat <- pData(mrexp$qiime) 
pdat$sam_names <- rownames(pdat)
pdat$dilution <- pdat$dilution %>% factor()
pData(mrexp$qiime) <- pdat
```


## Titration Comparisons

```{r}
## Variables for Comparisons 
titrations <- c(-1,0,1:5,10,15)
T1 <- rep(titrations, each = length(titrations)) %>% as.numeric()
T2 <- rep(titrations, length(titrations)) %>% as.numeric()

sampleID <- c("E01JH0004", "E01JH0017", "E01JH0038", "E01JH0011", "E01JH0016")

titration_comp_df <- sampleID %>% set_names(sampleID) %>% 
      ## excluding duplicate, inverse, logFC calculations
      map_df(~data_frame(T1,T2),.id = "sampleID") %>% filter(T1 < T2)
```


Comparison for all titration combinations for each biological replicate. 
Total of `r nrow(titration_comp_df)` sets of logFC calculations.

```{r}
DT::datatable(titration_comp_df)
```



## Subset MRexperiments
For each set of logFC calculations subset the MRexperiment objects 

```{r}
make_titration_comp_subset_df <- function(mrexp_obj, titration_comp_df, present = 1){
      ## mrexp_obj and mrexp_pdat are defined globally and not passed as arguments
      get_sams <- function(sample_id, T1, T2){
        mrexp_pdat %>% filter(dilution %in% c(T1, T2), sampleID == sample_id) %>%
          .$sam_names
      }
      
      subset_mrexp <- function(sams){
            mrexp_obj %>% {.[,which(colnames(.) %in% sams)]} %>% 
                 filterData(present = present, depth = 1)
      }
      
      mrexp_pdat <- mrexp_obj %>% pData()
      titration_comp_df %>% 
            mutate(sam_names = pmap(list(sampleID, T1, T2), get_sams),
                   mrexp_sub = map(sam_names, subset_mrexp))
}

```



### DADA2
```{r}
titration_comp_dada2_df <- mrexp$dada2 %>% 
  make_titration_comp_subset_df(titration_comp_df)
```

### Mothur
```{r}
titration_comp_mothur_df <- mrexp$mothur %>%
      make_titration_comp_subset_df(titration_comp_df)
```

### QIIME
```{r}
titration_comp_qiime_df <- mrexp$qiime %>%
      make_titration_comp_subset_df(titration_comp_df)
```

## Estimate logFC 
### MetagenomeSeq - fitFeatureModel  

Functions for calculating logFC using fitFeatureModel
```{r}
fit_model <- function(mrexp_obj, T1, T2, css_p = 0.75){
      ## Normalize count data
      mrexp_obj <- mrexp_obj %>% cumNorm(p = css_p)
      # print(paste("Titrations:", T1, T2))
      pd <- pData(mrexp_obj)
      # defining levels to control the numerator and denominator in logFC estimate
      pd$dilution <- factor(pd$dilution, levels = c(T1, T2)) 
      mod <- model.matrix(~1 + dilution, data = pd) 
      # print("fit model")
      fitFeatureModel(mrexp_obj, mod)
}

fit_safely <- safely(fit_model)
mrcoefs_safely <- safely(MRcoefs)
```

#### DADA2
```{r message = FALSE}
logFC_MgSeq_dada2_df <- titration_comp_dada2_df %>% 
      mutate(fit = pmap(list(mrexp_sub, T1, T2), fit_safely)) 

logFC_MgSeq_dada2_coefs_df <- logFC_MgSeq_dada2_df %>% 
      mutate(fit_coefs = map(fit, ~mrcoefs_safely(.$result, number = Inf))) 

logFC_MgSeq_dada2_coefs_tidy <- logFC_MgSeq_dada2_coefs_df %>% 
      select(sampleID, T1, T2, fit_coefs) %>% 
      mutate(fit_coefs = map(fit_coefs, ~.$result),
             fit_coefs_class = map_chr(fit_coefs, class)) %>% 
      filter(fit_coefs_class == "data.frame") %>% select(-fit_coefs_class) %>% 
      mutate(fit_coefs = map(fit_coefs, rownames_to_column, var = "featureNames")) %>% 
      unnest()
```

#### Mothur
```{r message = FALSE}
logFC_MgSeq_mothur_df <- titration_comp_mothur_df %>% 
      mutate(fit = pmap(list(mrexp_sub, T1, T2), fit_safely)) 

logFC_MgSeq_mothur_coefs_df <- logFC_MgSeq_mothur_df %>% 
      mutate(fit_coefs = map(fit, ~mrcoefs_safely(.$result, number = Inf))) 

logFC_MgSeq_mothur_coefs_tidy <- logFC_MgSeq_mothur_coefs_df %>% 
      select(sampleID, T1, T2, fit_coefs) %>% 
      ## filtering titration comparisons where the model failed
      mutate(fit_coefs = map(fit_coefs, ~.$result),
             fit_coefs_class = map_chr(fit_coefs, class)) %>% 
      filter(fit_coefs_class == "data.frame") %>% select(-fit_coefs_class) %>% 
      mutate(fit_coefs = map(fit_coefs, rownames_to_column, var = "featureNames")) %>% 
      unnest()
```

#### QIIME
```{r message = FALSE}
logFC_MgSeq_qiime_df <- titration_comp_qiime_df %>% 
      mutate(fit = pmap(list(mrexp_sub, T1, T2), fit_safely)) 

logFC_MgSeq_qiime_coefs_df <- logFC_MgSeq_qiime_df %>% 
      mutate(fit_coefs = map(fit, ~mrcoefs_safely(.$result, number = Inf))) 

logFC_MgSeq_qiime_coefs_tidy <- logFC_MgSeq_qiime_coefs_df %>% 
      select(sampleID, T1, T2, fit_coefs) %>% 
      ## excluding titration comparisons where the model failed
      mutate(fit_coefs = map(fit_coefs, ~.$result),
             fit_coefs_class = map_chr(fit_coefs, class)) %>% 
      filter(fit_coefs_class == "data.frame") %>% select(-fit_coefs_class) %>% 
      mutate(fit_coefs = map(fit_coefs, rownames_to_column, var = "featureNames")) %>% 
      unnest()
```

Cleanup fitfeatureModel
```{r}
rm(logFC_MgSeq_dada2_df)
rm(logFC_MgSeq_dada2_coefs_df)
rm(logFC_MgSeq_mothur_df)
rm(logFC_MgSeq_mothur_coefs_df)
rm(logFC_MgSeq_qiime_df)
rm(logFC_MgSeq_qiime_coefs_df)
```

### DESeq2 
Function for converting MRexperiment to DESeqDataSeq
```{r message = FALSE}
## Function based on phyloseq phyloseq_to_deseq2
mrexp_to_deseq2 <- function(mrexp_obj, design, ...){
      ## Extracting count data - no scaling or transformation 
      count_data <- mrexp_obj %>% MRcounts(sl = 1) %>% 
            as.matrix() %>% round(digits = 0)   
      
      ## Sample metadata
      col_data <- pData(mrexp_obj)
      
      ## generating DESeq2 object, performs 
      DESeqDataSetFromMatrix(count_data, col_data, design, ...)
}
```

#### DADA2
```{r}
logFC_DESeq2_dada2_df <- titration_comp_dada2_df %>% 
      mutate(fit = map(mrexp_sub, mrexp_to_deseq2, design = ~dilution),
             fit = map(fit, DESeq, test = "Wald", fitType = "local", quiet = TRUE)) 

logFC_DESeq2_dada2_coefs_df <- logFC_DESeq2_dada2_df %>% 
      mutate(fit_coefs = map(fit, results, cooksCutoff = FALSE)) 

logFC_DESeq2_dada2_coefs_tidy <- logFC_DESeq2_dada2_coefs_df %>% 
      select(sampleID, T1, T2, fit_coefs) %>% 
      mutate(fit_coefs = map(fit_coefs, as.data.frame), 
            fit_coefs = map(fit_coefs, rownames_to_column, var = "featureNames")) %>% 
      unnest()
```

#### Mothur
```{r}
logFC_DESeq2_mothur_df <-titration_comp_mothur_df %>%
      mutate(fit = map(mrexp_sub, mrexp_to_deseq2, design = ~dilution),
            fit = map(fit, DESeq, test = "Wald",quiet = TRUE))#, fitType = "local")) #using default fit 

logFC_DESeq2_mothur_coefs_df <- logFC_DESeq2_mothur_df %>% 
      mutate(fit_coefs = map(fit, results, cooksCutoff = FALSE)) 

logFC_DESeq2_mothur_coefs_tidy <- logFC_DESeq2_mothur_coefs_df %>% 
      select(sampleID, T1, T2, fit_coefs) %>% 
      mutate(fit_coefs = map(fit_coefs, as.data.frame), 
            fit_coefs = map(fit_coefs, rownames_to_column, var = "featureNames")) %>% 
      unnest()
```

Issue with DESeq for at least one of the biological replicate titration pairs. 
To troubleshoot, identifying the biological replicate and titration pairs that fail.
Using `safely` to identify failed tests 
Error message for failed tests - `simpleError in lfproc(x, y, weights = weights, cens = cens, base = base, geth = geth,     ...): newsplit: out of vertex space` 
Based on post on seqanswers (http://seqanswers.com/forums/showthread.php?t=46452) trying using the default fit rather than specifying fit to see if that addresses the error.

```{r}
# deseq_safe <- safely(DESeq)
# titration_DESeq2_logFC_mothur_df <-titration_comp_mothur_df %>%
#       mutate(fit = map(mrexp_sub, mrexp_to_deseq2, design = ~dilution),
#             fit = map(fit, deseq_safe, test = "Wald", fitType = "local"))

# titration_DESeq2_logFC_mothur_df$fit_result <- titration_DESeq2_logFC_mothur_df$fit %>% 
#       map(~.$result)  %>% map_chr(class) 
# titration_DESeq2_logFC_mothur_df$fit_result == "NULL" -> idx
# failed <- titration_DESeq2_logFC_mothur_df[idx,]
# failed$fit
```


#### QIIME
```{r message = FALSE}
logFC_DESeq2_qiime_df <- titration_comp_qiime_df %>% 
      mutate(fit = map(mrexp_sub, mrexp_to_deseq2, design = ~dilution),
             fit = map(fit, DESeq, test = "Wald", fitType = "local", quiet = TRUE)) 

logFC_DESeq2_qiime_coefs_df <- logFC_DESeq2_qiime_df %>% 
      mutate(fit_coefs = map(fit, results, cooksCutoff = FALSE)) 

logFC_DESeq2_qiime_coefs_tidy <- logFC_DESeq2_qiime_coefs_df %>% 
      select(sampleID, T1, T2, fit_coefs) %>% 
      mutate(fit_coefs = map(fit_coefs, as.data.frame), 
            fit_coefs = map(fit_coefs, rownames_to_column, var = "featureNames")) %>% 
      unnest()
```

##### Clean up DESeq2
```{r}
rm(logFC_DESeq2_dada2_df)
rm(logFC_DESeq2_dada2_coefs_df)
rm(logFC_DESeq2_mothur_df)
rm(logFC_DESeq2_mothur_coefs_df)
rm(logFC_DESeq2_qiime_df)
rm(logFC_DESeq2_qiime_coefs_df)
```

### EdgeR
Function for running EdgeR
```{r}
## Function based on phyloseq_to_edgeR code 
## http://joey711.github.io/phyloseq-extensions/edgeR.html
mrexp_to_edgeR <- function(mrexp_obj, group, method = "RLE", ...){
      ## Extracting count data - no scaling or transformation 
      x <- mrexp_obj %>% MRcounts(sl = 1) %>% as.matrix()   
      x <- x + 1 # add 1 to prevent log(0) issues
      
      ## Check `group` argument
      if(identical(all.equal(length(group), 1), TRUE) & ncol(mrexp_obj)> 1){
            ## Assumes grouop is a categorical sample variable name
            group <- pData(mrexp_obj) %>% .[,group]
      }
      
      ## Use taxonomy information at gene annotations
      taxonomy <- fData(mrexp_obj)
      if(!is.null(taxonomy)){
            taxonomy <- taxonomy %>% as.matrix() %>% data.frame()
      }
      
      
      ## Convert into a DGEList
      y <- DGEList(counts = x, group = group, genes = taxonomy, 
                   remove.zeros = TRUE, ...)
      
      
      ## Calc normalization factors 
      z <- edgeR::calcNormFactors(y, method = method)
      
      ## Check for division by zero inside `calcNormFactors`
      if( !all(is.finite(z$amples$norm.factors))){
            stop("Something wrong with edgeR::calcNormFactors on this data, 
                 non-finite $norm.factors, consider changing `method` argument.")
      }
      
      ## Estimate dispersions 
      z %>% estimateCommonDisp() %>% estimateTagwiseDisp()
}
```

#### DADA2
```{r message = FALSE}
logFC_edgeR_dada2_df <- titration_comp_dada2_df %>% 
      mutate(fit = map(mrexp_sub, mrexp_to_edgeR, group = "dilution"),
             fit = map(fit, exactTest)) 

logFC_edgeR_dada2_coefs_df <- logFC_edgeR_dada2_df %>% 
      mutate(fit_coefs = map(fit, topTags, n = Inf, adjust.method = "BH")) 

logFC_edgeR_dada2_coefs_tidy <- logFC_edgeR_dada2_coefs_df %>% 
      select(sampleID, T1, T2, fit_coefs) %>% 
      mutate(fit_coefs = map(fit_coefs, ~.@.Data[[1]])) %>% 
      unnest()
```

#### Mothur
```{r message = FALSE}
logFC_edgeR_mothur_df <- titration_comp_qiime_df %>% 
      mutate(fit = map(mrexp_sub, mrexp_to_edgeR, group = "dilution"),
             fit = map(fit, exactTest)) 

logFC_edgeR_mothur_coefs_df <- logFC_edgeR_mothur_df %>% 
      mutate(fit_coefs = map(fit, topTags, n = Inf, adjust.method = "BH")) 

logFC_edgeR_mothur_coefs_tidy <- logFC_edgeR_mothur_coefs_df %>% 
      select(sampleID, T1, T2, fit_coefs) %>% 
      mutate(fit_coefs = map(fit_coefs, ~.@.Data[[1]])) %>% 
      unnest()
```

#### QIIME
```{r message = FALSE}
logFC_edgeR_qiime_df <- titration_comp_qiime_df %>% 
      mutate(fit = map(mrexp_sub, mrexp_to_edgeR, group = "dilution"),
             fit = map(fit, exactTest)) 

logFC_edgeR_qiime_coefs_df <- logFC_edgeR_qiime_df %>% 
      mutate(fit_coefs = map(fit, topTags, n = Inf, adjust.method = "BH")) 

logFC_edgeR_qiime_coefs_tidy <- logFC_edgeR_qiime_coefs_df %>% 
      select(sampleID, T1, T2, fit_coefs) %>% 
      mutate(fit_coefs = map(fit_coefs, ~.@.Data[[1]])) %>% 
      unnest()
```


## Save logFC tables 
Combine tables for the different pipelines into individual data frames for each differential abundance methods and save as RDS files.  
```{r}
list(dada2 = logFC_MgSeq_dada2_coefs_tidy,
     mothur = logFC_MgSeq_mothur_coefs_tidy,
     qiime = logFC_MgSeq_qiime_coefs_tidy) %>% 
      bind_rows(.id = "pipe") %>% saveRDS("../data/logFC_MgSeq_df.rds") 

list(dada2 = logFC_DESeq2_dada2_coefs_tidy,
     mothur = logFC_DESeq2_mothur_coefs_tidy,
     qiime = logFC_DESeq2_qiime_coefs_tidy) %>% 
      bind_rows(.id = "pipe") %>% saveRDS("../data/logFC_DESeq2_df.rds") 

list(dada2 = logFC_edgeR_dada2_coefs_tidy,
     mothur = logFC_edgeR_mothur_coefs_tidy,
     qiime = logFC_edgeR_qiime_coefs_tidy) %>% 
      bind_rows(.id = "pipe") %>% saveRDS("../data/logFC_edgeR_df.rds")
```


## Next Steps 

- Implement permutation based method for calculating logFC and variance estimates  
- Work out methods for using different normalization methods  
- Look into titrations comparisons with build errors 

## Session Information 
```{r}
s_info <- devtools::session_info()
print(s_info$platform)
knitr::kable(s_info$packages)
```
