---
title: "Feature Inference DADA2 Oligotype Analysis"
author: "Nate Olson"
date: '`r Sys.Date()`'
output:
  pdf_document: default
  html_document: default
---

```{r setup, message = FALSE, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dada2)
library(tidyverse)
library(Biostrings)
dada_pipe_dir <- "/Users/nolson/Projects/16S_etec_mix_study/analysis/pipelines/dada2"
dada_data_dir <- file.path(dada_pipe_dir, "processed_data")
```

## Objective 
Characterization of DADA2 features. 
Due to the relatively low number of features per 

### Oligotype Analysis for most abundant feature

Extracting sequences for the most abundant feature and saving as fasta.
```{r}
## seqTable generated in `2017-03-02-Feature-Inference-Unique-Seq-Distribution.Rmd`
seqTable <- readRDS("E01JH00011_plate1_half1_dada_seq_tbl.rds")
abu_seq <- seqTable %>% group_by(seq) %>% summarise(count = n()) %>% 
      arrange(desc(count)) %>% .$seq %>% .[1]
names(abu_seq) <- "top_feature"
abu_seqTable <- seqTable %>% filter(seq == abu_seq)
seqs <- abu_seqTable$derepSeq 
names(seqs) <- paste0(abu_seqTable$sampleID,"-", abu_seqTable$id)

## saving sequences to file
fasta_file <- file.path(dada_pipe_dir,"oligotype_analysis", "E01JH00011_top_feature.fasta")
# DNAStringSet(seqs) %>% writeXStringSet(fasta_file)

## Saving unique sequences
fasta_file <- file.path(dada_pipe_dir,"oligotype_analysis", "E01JH00011_top_feature_unique.fasta") 
seq_df <- data_frame(seqs = seqs, seq_name = names(seqs)) %>% 
     group_by(seqs) %>% sample_n(1)
seqs <- seq_df$seqs
names(seqs) <- seq_df$seq_name
seqs %>% DNAStringSet() %>% writeXStringSet(fasta_file)

## Saving feature representative sequence
fasta_file <- file.path(dada_pipe_dir,"oligotype_analysis", "E01JH00011_top_feature_seq.fasta")
# DNAStringSet(x = abu_seq) %>% writeXStringSet(fasta_file)
```

#### Performing oligotype analysis

Oligotype analysis performed on the command line
```
## Calculate Entropy
entropy-analysis E01JH00011_top_feature.fasta 

## Oligotyping
oligotype E01JH00011_top_feature.fasta \\
      E01JH00011_top_feature.fasta-ENTROPY \\
      --quick -c 1 -M 20 -A 30 -s 4
## Entropy analysis on oligotypes
for i in 00000_C 00001_G 00002_A 00003_T; do entropy-analysis $i --no-display; done
for i in 00000_TC 00001_TG 00002_GC 00003_GG 00004_TA 00005_TT 00006_AC 00007_GA ; do entropy-analysis $i --no-display; done
```

```{r fig.cap = "Shannon entropy by sequence position for all sequences assigned to the most abundant feature.", fig.width=4}
fasta_file <- file.path(dada_pipe_dir,"oligotype_analysis", "E01JH00011_top_feature.fasta")
knitr::include_graphics(paste0(fasta_file, "-ENTROPY.png"))
```

After entropy decomposition (oligotyping) using a single base position the most abundant oligotype has low entropy values( ~0.2). 
These values are defined as the noise level for Illumina sequencing data in the Oligotyping article. 
For the other oligotypes the entropy level is higher. 
Inspection of the sequences indicates that the sequences have different starting positions which resulted in high entropy values. 
```{r fig.width = 3}
oligorep_dir <- "E01JH00011_top_feature-sc1-s1-a0.0-A0-M0/OLIGO-REPRESENTATIVES/"
entropy_figs <- file.path(dada_pipe_dir,"oligotype_analysis", oligorep_dir) %>% 
      paste0(., "0000",c("0_G","1_A","2_T","3_C"), "-ENTROPY.png")
knitr::include_graphics(entropy_figs)
```


## Unique Sequence Analysis 
Greater than expected sequence diversity was observed for unique sequences in the top feature.
To evaluate the taxonomic diversity of sequences assigned to the top feature using mothur 
```
classify.seqs(fasta=E01JH00011_top_feature.fasta, reference=/Users/nolson/Projects/16S_etec_mix_study/analysis/pipelines/mothur/data/references/trainset.10_082014.V34.fasta, taxonomy=/Users/nolson/Projects/16S_etec_mix_study/analysis/pipelines/mothur/data/references/trainset.10_082014.V34.tax, cutoff=80, processors=7)
```

```{r}
tax_class <- read_tsv(file.path(dada_pipe_dir,"oligotype_analysis", "E01JH00011_top_feature.V34.wang.tax.summary"))
```

Of the 62075 total sequences assigned to the feature, 67072 were assigned to the Bacterial phylum and 3 to the Archaea phylum, while 110 sequences were unable to be classified. 
The unclassified sequences are non-16S sequences and represent high error, artifact reads being assigned to the feature. 

```{r}
tax_class %>% filter(taxlevel == 2)
```

Mothur code for removing unclassified lineages
```
remove.lineage(taxonomy=E01JH00011_top_feature.V34.wang.taxonomy, taxon=unknown)
```

__Include alignment of unclassified sequences to representative sequence__

Of the classified sequences, the sequences were assigned to 7 different phyla, along with 230 seqeunces that could not be classified past the division level (Bacteria_unclassified), and the 110 unclassified sequences. 
```{r}
tax_class %>% filter(taxlevel == 3)
```

Removing unclassified sequences from the sequence set
```{r}
tax_classified <-  file.path(dada_pipe_dir,
                             "oligotype_analysis", 
                             "E01JH00011_top_feature.V34.wang.pick.taxonomy") %>% 
      read_tsv(col_names = FALSE) %>% 
      mutate(X1 = gsub("_",":", x= X1))
```

```{r}
## Total Classified Sequences with Frequency
fasta_file <- file.path(dada_pipe_dir,"oligotype_analysis", "E01JH00011_top_feature_classified.fasta")
seqs <- abu_seqTable$derepSeq 
names(seqs) <- paste0(abu_seqTable$sampleID,"-", abu_seqTable$id)
classified_seqs <- seqs[names(seqs) %in% tax_classified$X1]
classified_seqs %>% DNAStringSet() %>% writeXStringSet(fasta_file)

## Unique Classified Sequences
fasta_file <- file.path(dada_pipe_dir,"oligotype_analysis", "E01JH00011_top_feature_unique_classified.fasta")
seq_df <- data_frame(seqs = classified_seqs, seq_name = names(classified_seqs)) %>% 
      group_by(seqs) %>% sample_n(1)
seq_df$seqs %>% length()
unique_classified_seqs <- seq_df$seqs
names(unique_classified_seqs) <- seq_df$seq_name
unique_classified_seqs %>% DNAStringSet() %>% writeXStringSet(fasta_file)
```

__Next Steps__ Does it make sense to do entropy analysis on the remain sequences, taxonomic classification indicates the within feature sequence diversity.

## Multiple Sequence Alignment 
Issue relating to different starting position for dereplicated sequences performing multiple sequence alignment to see if that clears up the issue.
Using mothur align.seqs with SILVA referencde alignment to address multiple starting positions issue. 
Issue with long and short sequence, removing using screen.seqs. 
Using filter.seqs to remove gap positions, may want to filter end positions with "." as well.  

```
align.seqs(fasta=E01JH00011_top_feature_unique_classified.fasta, reference=/Users/nolson/Projects/16S_etec_mix_study/analysis/pipelines/mothur/data/references/silva.nr_v119.V34.align, flip=TRUE, processors = 7)
summary.seqs(fasta=current)
filter.seqs(fasta=E01JH00011_top_feature.align, vertical=T, trump=-)

```

Entropy analysis for aligned sequences. 
```{r fig.cap = "Shannon entropy by sequence position for all sequences assigned to the most abundant feature.", fig.width=4}
aligned_entropy_file <- file.path(dada_pipe_dir,"oligotype_analysis", "E01JH00011_top_feature_classified.fasta")
knitr::include_graphics(paste0(aligned_entropy_file, "-ENTROPY.png"))
```



## Within Feature Pairwise Seq Distances
Using the MSA of the classified seuqences calculated the pairwise distances between sequenes using mothur's `dist.seq` (see commands above).

```{r}
dist_mat <- read_delim("/Users/nolson/Projects/16S_etec_mix_study/analysis/pipelines/dada2/oligotype_analysis/E01JH00011_top_feature_unique_curated.filter.dist",col_names = FALSE, delim = " ")
```
```{r}
dist_mat %>% ggplot() + geom_density(aes(x = X3)) + 
      theme_bw() + labs(x = "Distance", y = "Density") + 
      geom_vline(aes(xintercept = 0.03), color = "darkorange")
```


