---
title: "Feature Inference Discrepancies - DADA2"
author: "Nate Olson"
date: '`r Sys.Date()`'
output:
  html_document: default
  pdf_document: default
---

```{r setup, message = FALSE, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# library(ProjectTemplate)
# cwd <- getwd()
# setwd("../")
# load.project()
# setwd(cwd)
library(dada2)
library(tidyverse)
library(Biostrings)
dada_pipe_dir <- "/Users/nolson/Projects/16S_etec_mix_study/analysis/pipelines/dada2/"
dada_data_dir <- file.path(dada_pipe_dir, "processed_data")
```

## Objective 
A relatively low number of inferred features per samples identified using the sequence inference pipeline (DADA2) relative to the open-reference clustering (QIIME) and de-novo clustering (Mothur) pipelines. 
This low number of inferred features could be due to grouping sequences that are representative of distinct biological units. 

## Approach
To evaluate the composition of sequences in the DADA2 features. 
Initial analysis of biological replicate E01JH00011, PCR replicates from half of plate 1. 

1. Investigate the within feature pairwise sequence distances.  
2. Characterize the distribution of sequences assigned to a feature across titrations, the assumption is that unrelated sequences will have different distributions.  


## Getting Data for Analysis
The `dada_to_seq_table` function generates a table with the denoised sequence, unique sequence, and the id for the input sequence from the fastq file. 


```{r}
derepFs <- file.path(dada_data_dir, "derepFs-2016-11-07.rds") %>% readRDS()
dadaFs <- file.path(dada_data_dir, "dadaFs-single-inference-2016-11-07.rds") %>% 
      readRDS()
```

```{r}
get_seqtable <- function(sam){
      sr <- ShortRead::readFastq("~/Projects/16S_etec_mix_study/analysis/pipelines/dada2/processed_data/",
                           pattern = pats <- paste0(sam,".*R1.*filt.fastq.gz"))
      dadaRes <- dadaFs[[sam]]
      derep <- derepFs[[sam]]
      dada2:::dada_to_seq_table(dadaRes = dadaRes, derep = derep, sr = sr)
}
```

```{r}
## sample IDs
sams <- paste0("1-",LETTERS[1:8],"2") %>% c(.,"1-B2")
sam_list <- as.list(sams) %>% set_names(sams)
seqTable <- map_df(.x = sam_list, .f = get_seqtable, .id = "sampleID")
saveRDS(seqTable, "E01JH00011_plate1_half1_dada_seq_tbl.rds")
```

seqTable data description
sampleID - PCR (sample) the sequence is from
seq - representative sequence, most abundant sequence in the feature
n0 - could not find in dada2 documentation
uniqueIndex - dada2 assigned index
genotypeIndex - dada2 assigned index 
derepSeq - unique sequence
id - sequence id for source read

```{r}
seqTable %>% glimpse()
```


### Distribution of dereplicated sequences across samples.
Extracting the sequence top 4 and most abundant features.  

```{r}
top_seqs <- seqTable %>% group_by(seq) %>% summarise(count = n()) %>% 
      arrange(desc(count)) %>% .$seq %>% .[1:4]
top_seqTable <- seqTable %>% filter(seq %in% top_seqs)
top_sam_seqTable <- top_seqTable %>% group_by(sampleID, seq, derepSeq) %>% summarise(count = n())
```

Sequences with the abundances consistently above 10 tend to have similar distribution patterns to the most abundant supporting the hypothesis that the sequences in the feature are representatives of the sample biological units. 
May want to consider testing for consistencies in abundances distributions between the samples. 

```{r fig.cap="Distribution of unique sequences assigned to the 10 most abundant features for biological replicate 2, samples from the first half of PCR plate 1. Each line represents the abundance of a unique (dereplicated) sequence across samples."}
top_sam_seqTable %>%  ggplot() + 
      geom_path(aes(x = sampleID, y = count, group = derepSeq)) + 
      scale_y_log10() + theme_bw() + facet_wrap(~seq, nrow = 1) + 
      theme(axis.text.x = element_text(angle = 90))
```



## Ranking features
Based on conversation with Hector (3/1), for each feature 
1. center and scale unique sequence abundance distribution across samples 
2. center values by sample 
3. calculate total feature variance as $\sqrt{SS}$

```{r fig.cap="Diagram of variance ranking method"}
knitr::include_graphics("IMG_2467.JPG")
```

Use total variance calculation to rank features, looking at the features with the greatest variance. 

### Oligotype Analysis for most abundant feature

Extracting sequences for the most abundant feature and saving as fasta.
```{r}
abu_seq <- seqTable %>% group_by(seq) %>% summarise(count = n()) %>% 
      arrange(desc(count)) %>% .$seq %>% .[1]
names(abu_seq) <- "top_feature"
abu_seqTable <- seqTable %>% filter(seq == abu_seq)
seqs <- abu_seqTable$derepSeq 
names(seqs) <- paste0(abu_seqTable$sampleID," ", abu_seqTable$id)

## saving sequences to file
fasta_file <- file.path(dada_pipe_dir,"oligotype_analysis", "E01JH00011_top_feature.fasta")
DNAStringSet(seqs) %>% writeXStringSet(fasta_file)

## Saving unique sequences
fasta_file <- file.path(dada_pipe_dir,"oligotype_analysis", "E01JH00011_top_feature_unique.fasta")
unique(seqs) %>% DNAStringSet() %>% writeXStringSet(fasta_file)

## Saving feature representative sequence
fasta_file <- file.path(dada_pipe_dir,"oligotype_analysis", "E01JH00011_top_feature_seq.fasta")
DNAStringSet(x = abu_seq) %>% writeXStringSet(fasta_file)
```

#### Performing oligotype analysis

Oligotype analysis performed on the command line
```
## Calculate Entropy
entropy-analysis E01JH00011_top_feature.fasta 

## Oligotyping
oligotype E01JH00011_top_feature.fasta \\
      E01JH00011_top_feature.fasta-ENTROPY \\
      --quick -c 1 -M 20 -A 30 -s 4
## Entropy analysis on oligotypes
for i in 00000_C 00001_G 00002_A 00003_T; do entropy-analysis $i --no-display; done
for i in 00000_TC 00001_TG 00002_GC 00003_GG 00004_TA 00005_TT 00006_AC 00007_GA ; do entropy-analysis $i --no-display; done
```


Issues with the oligotyping based approach - high entropy values even after splitting into different oligotypes. 

```{r fig.cap = "Shannon entropy by sequence position for all sequences assigned to the most abundant feature.", fig.width=4}
fasta_file <- file.path(dada_pipe_dir,"oligotype_analysis", "E01JH00011_top_feature_unique.fasta")
knitr::include_graphics(paste0(fasta_file, "-ENTROPY.png"))
```

After entropy decomposition (oligotyping) using a single base position the most abundant oligotype has low entropy values( ~0.2). 
These values are defined as the noise level for Illumina sequencing data in the Oligotyping article. 
For the other oligotypes the entropy level is higher. 
Inspection of the sequences indicates that the sequences have different starting positions which resulted in high entropy values. 
```{r fig.width = 3}
oligorep_dir <- "E01JH00011_top_feature-c1-s4-a0.0-A30-M20/OLIGO-REPRESENTATIVES/"
entropy_figs <- file.path(dada_pipe_dir,"oligotype_analysis", oligorep_dir) %>% 
      paste0(., "0000",c("0_C","1_G","2_A","3_T"), "-ENTROPY.png")
knitr::include_graphics(entropy_figs)
```

Issue relating to different starting position for dereplicated sequences performing multiple sequence alignment to see if that clears up the issue
Using mothur align.seqs with SILVA referencde alignment to address multiple starting positions issue. 
Issue with long and short sequence, removing using screen.seqs. 
Using filter.seqs to remove gap positions, may want to filter end positions with "." as well.  

```
align.seqs(fasta=E01JH00011_top_feature.fasta, reference=/Users/nolson/Projects/16S_etec_mix_study/analysis/pipelines/mothur/data/references/silva.nr_v119.V34.align, processors = 7)
summary.seqs(fasta=current)
screen.seqs(fasta=E01JH00011_top_feature.align, minlength=270, maxlength=275)
filter.seqs(fasta=E01JH00011_top_feature.good.align, vertical=T, trump=-)
```

Entropy analysis for aligned sequences, not sure why so many fewer sequences after mothur might be filtering sequences intead of columns....
```{r fig.cap = "Shannon entropy by sequence position for all sequences assigned to the most abundant feature.", fig.width=4}
aligned_entropy_file <- file.path(dada_pipe_dir,"oligotype_analysis", "E01JH00011_top_feature.good.filter.fasta")
knitr::include_graphics(paste0(aligned_entropy_file, "-ENTROPY.png"))
```


Entropy analysis for Oligotypes after sequence alignment
```{r fig.width = 3}
oligorep_dir <- "E01JH00011_top_feature-c1-s4-a0.0-A30-M20/OLIGO-REPRESENTATIVES/"
entropy_figs <- file.path(dada_pipe_dir,"oligotype_analysis", oligorep_dir) %>% 
      paste0(., "0000",c("0_C","1_G","2_A","3_T"), "-ENTROPY.png")
knitr::include_graphics(entropy_figs)
```

Oligotyping unaligned reads at base 31
```{r}
oligorep_dir <- "E01JH00011_top_feature-sc1-s1-a0.0-A0-M0/OLIGO-REPRESENTATIVES/"
entropy_figs <- file.path(dada_pipe_dir,"oligotype_analysis", oligorep_dir) %>% 
      paste0(., "0000",c("0_G","1_A","2_T","3_C"), "-ENTROPY.png")
knitr::include_graphics(entropy_figs)
```


## Investigation of Top Feature Sequences
Saved the unique sequences for the most abundant feature. 
Manually curated the sequences, removing sequences with a large number of differences from the feature representative sequence. 

Entropy analysis of curated sequences (does not include sequence abundances). 
Similar to the entropy analysis for all sequences. 
There is a clear peak at position 31 and the entropy increases towards the end of the sequence. 
This increase in entropy towards the end of the sequence correlates with the increase in sequencing errors toward the end of the forward reads. 

```{r fig.cap = "Shannon entropy by sequence position for all sequences assigned to the most abundant feature.", fig.width=4}
aligned_entropy_file <- file.path(dada_pipe_dir,"oligotype_analysis",
                                  "E01JH00011_top_feature_unique_manual_cleanup.fasta")
knitr::include_graphics(paste0(aligned_entropy_file, "-ENTROPY.png"))
```

Performing multiple sequence alignment (MSA) for the sequences that were removed from the unique sequence list. 
Based on the MSA, what appear to be non-16S reads were assigned to the feature. 


## Within Feature Pairwise Seq Distances
1. Obtain unique sequences assigned to individual features 
2. Calculate pairwise distances on unique sequences - may need to do on forward and reverse reads individually 
      1. calculate pairwise distances using ..... Mothur with sequence alignment?

```{r}
dist_mat <- read_delim("/Users/nolson/Projects/16S_etec_mix_study/analysis/pipelines/dada2/oligotype_analysis/E01JH00011_top_feature.good.filter.dist",col_names = FALSE, delim = " ")
```
```{r}
dist_mat %>% ggplot() + geom_density(aes(x = X3)) + theme_bw() + labs(x = "Distance", y = "Density")
```


```{r}
dist_mat %>% arrange(desc(X3))
```












