---
title: "Feature-Level Relative Abundance Error Analysis"
output: html_notebook
---

* Next steps add taxonomy to phylo data 
* phylo analysis for other pipelines - may want to generate functions for this ...
* mean-variance correlation  
* summarize results

```{r fLevelRelAbuSetup, include=FALSE}
## TODO feature count table
library(tidyverse)
library(forcats)
library(phylosignal)
library(adephylo)
library(ape)
library(phylobase)
library(ggtree)
nb_counts <- readRDS("~/Desktop/nb_counts_titrations.RDS")
pa_summary_anno_df <- readRDS("~/Desktop/pa_summary_anno_df.RDS")
theta_est <- readRDS("~/Desktop/bootstrap_theta_estimates.rds")
```

### Overview
Feature-level characterization of relative abundance outlier features.

- Is there a taxonomic specific effect?
      - Outlier taxonomy 
      - phylogenetic signal for Median Error Rate and RCOV  

- Can the higher error rate be attributed to lower relative abundance/ signal?  
      - Correlation with median relative abundance  
- Is the higher feature-level error rate due to lower accuracy pre-post relative abundance estiamtes 
      - Are these featues outlier in the Mean-variance pre-post 
      
      
## Approach
* Feature-level metrics 
- from relative abundance results Rmd 
      - Calculate feature-level median relative abundance 

## Conclusions
* When comparing the median feature-level relative abundance to the median error rate and RCOV. The median relative abundance was lower for features identified as outliers based on the feature-level error rate, but only when considering features with positive error rates. This is likely due to the more extreme error rates all being positive. For the RCOV the feature-level median relative abundance values were not significantly different between the outlier and non-outlier features. 
* 
```{r relAbuMunge, echo = FALSE, message = FALSE, warning = FALSE}
pre_post_prop <- nb_counts %>% 
      ungroup() %>% 
      filter(t_fctr %in% c(0,20)) %>% 
      mutate(end_point = if_else(t_fctr == 0 , "post", "pre")) %>% 
      dplyr::select(-t_fctr) %>% 
      ## setting values to 0 when one or more of the PCR replicates are 0 for titration end-points
      spread(end_point,nb_prop, fill = 0)

prop_inferred <- theta_est %>% 
      filter(pipe == "unclustered") %>% 
      ungroup() %>%
      mutate(t_fctr = factor(t_fctr, levels = c(0:5, 10, 15, 20))) %>% 
      dplyr::select(biosample_id, theta_hat_mean, t_fctr) %>% 
      right_join(nb_counts) %>% right_join(pre_post_prop) %>% 
      filter(t_fctr %in% c(1:5,10,15)) %>% 
      ## Using inferred theta estimates to calculate expected values
      mutate(inferred_prop = post * theta_hat_mean + pre * (1-theta_hat_mean))

## Excluding mix and unmix specific features
## Only including features observed in all or none of the four pre- post- PCR replicates
## Features with relative abundance estimates and expected values less than 1e-5, these are features that we would not expect to consistently observe in a PCR replicate for the given sequencing depth, ~100k 
## Excluding titrations where the inferred theta values are less than 1
pa_filter <- pa_summary_anno_df %>% 
      filter(pa_specific == "unspecific") %>% 
      dplyr::select(biosample_id, pipe, feature_id, full_pre, T00, T20, pa_mixed) %>% 
      filter(T00 %in% c(0,4), T20 %in% c(0,4))

prop_inferred <- prop_inferred %>% 
      right_join(pa_filter) %>% 
      filter(nb_prop > 1e-5, 
             inferred_prop > 1e-5,
             theta_hat_mean > 0)


#### Error Rate Calculations
rel_abu_error <- prop_inferred %>% 
      mutate(t_fctr = factor(t_fctr, levels = c(1:5, 10, 15))) %>% 
      mutate(inferred_error = nb_prop - inferred_prop,
             inferred_error_rate = inferred_error/inferred_prop) 

rel_abu_error_summary <-  rel_abu_error %>% 
      group_by(pipe, biosample_id, feature_id) %>% 
      summarise(median_rel_abu = median(nb_prop),
            median_error = median(inferred_error_rate),
                iqr_error = IQR(inferred_error_rate),
                rcov_error = iqr_error/abs(median_error), 
                mean_error = mean(inferred_error_rate),
                var_error = var(inferred_error_rate),
                cov_error = var_error/mean_error) 

### Error rate boxplot and outlier annotation
error_boxplot <- rel_abu_error %>% group_by(pipe, biosample_id, feature_id) %>% 
      summarise(median_error = median(inferred_error_rate)) %>%
      ggplot() + 
      geom_boxplot(aes(x = pipe, y = median_error), outlier.shape = NA) + 
      facet_wrap(~biosample_id, nrow = 1) + 
      theme_bw() + theme(axis.text.x = element_text(angle = 90)) +
      labs(x = "Individual", y = "Median")

## Annotating features as outliers based on boxplot
error_plot_dat <- ggplot_build(error_boxplot)$data[[1]] %>% 
      mutate(pipe = fct_recode(factor(group), 
                               dada2 = "1", 
                               mothur = "2", 
                               qiime = "3",
                               unclustered = "4"),
             biosample_id = fct_recode(PANEL, 
                                       E01JH0004 = "1", 
                                       E01JH0011 = "2", 
                                       E01JH0016 = "3", 
                                       E01JH0017 = "4", 
                                       E01JH0038 = "5"))
outlier_error_dat <- error_plot_dat %>% 
      dplyr::select(ymin, ymax, pipe, biosample_id)

rel_error_outlier_cat <- rel_abu_error_summary %>% 
      left_join(outlier_error_dat) %>% 
      mutate(outlier_cat = if_else(median_error < ymin | median_error > ymax, "outlier","inlier")) 

## Robust COV Analysis
rcov_boxplot <- rel_abu_error_summary %>%
      ggplot() + geom_boxplot(aes(x = pipe, y = rcov_error), outlier.shape = NA) + 
      facet_wrap(~biosample_id, nrow = 1) + 
      theme_bw() + theme(axis.text.x = element_text(angle = 90)) +
      labs(x = "Individual", y = "RCOV") 


## Annotating features as outliers based on boxplot
rcov_plot_dat <- ggplot_build(rcov_boxplot)$data[[1]] %>% 
     mutate(pipe = fct_recode(factor(group), 
                               dada2 = "1", 
                               mothur = "2", 
                               qiime = "3",
                               unclustered = "4"),
             biosample_id = fct_recode(PANEL, 
                                       E01JH0004 = "1", 
                                       E01JH0011 = "2", 
                                       E01JH0016 = "3", 
                                       E01JH0017 = "4", 
                                       E01JH0038 = "5"))
outlier_rcov_dat <- rcov_plot_dat %>% 
      dplyr::select(ymin, ymax, pipe, biosample_id)

rcov_outlier_cat <- rel_abu_error_summary %>% 
      left_join(outlier_rcov_dat) %>% 
      mutate(outlier_cat = if_else(rcov_error < ymin | rcov_error > ymax, "outlier","inlier"))

### Plot Code -----------------------------------------------------------------
## Observed v. Expected Scatter plot
relAbuOvE <- prop_inferred %>% 
      ggplot() + 
      geom_point(aes(x = inferred_prop, y = nb_prop), alpha = 0.15) +
      geom_abline(aes(intercept = 0, slope = 1), color = "darkorange") +
      geom_smooth(aes(x = inferred_prop, y = nb_prop)) +
      facet_grid(pipe~biosample_id) +
      scale_y_log10() + scale_x_log10() +
      theme_bw() +
      labs(x = "Expected", 
           y = "Observed")

# ggsave(relAbuOvE, "~/Desktop/quant_exp_vs_obs.png", dpi = 450)

## Median Error Pipeline Comparison
ymin <- ggplot_build(error_boxplot)$data[[1]]$ymin %>% min()
ymax <- ggplot_build(error_boxplot)$data[[1]]$ymax %>% max()
error_boxplot <- error_boxplot + coord_cartesian(ylim = c(ymin, ymax))

# ggsave(error_boxplot, "~/Desktop/quant_bias.png", dpi = 450) 

## RCOV Error Pipeline Comparison
ymin <- ggplot_build(rcov_boxplot)$data[[1]]$ymin %>% min()
ymax <- ggplot_build(rcov_boxplot)$data[[1]]$ymax %>% max()
rcov_boxplot <- rcov_boxplot + coord_cartesian(ylim = c(ymin, ymax))

# ggsave(rcov_boxplot, "~/Desktop/quant_variance.png", dpi = 450)
```

```{r}
rel_error_summary <- rel_error_outlier_cat %>% 
      rename(error_cat = outlier_cat) %>% 
      select(-ymin, -ymax) %>% 
      left_join(rcov_outlier_cat) %>% 
      rename(rcov_cat = outlier_cat)
```

## Relationship between the median relative abundance and median error 
Higher median error rates can not be attributed lower relative abundance. 
```{r}
rel_error_summary %>% 
      ggplot() + 
      geom_point(aes(x = median_rel_abu, y = median_error, fill = error_cat), shape = 21) + 
      theme_bw() + 
      scale_x_log10() +       
      labs(x = "Median Relative Abundance", y = "Median Error Rate", fill = "Boxplot Outlier")
```

Comparing the median feature-level relative abundance between outlier and non-outliers (inliers)
```{r}
rel_error_summary %>% filter(median_error > 0) %>% 
      ggplot() + 
      geom_boxplot(aes(x = error_cat, y = median_rel_abu)) + 
      theme_bw() + 
      scale_y_log10()
```
Using a non-parametric test as the data are not normally distributed. 
The median relative abundance error rate is not significantly different between the two sets of features.
```{r}
kruskal.test(log10(median_rel_abu) ~ factor(error_cat), data = rel_error_summary)
```

However, when only including features with median error rates > 0, the difference is statistically significant.
```{r}
kruskal.test(log10(median_rel_abu) ~ factor(error_cat), data = filter(rel_error_summary, median_error > 0))
```




## Relative RCOV and Median Relative abundance
The error RCOV and median relative abundance were correlated
```{r}
rel_error_summary %>%  
      ggplot() + 
      geom_point(aes(x = median_rel_abu, y = rcov_error, fill = rcov_cat), shape = 21) + 
      theme_bw() + 
      scale_x_log10() + scale_y_log10() +    
      labs(x = "Median Relative Abundance", y = "Error RCOV", fill = "Boxplot Outlier")
```

Comparing the median feature-level relative abundance between outlier and non-outliers (inliers) for RCOV
```{r}
rel_error_summary %>%
      ggplot() + 
      geom_boxplot(aes(x = rcov_cat, y = median_rel_abu)) + 
      theme_bw() + 
      scale_y_log10()
```

Using a non-parametric test as the data are not normally distributed. 
The median relative abundance error rate is not significantly different between the two sets of features.
```{r}
kruskal.test(log10(median_rel_abu) ~ factor(error_cat), data = rel_error_summary)
```

## Phylogenetic signal code

```{r message = FALSE}
dada_tree <- readRDS("~/Projects/mgtst_pipelines/dada2/dada_tree_GTR.rds") %>% .$tree

dada_metrics <- rel_error_summary %>%  
      filter(feature_id %in% dada_tree$tip.label) %>% 
      filter(biosample_id == "E01JH0016")

tips_to_drop <- dada_tree$tip.label[!(dada_tree$tip.label %in% dada_metrics$feature_id)] 

dada_trim <- drop.tip(dada_tree,tip = tips_to_drop)

dada_labels <-  data_frame(tip.label = dada_trim$tip.label)
dada_dat <- list()
dada_dat$median_error <- dada_metrics$median_error
dada_dat$rcov_error <- dada_metrics$rcov_error
dada_dat$tip.label <- dada_metrics$feature_id
dada_dat <- as.data.frame(dada_dat) %>% right_join(dada_labels) %>% 
      column_to_rownames(var = "tip.label")
dada_p4d <- phylo4d(dada_trim, dada_dat)
dada_lipa <- lipaMoran(dada_p4d)
barplot.phylo4d(dada_p4d, bar.col=(dada_lipa$p.value < 0.05) + 1, center = FALSE , scale = FALSE)
```

a significant phylogenetic signal was detected
```{r message = FALSE}
phyloSignal(p4d = dada_p4d, method = c("I","Cmean"))
``` 

```{r message = FALSE}
dada_fData <- readRDS("../../mgtst_pipelines/dada2/dada_mrexp.rds") %>% .@featureData

dada_tax <- dada_fData@data %>% 
      as.data.frame() %>% 
      rename(id = OTUname) %>% 
      filter(id %in% dada_trim$tip.label)

dada_labels <-  data_frame(id = dada_trim$tip.label)

dada_dat <- dada_metrics %>% 
      rename(id = feature_id) %>% 
      right_join(dada_labels) %>% 
      dplyr::select(id, median_error, rcov_error) %>% 
      right_join(dada_tax)

dada_trim <- groupOTU(dada_trim, split(dada_tax$id, dada_tax$Rank3), group_name = "Class")

p <- ggtree(dada_trim, aes(color = Class))
p + theme(legend.position = "right")
## Code is not working ...:(
# 
# p <- facet_plot(p, panel = 'Median Error', data = dada_dat, 
#                 geom=geom_segment, aes(x=0, xend=median_error, y=y, yend=y), size=1)
# 
# facet_plot(p, panel = 'RCOV', data = dada_dat, 
#            geom=geom_segment, aes(x=0, xend=rcov_error, y=y, yend=y), size=1) +
#       theme_tree2() + theme(legend.position = "bottom")
```

## QIIME
```{r message = FALSE}
qiime_tree <- read.tree("~/Projects/mgtst_pipelines/qiime/otus_uc_fast/rep_set.tre")
qiime_metrics <- feature_metrics %>%
      # filter(feature_id %in% qiime_tree$tip.label) %>%
      filter(pipe == "qiime", biosample_id == "E01JH0016")

## Only including tips in logFC feature summary (pre)
tips_to_drop <- qiime_tree$tip.label[!(qiime_tree$tip.label %in% qiime_metrics$feature_id)]
qiime_trim <- drop.tip(qiime_tree,tip = tips_to_drop)

## Data frame with metrics for comparison
qiime_labels <-  data_frame(tip.label = qiime_trim$tip.label)
qiime_dat <- qiime_metrics %>% rename(tip.label = feature_id) %>% 
      right_join(qiime_labels) %>% 
      dplyr::select(tip.label, slope_metric, adj.r.squared) %>% 
      as.data.frame() %>% dplyr::select(-tip.label)

qiime_p4d <- phylo4d(qiime_trim, qiime_dat)
qiime_lipa <- lipaMoran(qiime_p4d)
barplot.phylo4d(qiime_p4d, bar.col=(qiime_lipa$p.value < 0.05) + 1, center = FALSE , scale = FALSE)
```

```{r message = FALSE}
phyloSignal(p4d = qiime_p4d, method = c("I","Cmean"))
``` 

```{r message = FALSE}
qiime_labels <-  data_frame(id = qiime_trim$tip.label)
qiime_dat <- qiime_metrics %>% 
      rename(id = feature_id) %>% 
      right_join(qiime_labels) %>% 
      dplyr::select(id, slope_metric, adj.r.squared) 

qiime_tax <- logFC_pre_taxa %>% 
      filter(pipe == "qiime") %>% 
      rename(id = feature_id) %>% 
      left_join(qiime_labels)

qiime_trim <- groupOTU(qiime_trim, split(qiime_tax$id, qiime_tax$Rank3), group_name = "Class")



p <- ggtree(qiime_trim, aes(color = Class))

p <- facet_plot(p, panel = 'Slope Metric', data = qiime_dat, 
                geom=geom_segment, aes(x=0, xend=slope_metric, y=y, yend=y), size=1)

facet_plot(p, panel = 'R2', data = qiime_dat, 
           geom=geom_segment, aes(x=0, xend=adj.r.squared, y=y, yend=y), size=1) +
      theme_tree2() + theme(legend.position = "bottom")
```

## Mothur
Mothur tree node ids and cluster ids do not match up....
```{r message = FALSE}
seq_id_df <- read_lines("~/Projects/mgtst_pipelines/mothur/mgtst.trim.contigs.good.unique.good.filter.unique.precluster.pick.opti_mcc.unique_list.0.03.rep.fasta") %>% grep(pattern = ">", value = TRUE) %>% 
      str_replace("\\|.*","") %>% 
      str_replace(">","") %>% 
      enframe(name = "X", value = "seq_id") %>% 
      select(-X) %>% 
      separate(seq_id, into = c("seq_id", "feature_id"),sep = "\t")


mothur_tree <- read.tree("~/Projects/mgtst_pipelines/mothur/mgtst.trim.contigs.good.unique.good.filter.unique.precluster.pick.opti_mcc.unique_list.0.03.rep.tre")

seq_tree_ids <- data_frame(seq_id = mothur_tree$tip.label) %>% 
      left_join(seq_id_df)

mothur_tree$tip.label <- seq_tree_ids$feature_id

mothur_metrics <- feature_metrics %>%
      filter(feature_id %in% mothur_tree$tip.label) %>%
      filter(biosample_id == "E01JH0016")

tips_to_drop <- mothur_tree$tip.label[!(mothur_tree$tip.label %in% mothur_metrics$feature_id)]

mothur_trim <- drop.tip(mothur_tree,tip = tips_to_drop)

mothur_labels <-  data_frame(tip.label = mothur_trim$tip.label)
mothur_dat <- list()
mothur_dat$slope_metric <- mothur_metrics$slope_metric
mothur_dat$adj.r.squared <- mothur_metrics$adj.r.squared
mothur_dat$tip.label <- mothur_metrics$feature_id
mothur_dat <- as.data.frame(mothur_dat) %>% right_join(mothur_labels) %>%
      column_to_rownames(var = "tip.label")
mothur_p4d <- phylo4d(mothur_trim, mothur_dat)
mothur_lipa <- lipaMoran(mothur_p4d)
barplot.phylo4d(mothur_p4d, bar.col=(mothur_lipa$p.value < 0.05) + 1, center = FALSE , scale = FALSE)
```

```{r}
phyloSignal(p4d = mothur_p4d, method = c("I","Cmean"))
``` 

```{r message = FALSE}
mothur_labels <-  data_frame(id = mothur_trim$tip.label)
mothur_dat <- mothur_metrics %>% 
      rename(id = feature_id) %>% 
      right_join(mothur_labels) %>% 
      dplyr::select(id, slope_metric, adj.r.squared) 

mothur_tax <- logFC_pre_taxa %>% 
      filter(pipe == "mothur") %>% 
      rename(id = feature_id) %>% 
      left_join(mothur_labels)

mothur_trim <- groupOTU(mothur_trim, split(mothur_tax$id, mothur_tax$Rank3), group_name = "Class")



p <- ggtree(mothur_trim, aes(color = Class))

p <- facet_plot(p, panel = 'Slope Metric', data = mothur_dat, 
                geom=geom_segment, aes(x=0, xend=slope_metric, y=y, yend=y), size=1)

facet_plot(p, panel = 'R2', data = mothur_dat, 
           geom=geom_segment, aes(x=0, xend=adj.r.squared, y=y, yend=y), size=1) +
      theme_tree2() + theme(legend.position = "bottom")
```



## Unclustered 
```{r message = FALSE}
unclustered_tree <- read.tree("~/Projects/mgtst_pipelines/unclustered/unclustered_seqs_set.tre")
unclustered_metrics <- feature_metrics %>%
      filter(pipe == "unclustered", biosample_id == "E01JH0016")

## Only including tips in logFC feature summary (pre)
tips_to_drop <- unclustered_tree$tip.label[!(unclustered_tree$tip.label %in% unclustered_metrics$feature_id)]
unclustered_trim <- drop.tip(unclustered_tree,tip = tips_to_drop)

## Data frame with metrics for comparison
unclustered_labels <-  data_frame(tip.label = unclustered_trim$tip.label)
unclustered_dat <- unclustered_metrics %>% rename(tip.label = feature_id) %>% 
      right_join(unclustered_labels) %>% 
      dplyr::select(tip.label, slope_metric, adj.r.squared) %>% 
      as.data.frame() %>% dplyr::select(-tip.label)

unclustered_p4d <- phylo4d(unclustered_trim, unclustered_dat)
unclustered_lipa <- lipaMoran(unclustered_p4d)
barplot.phylo4d(unclustered_p4d, bar.col=(unclustered_lipa$p.value < 0.05) + 1, center = FALSE , scale = FALSE)
```

```{r}
phyloSignal(p4d = unclustered_p4d, method = c("I","Cmean"))
``` 


```{r message = FALSE}
unclustered_labels <-  data_frame(id = unclustered_trim$tip.label)
unclustered_dat <- unclustered_metrics %>% 
      rename(id = feature_id) %>% 
      right_join(unclustered_labels) %>% 
      dplyr::select(id, slope_metric, adj.r.squared) 

unclustered_tax <- logFC_pre_taxa %>% 
      filter(pipe == "unclustered") %>% 
      rename(id = feature_id) %>% 
      left_join(unclustered_labels)

unclustered_trim <- groupOTU(unclustered_trim, split(unclustered_tax$id, unclustered_tax$Rank3), group_name = "Class")

p <- ggtree(unclustered_trim)

p <- facet_plot(p, panel = 'Slope Metric', data = unclustered_dat, 
                geom=geom_segment, aes(x=0, xend=slope_metric, y=y, yend=y, color = Class), size=1)

facet_plot(p, panel = 'R2', data = unclustered_dat, 
           geom=geom_segment, aes(x=0, xend=adj.r.squared, y=y, yend=y, color = Class), size=1) +
      theme_tree2() + theme(legend.position = "bottom")
```



* Mean-variance plot - highlight features identified as outliers