---
title: "Figures for M3 Slides"
author: "Nate Olson"
date: "1/5/2018"
output: pdf_document
---

## E coli 
```{r include = FALSE}
library(tidyverse)
library(ggpubr)
```


These features either represent different copies of the 16S gene, different Escherichia species, or sequencing errors generated by the high abundance E. coli. 

```{r eschData, include = FALSE}
## logFC 
logFC_edgeR <- readRDS("data/logFC_edgeR_df.RDS") 
logFC_edgeR <- logFC_edgeR %>% rename(feature_id = OTUname)
logFC_esch <- logFC_edgeR %>% 
      filter(Rank5 == "f__Enterobacteriaceae" | Rank6 == "Escherichia/Shigella")

## Escherichia Features (QIIME Enterobacteriaceae)
esch_features <- logFC_esch %>% 
      filter(T1 == 0, T2 == 20) %>% 
      group_by(pipe, biosample_id) %>% 
      top_n(n = 3, logCPM) %>% 
      ungroup() %>% 
      select(pipe, feature_id) %>% 
      unique()


## Raw Counts
raw_counts <- readRDS("data/raw_counts.RDS")
raw_counts_esch <- raw_counts %>% 
      right_join(esch_features) %>% 
      filter(biosample_id != "NTC") %>% 
      group_by(biosample_id, feature_id) %>% 
      mutate(med_count = median(count)) %>% 
      filter(med_count != 0)

## Enterobacteriaceae Features Updated
esch_features <- raw_counts_esch %>% ungroup() %>%
      select(pipe, feature_id) %>% unique()

## NB Counts  
nb_counts <- readRDS("data/nb_counts_titrations.RDS")
nb_counts_esch <- nb_counts %>% 
      right_join(esch_features) %>% 
      filter(biosample_id != "NTC")

## Count Data Frame
esch_counts <- left_join(raw_counts_esch, nb_counts_esch)

## updated logFC
logFC_esch <- left_join(esch_features, logFC_esch)

### Relative Abundance Plot
scaleFUN <- function(x) sprintf("%.3f", x)
nb_plot <- nb_counts_esch %>% ungroup() %>% 
      mutate(pipe = factor(pipe, levels = c("unclustered","dada2","mothur","qiime"))) %>% 
      mutate(t_fctr = as.numeric(as.character(t_fctr))) %>% 
      ggplot() + 
      geom_line(aes(x = t_fctr, y = nb_prop, color = feature_id)) + 
      geom_point(aes(x = t_fctr, y = nb_prop, fill = feature_id), shape = 21) + 
      facet_grid(biosample_id ~ pipe, scales = "free") + theme_bw() +
      scale_y_continuous(trans = "log2", labels = scaleFUN, name = "Relative Abundance") +
      labs(x = "Titration")

### LogFC plot
logFC_plot <- logFC_esch %>% filter(T1 == 1, T2 != 20) %>% 
      ungroup() %>% 
      mutate(pipe = factor(pipe, levels = c("unclustered","dada2","mothur","qiime"))) %>% 
      mutate(T2 = as.numeric(as.character(T2))) %>% 
      ggplot() + 
      geom_line(aes(x = T2, y = logFC, color = feature_id)) + 
      geom_point(aes(x = T2, y = logFC, fill = feature_id), shape = 21) + 
      facet_grid(biosample_id~pipe) + 
      theme_bw() + labs(x = "Titration")
```

## Figures only for E01JH0011 

```{r}
ec_11_counts <- esch_counts %>% filter(biosample_id == "E01JH0011")
```

```{r}
ec_11_counts %>% filter(pipe == "dada2") %>% 
      mutate(pipe = factor(pipe, levels = c("unclustered","dada2","mothur","qiime"))) %>% 
      mutate(t_fctr = as.numeric(as.character(t_fctr))) %>% 
      ggplot() +
      geom_line(aes(x = t_fctr, y = count, color = feature_id)) +
      geom_point(aes(x = t_fctr, y = count, fill = feature_id), shape = 21) +
      # facet_grid(. ~ pipe, scales = "free") + 
      theme_bw() +
      scale_y_continuous(trans = "log2", labels = scaleFUN, name = "Relative Abundance") +
      labs(x = "Titration") + theme(legend.position = "none")
```

```{r}
ec_11_counts %>% filter(pipe == "dada2") %>% 
      mutate(pipe = factor(pipe, levels = c("unclustered","dada2","mothur","qiime"))) %>% 
      mutate(t_fctr = as.numeric(as.character(t_fctr))) %>% 
      ggplot() +
      geom_line(aes(x = t_fctr, y = nb_prop, color = feature_id)) +
      geom_point(aes(x = t_fctr, y = nb_prop, fill = feature_id), shape = 21) +
      # facet_grid(. ~ pipe, scales = "free") + 
      theme_bw() +
      scale_y_log10() +
      labs(x = "Titration", y = "Relative Abundance") + theme(legend.position = "none")
ggsave("~/Desktop/ecoli_rel_abu.png", height = 3, width = 3)
```

```{r}
ec_11_prepost <- ec_11_counts %>% filter(t_fctr %in% c(0,20)) %>% 
      select(pipe, biosample_id, feature_id, t_fctr, nb_prop) %>% 
      distinct() %>% 
      mutate(t_fctr = if_else(t_fctr == 0, "post","pre")) %>% 
      spread(t_fctr, nb_prop)
ec_11_exp_counts <- ec_11_counts %>% 
      filter(t_fctr %in% c(1:15)) %>% 
      select(biosample_id, pipe, feature_id, t_fctr, nb_prop) %>% 
      distinct() %>% 
      left_join(ec_11_prepost) %>% 
      mutate(t_fctr = as.numeric(as.character(t_fctr)),
            exp_prop = post * 2^-t_fctr + pre * (1 - 2^-t_fctr))
```
```{r}
ec_11_exp_counts %>% 
      filter(pipe == "dada2") %>% 
      ggplot() + 
      geom_point(aes(x = exp_prop, y = nb_prop, fill = feature_id), shape = 21) + 
      geom_abline(aes(intercept = 0, slope = 1), linetype = 2, color = "grey60") + 
      # facet_wrap(~feature_id, nrow = 1) + 
      theme_bw() + 
      scale_x_log10() + scale_y_log10() + 
      labs(x = "Expected Relative Abundance", y = "Relative Abundance") + 
      theme(legend.position = "none") 
ggsave("~/Desktop/ecoli_obs_v_exp.png", height = 3, width = 3)
```

```{r}
ec_11_feat_metrics <- ec_11_exp_counts %>% 
      filter(pipe == "dada2", feature_id == "SV1") %>% 
      mutate(error_rate = abs(exp_prop - nb_prop)/exp_prop) %>% 
      group_by(feature_id) %>% 
      summarise(med_error = median(error_rate),
                rcov_error = IQR(error_rate)/med_error)

ec_11_exp_counts %>% filter(pipe == "dada2", feature_id == "SV1") %>% 
      mutate(error_rate = abs(exp_prop - nb_prop)/exp_prop) %>% 
      mutate(pipe = factor(pipe, levels = c("unclustered","dada2","mothur","qiime"))) %>% 
      mutate(t_fctr = as.numeric(as.character(t_fctr))) %>% 
      ggplot() +
      geom_linerange(aes(x = t_fctr, ymin = nb_prop, ymax = exp_prop)) + 
      geom_line(aes(x = t_fctr, y = nb_prop, color = feature_id)) + 
      geom_point(aes(x = t_fctr, y = nb_prop, fill = feature_id), shape = 21) +
      
      geom_text(aes(x = t_fctr, y = nb_prop * 1.35, label = signif(error_rate, digits = 2))) + 
      geom_text(data = ec_11_feat_metrics, 
                aes(x = 12, y = 0.75, 
                    label = paste0("Median Error: ", signif(med_error, digits = 2)))) +
            geom_text(data = ec_11_feat_metrics, 
                aes(x = 12, y = 0.5, 
                    label = paste0("RCOV Error: ", signif(rcov_error, digits = 2)))) +
      # facet_wrap(~ feature_id) + 
      theme_bw() +
      scale_y_continuous(trans = "log2", labels = scaleFUN, name = "Relative Abundance") +
      labs(x = "Titration") + theme(legend.position = "none")
ggsave("~/Projects/MGTST-Manuscript/ecoli_metrics.pdf", height = 3, width = 5)
```
## Pipeline comparisons 
```{r relAbuSetup, include=FALSE}
## TODO feature count table
library(nlme)
library(ape)
library(multcomp)
library(kableExtra)
library(tidyverse)
library(forcats)
library(stringr)
library(ggpubr)
library(ggridges)
nb_counts <- readRDS("data/nb_counts_titrations.RDS")
pa_summary_anno_df <- readRDS("data/pa_summary_anno_df.RDS")
theta_est <- readRDS("data/bootstrap_theta_estimates.rds")
```

```{r relAbuMunge, echo = FALSE, message = FALSE, warning = FALSE}
### TODO - move to separate Rmd and generate data_frame
pre_post_prop <- nb_counts %>% 
    ungroup() %>% 
    filter(t_fctr %in% c(0,20)) %>% 
    mutate(end_point = if_else(t_fctr == 0 , "post", "pre")) %>% 
    dplyr::select(-t_fctr) %>% 
    ## setting values to 0 when one or more of the PCR replicates are 0 for titration end-points
    spread(end_point,nb_prop, fill = 0)

prop_inferred <- theta_est %>% 
    filter(pipe == "unclustered") %>% 
    ungroup() %>%
    mutate(t_fctr = factor(t_fctr, levels = c(0:5, 10, 15, 20))) %>% 
    dplyr::select(biosample_id, theta_hat_mean, t_fctr) %>% 
    right_join(nb_counts) %>% right_join(pre_post_prop) %>% 
    filter(t_fctr %in% c(1:5,10,15)) %>% 
    ## Using inferred theta estimates to calculate expected values
    mutate(inferred_prop = post * theta_hat_mean + pre * (1 - theta_hat_mean))

## Excluding mix and unmix specific features
## Only including features observed in all or none of the four pre- post- PCR replicates
## Features with relative abundance estimates and expected values less than 1e-5, these are features that we would not expect to consistently observe in a PCR replicate for the given sequencing depth, ~100k 
## Excluding titrations where the inferred theta values are less than 1
pa_filter <- pa_summary_anno_df %>% 
    filter(pa_specific == "unspecific") %>% 
    dplyr::select(biosample_id, pipe, feature_id, full_pre, T00, T20, pa_mixed) %>% 
    filter(T00 %in% c(0,4), T20 %in% c(0,4))

prop_inferred <- prop_inferred %>% 
    right_join(pa_filter) %>% 
    # filter(nb_prop > 1e-5, 
    #        inferred_prop > 1e-5,
    #        theta_hat_mean > 0)
    ## Filtering absed on 1/median library size
    filter(nb_prop > 1/73571,
           inferred_prop > 1/73571,
           theta_hat_mean > 0)

## Save for supplemental table
saveRDS(prop_inferred, "data/prop_inferred.RDS")


#### Error Rate Calculations
rel_abu_error <- prop_inferred %>% 
    mutate(t_fctr = factor(t_fctr, levels = c(1:5, 10, 15))) %>% 
    mutate(inferred_error = abs(nb_prop - inferred_prop),
           inferred_error_rate = inferred_error/inferred_prop) 

rel_abu_error_summary <-  rel_abu_error %>% 
    group_by(pipe, biosample_id, feature_id) %>% 
    summarise(median_rel_abu = median(nb_prop),
              median_error = median(inferred_error_rate),
              iqr_error = IQR(inferred_error_rate),
              rcov_error = iqr_error/median_error, 
              mean_error = mean(inferred_error_rate),
              var_error = var(inferred_error_rate),
              cov_error = var_error/mean_error) 
```


```{r}
prop_inferred %>% filter(biosample_id == "E01JH0004", pipe != "unclustered") %>% 
      mutate(pipe = factor(pipe, levels = c("unclustered","dada2","mothur","qiime"))) %>% 
    ggplot() +
      geom_hex(aes(x = inferred_prop, y = nb_prop)) + 
    geom_smooth(aes(x = inferred_prop, y = nb_prop), color = "darkorange", method = "lm") +
    # geom_abline(aes(intercept = 0, slope = 1), color = "grey80", linetype = 2) +
    facet_wrap(~pipe, nrow = 1) +
    scale_y_log10() + scale_x_log10() +
    theme_bw() +
    labs(x = "Expected",
         y = "Observed", 
         color = "Pipeline") + 
      theme(legend.position = "right")
ggsave("~/Desktop/obs_v_exp_plot.png", height = 4, width = 8)
```

```{r}
error_boxplot <- rel_abu_error %>% group_by(pipe, biosample_id, feature_id) %>% 
      filter(biosample_id == "E01JH0004", pipe != "unclustered") %>% 
    summarise(median_error = median(inferred_error_rate)) %>%
            ungroup() %>% 
      mutate(pipe = factor(pipe, levels = c("unclustered","dada2","mothur","qiime"))) %>% 
    ggplot() + 
    geom_boxplot(aes(x = pipe, y = median_error, color = pipe), outlier.shape = NA) + 
    # facet_wrap(~biosample_id, ncol = 1) + 
    theme_bw() + theme(legend.position = "none") +
    labs(x = "Pipeline", y = "Median Error Rate", color = "Pipeline")

ymin <- ggplot_build(error_boxplot)$data[[1]]$ymin %>% min()
ymax <- ggplot_build(error_boxplot)$data[[1]]$ymax %>% max()
error_boxplot + coord_cartesian(ylim = c(ymin, ymax))
ggsave("~/Desktop/med_error_plot.png", height = 4, width = 5)
```

```{r}
rcov_boxplot <- rel_abu_error_summary %>% filter(biosample_id == "E01JH0004") %>% 
      ungroup() %>% 
      mutate(pipe = factor(pipe, levels = c("unclustered","dada2","mothur","qiime"))) %>% 
      filter(pipe != "unclustered") %>% 
    ggplot() + 
    geom_boxplot(aes(x = pipe, y = rcov_error, color = pipe), outlier.shape = NA) + 
    # facet_wrap(~biosample_id, nrow = 1) + 
    theme_bw() + theme(legend.position = "none") +
    labs(x = "Pipeline", y = "RCOV Error Rate", color = "Pipeline") 

ymin <- ggplot_build(rcov_boxplot)$data[[1]]$ymin %>% min()
ymax <- ggplot_build(rcov_boxplot)$data[[1]]$ymax %>% max()
rcov_boxplot + coord_cartesian(ylim = c(ymin, ymax))
ggsave("~/Desktop/rcov_error_plot.png", height = 4, width = 5)
```
