---
title: "NIST Run 1 Seq QA"
author: "Nate Olson"
date: '`r Sys.Date()`'
always_allow_html: yes
output:
  pdf_document: default
  html_document: default
---

```{r setup, warning=FALSE, message=FALSE, echo = FALSE}
library(ProjectTemplate)
cwd <- getwd()
setwd("../")
load.project()
setwd(cwd)
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
```


```{r sav_data_files}
project_dir <- "~/Projects/16S_etec_mix_study/"
sav_dir <- paste0(project_dir,"data/basespace_sav/")
sav_list <- list(
    march2016_v3 = paste0(sav_dir,"run_18071097_sav"),
    basespace_v3 = paste0(sav_dir,"run_3861867_sav"),
    april2016_v3 = paste0(sav_dir,"run_18527531_sav"),
    nist2017_v3 = paste0(sav_dir, "run_170209_sav")
)
```


```{r fq_data_files}
fastq_dir <- file.path(project_dir, "data/raw_seq/nist_run1/")
dat_files <- list.files(path = fastq_dir,pattern = "001.fastq.gz$",
                           recursive = TRUE, full.names = TRUE)
seq_ds_id <- dat_files %>% str_split("/") %>% flatten_chr() %>% 
    grep(pattern = "fastq.gz", .,value = TRUE) %>%
    str_replace(".fastq.gz","") %>% paste0("Fq_",.) %>% 
    str_replace("-",".")
names(dat_files) <- seq_ds_id
```

```{r munge_error_dat}
error_df <- sav_list %>% map(savR) %>%
    map_df(errorMetrics,.id = "ds") %>%
    separate(ds, c("ds","chemistry"),sep = "_") %>%
    mutate(read_len = if_else(chemistry == "v3",301, 251),
           read = if_else(cycle < read_len, "R1","R2"))
```

```{r runQA, eval=FALSE}
# Generated R script `run_rqca.R` to generate and save a list with the rqcA output. 
qa_list <- list()
step_size <- 6 #for all data

n_files <- length(dat_files)
for(i in 0:((n_files/step_size) - 1)){
    print(i)
    qa_list <- c(qa_list,
                      rqcQA(dat_files[(step_size * i+1):(step_size*(i + 1))],
                          group = read_groups[(step_size *i+1):(step_size *(i+1))],
                          workers = step_size))
}

names(qa_list) <- names(dat_files)
names(read_groups) <- names(qa_list)
```

```{r loadQA}
# from script but not saved
read_groups <- rep(NA,length(dat_files))
read_groups[grepl("/NIST-1-.*_R1", dat_files)] <- "plate1_R1"
read_groups[grepl("/NIST-1-.*_R2", dat_files)] <- "plate1_R2"
read_groups[grepl("/NIST-2-.*_R1", dat_files)] <- "plate2_R1"
read_groups[grepl("/NIST-2-.*_R2", dat_files)] <- "plate2_R2"

qa_list <- readRDS("~/Projects/16S_etec_mix_study/analysis/nist_run1_qa/rqcQA_list.rds")
```

```{r metadat}
# Tidy data

## Run 
grp_df <- data_frame(read_group = read_groups, 
                     seq_ds_id, 
                     filename = basename(dat_files)) %>%
    separate(read_group,c("plate","Read")) %>% 
    mutate(ill_id = str_replace(filename, "_.*",""),
           ill_id = str_replace(ill_id, "NIST-",""))

## read count data 
qa_file_info <- perFileInformation(qa_list) %>% 
    select(-format,-path)

## study metadata
meta_df <- sampleSheet %>% 
    mutate(pos_ns = str_replace(pos, "_",""),
           ill_id = paste(pcr_16S_plate, pos_ns, sep = "-")) %>% 
    filter(seq_lab == "NIST", barcode_lab == "NIST") %>% 
    mutate(pcr_16S_plate = as.character(pcr_16S_plate)) %>% 
    left_join(grp_df) %>% left_join(qa_file_info)
```

```{r read_metrics}
# __Read Level Metrics__
qa_read_df <- qa_list %>% map_df(perReadWidth) %>% left_join(meta_df) %>% 
    mutate(len_prop = count/reads)

qa_read_q_df <- qa_list %>% map_df(perReadQuality) %>% left_join(meta_df)

qa_read_freq <- qa_list %>% map_df(perReadFrequency) %>% left_join(meta_df)
```

```{r cycle_metrics}
## amplicon position __Cycle Level Metrics__
amp_pos_df <- data_frame(cycle = rep(1:300, 2), 
                         Read = rep(c("R1","R2"), each = 300),
                         amp_pos = c(1:300,c(460 - 1:300)))

qa_cycle_q_df <- qa_list %>% map_df(perCycleQuality) %>% 
    as_data_frame() %>% 
    mutate(cycle = as.numeric(as.character(cycle))) %>% 
    filter(count != 0) %>% # not sure if this impacts the smoothing function ...
    left_join(meta_df) %>% left_join(amp_pos_df)
```

### Sequencing Data Quality Assessment
To generate summaries of QA metrics for the 384 datasets in the study (192 samples with forward and reverse reads) used the bioconductor `Rqc` package (REF) to calculate the quality metrics used in the following analysis. 

#### Read Counts   
Two barcoded experimental sample has less than 50,000 reads. The rest of the samples with less than 50,000 reads are negative PCR controls (NTC). Sample E01JH0016 titration 5 position F9 of plate 1 initial 16S PCR failed. 
__TODO__ Figure out why E01JH0011 titration 3 position D2 plate 2 is also low, look at picogreen post normaization data. 


```{r table_count}
# meta_df %>% mutate(exp_ntc = ifelse(biosample_id == "NTC", "NTC","EXP")) %>% 
#     group_by(exp_ntc, Read, plate) %>% 
#     summarise(mean_lib_size = mean(reads),
#               min_lib_size = min(reads),
#               median = median(reads),
#               max_lib_size = max(reads)) %>% 
#     kable(caption = "Summary statistics for experimental and no template control samples by PCR plate and read.")
```

Potential issue with high seq number for no template controls. 

```{r fig_count, fig.cap="Number of reads per barcoded sample (Library Size), by read direction (X-facet) and replicate 16S PCR plate (Y-facet). Vertical line indicates 50,000 reads per barcoded sample."}
meta_df %>% # filter(Read == "R1") %>%
    ggplot() + 
        geom_histogram(aes(x = reads), position = "dodge") + 
        facet_grid(plate~Read) + scale_x_log10() + 
        geom_rug(aes(x = reads, color = biosample_id)) +
        geom_vline(aes(xintercept = 50000), color = "grey80") +
        geom_vline(aes(xintercept = 10000), linetype = 2, color = "grey80") +
        theme_bw() + labs(x = "Library Size (log10)", 
                          y = "Count", color = "Sample ID") +
        theme(legend.position = "bottom")
```

```{r}
# ggsave("~/Desktop/library_size.png", width = 5.46, height = 7.48, dpi = 450)
```

```{r count_tbl}
# meta_df <- meta_df %>% mutate(sampleID = biosample_id, dilution = titration)
# meta_df  %>% filter(sampleID != "NTC") %>%
#     select(sampleID, dilution, pos, plate, Read, reads) %>%
#     spread(Read, reads) %>%
#     kable(caption = "Barcoded experimental samples")
``` 

#### PhiX Error Rate 
The sequencing error rate data was obtained from the Basespace sequencing run report downloaded from Basespace (SAV file).
Error rate is compared to the first sequencing run and a 16S public dataset on basespace ( 16S-Metagenomic-Library-Prep run id 3861867). 
The error rate for the second run was lower for both R1 and R2 compared to the first run but still higher than the error rate for the public dataset. __NOTE__ Not sure whether or not to include in publication as supplemental material.

```{r phix_error_fig, fig.cap = "PhiX error rate for initial and reseuqencing of JHU barcoded samples compared to the public dataset."}
ggplot(error_df) +
    geom_point(aes(x = cycle, y = errorrate, color = ds), alpha = 0.01) +
    geom_smooth(aes(x = cycle, y = errorrate, color = ds)) +
    theme_bw() +
    labs(x = "Cycle", y = "PhiX Error Rate", color = "Dataset") +
    facet_grid(chemistry~read, scales = "free_x")
```

```{r}
# ggsave("nist_run_error_rate.png")
```

#### Base Quality Score 
##### Read BQ
Differences in forward and reverse read average base quality score distributions consistent between replicate plates. 
A distinct population of barcoded datasets, NTC vs experimental samples, with a higher proportion of lower base quality scores for forward read datasets. 
For reverse reads the popultion of datasets with lower base quality scores is more hetergeneous.  

```{r fig_readq, fig.cap = "Distribution of base quality scores per barcoded samples."}
qa_read_q_df %>% filter(!is.na(plate)) %>% 
    ggplot() + geom_density(aes(x = average, fill = biosample_id,
                                group = filename), alpha = 0.25) + 
    facet_grid(plate~Read) + theme_bw() + 
    labs(x = "Read Average Base Quality Score", y = "Density", fill = "Sample ID") + 
    theme(legend.position = "bottom")
```

#### Cycle BQ
Cycle base quality score is more homogeneous from PCR plate 2 samples than plate 1. For the expected overlap region, based on primer positions and read lengths (16S PCR fig), the forward read has consistently higher base quality scores relative to the reverse read. 
__NOTE__ Figure out which sample is the low reverse read for plate 1 sample id 17.  

```{r fig_lowcycleq, fig.cap = "Smoothing spline of the base quality score by sequencing cycle. Vertical lines indicate approximate overlap region between forward and reverse reads. This is not a read level analysis but average quality score for individual barcoded datasets."}
qa_cycle_q_df %>%# filter(biosample_id != "NTC") %>%
  ggplot(aes(x = amp_pos, y = score)) + 
    geom_vline(aes(xintercept = 300), color = "grey60") +
    geom_vline(aes(xintercept = 160), color = "grey60") +
    geom_smooth(aes(weight = count, color = biosample_id, group = filename)) + 
    geom_vline(aes(xintercept = 280), color = "grey40", linetype = 2) +
    facet_grid(plate~.) + theme_bw() + 
    scale_x_continuous(breaks = c(0,150,300, 450)) +
    labs(x = "Amplicon Position", 
         y = "Base Quality Score", color = "Sample ID") + 
    theme(legend.position = "bottom")
```

```{r}
# ggsave("~/Desktop/read_base_quality.png", width = 5.46, height = 7.48, dpi = 450)
```

