---
title: "Loading Mix Study Pipeline Results"
output:
  html_document: default
  html_notebook: default
---

## Data description
Sequencing JHU Run 2 from the 16S two sample titration study was processed using QIIME, Mothur, and DADA2. QIIME was run with four different pipelines, including reference based open clustering or _de novo_ clustering, with and without chimera filtering.


```{r setup, warning=FALSE, message=FALSE, echo = FALSE}
library(mgtst)
library(readr)
library(tidyr)
library(stringr)
library(magrittr)
library(knitr)
library(sapkotaUtils)
library(dplyr)
library(tibble)
library(purrr)
library(metagenomeSeq)
library(phyloseq)
library(biomformat)
opts_chunk$set(cache=TRUE)
opts_chunk$set(fig.align="center")
```

File locations
```{r prj_paths, warning=FALSE, message=FALSE, echo = FALSE}
proj_dir <- "~/Projects/16S_etec_mix_study"
pipe_dir <- file.path(proj_dir, "analysis","pipelines")
qiime_dir <- file.path(pipe_dir, "qiime")
mothur_dir <- paste0(pipe_dir, "/mothur/")
```

### Sample Data 
```{r}
data(sample_sheet)
sample_sheet_ns <- sample_sheet %>% 
    mutate(pos_ns = str_replace(pos, "_",""))
```


### Functions for loading QIIME Data
#### Load as phyloseq
```{r}
load_phyloseq_mgtst_qiime_biom <- function(biom_file, tree_file, refseq_file, sample_sheet_ns){
      ps_qiime <- import_biom(BIOMfilename = biom_file,
                              treefilename = tree_file,
                              refseqfilename = refseq_file)
      sam_names <- sample_names(ps_qiime) %>% str_replace("centroid=","")
      sample_id <- data_frame(seq_ids = sam_names) %>%
            separate(col = seq_ids, 
                     into = c("pcr_16S_plate","pos_ns"), sep = "-") %>%
            mutate(pcr_16S_plate = as.numeric(pcr_16S_plate))

      sample_df <- left_join(sample_id, sample_sheet_ns) %>%
            filter(seq_lab == "JHU", barcode_lab == "JHU") %>%
            mutate(pcr_16S_plate = as.character(pcr_16S_plate)) %>% 
            as.data.frame()
      samdat <- sample_data(sample_df)
      sample_names(samdat) <- sample_names(ps_qiime)
      sample_data(ps_qiime) <- samdat
      
      ps_qiime
}
```

#### Load as MRexperiment
```{r}
load_mrexp_mgtst_qiime_biom <- function(biom_file, sample_df){
      mrexp_qiime <- load_biom(biom_file)
      sam_names <- colnames(mrexp_qiime@assayData$counts) %>% str_replace("centroid=","")
      colnames(mrexp_qiime@assayData$counts) <- sam_names

      sample_id <- data_frame(seq_ids = sam_names) %>%
            separate(col = seq_ids, 
                     into = c("pcr_16S_plate","pos_ns"), sep = "-") %>%
            mutate(pcr_16S_plate = as.numeric(pcr_16S_plate))

      sample_df <- left_join(sample_id, sample_sheet_ns) %>%
            filter(seq_lab == "JHU", barcode_lab == "JHU") %>%
            mutate(pcr_16S_plate = as.character(pcr_16S_plate)) %>% 
            as.data.frame()

      rownames(sample_df) <- sam_names
      mrexp_qiime@phenoData <-  sample_df %>% AnnotatedDataFrame()
      
      mrexp_qiime
}
```

## QIIME Reference Clustering No Chimera Filtering
```{r}
ds_dir <- file.path(qiime_dir,"otus_uc_fast")
biom_file <- file.path(ds_dir, "otu_table_mc2_w_tax_no_pynast_failures.biom") 
tree_file <- file.path(ds_dir,"rep_set.tre")
refseq_file <- file.path(ds_dir,"new_refseqs.fna.gz")
```

```{r}
ps_qiime_refclus_nochimera <- load_phyloseq_mgtst_qiime_biom(biom_file,tree_file, refseq_file, sample_sheet_ns)
saveRDS(ps_qiime_refclus_nochimera, file = "../data/ps_qiime_refclus_nochimera.RDS")
```

```{r}
mrexp_qiime_refclus_nochimera <- load_mrexp_mgtst_qiime_biom(biom_file, sample_sheet_ns)
saveRDS(mrexp_qiime_refclus_nochimera, file = "../data/mrexp_qiime_refclus_nochimera.RDS")
```

## QIIME Reference Clustering Chimera Filtering
Command used to run pipeline for open reference clustering with chimera filtered input sequences

```
pick_open_reference_otus.py -o otus_uc_fast_no_chimera -i split_libs_chimera/seqs.fna_consensus_fixed.fasta -p uc_fast_params.txt
```
Unsure of the error only using `otu_table_mc2_w_tax.biom`

```{r}
ds_dir <- file.path(qiime_dir,"otus_uc_fast_no_chimera")
biom_file <- file.path(ds_dir, "otu_table_mc2_w_tax_no_pynast_failures_json.biom") 
tree_file <- file.path(ds_dir,"rep_set.tre")
refseq_file <- file.path(ds_dir,"new_refseqs.fna")
```

```{r}
ps_qiime_refclus_chimera_filt <- load_phyloseq_mgtst_qiime_biom(biom_file,tree_file, refseq_file, sample_sheet_ns)
saveRDS(ps_qiime_refclus_chimera_filt, file = "../data/ps_qiime_refclus_chimera_filt.RDS")
```


```{r}
mrexp_qiime_refclus_chimera_filt <- load_mrexp_mgtst_qiime_biom(biom_file, sample_sheet_ns)
saveRDS(mrexp_qiime_refclus_chimera_filt, file = "../data/mrexp_qiime_refclus_chimera_filt.RDS")
```

## QIIME De Novo Clustering No Chimera Filtering
```{r}
ds_dir <- file.path(qiime_dir,"otus_uc_fast_denovo")
biom_file <- file.path(ds_dir, "otu_table.biom") 
tree_file <- file.path(ds_dir,"rep_set.tre")
refseq_file <- NULL # ref seq not produced with de novo qiime pipeline
```

```{r}
ps_qiime_denovo_nochimera <- load_phyloseq_mgtst_qiime_biom(biom_file, tree_file, 
                                                            refseq_file, sample_sheet_ns)
saveRDS(ps_qiime_denovo_nochimera, file = "../data/ps_qiime_denovo_nochimera.RDS")
```


```{r}
mrexp_qiime_denovo_nochimera <- load_mrexp_mgtst_qiime_biom(biom_file, sample_sheet_ns)
saveRDS(mrexp_qiime_denovo_nochimera, file = "../data/mrexp_qiime_denovo_nochimera.RDS")
```


## QIIME De Novo Clustering Chimera Filtering
```{r}
ds_dir <- file.path(qiime_dir,"otus_uc_fast_denovo_no_chimera")
biom_file <- file.path(ds_dir, "otu_table.biom") 
tree_file <- file.path(ds_dir,"rep_set.tre")
refseq_file <- NULL # ref seq not produced with de novo qiime pipeline
```

```{r}
ps_qiime_denovo_chimera_filt <- load_phyloseq_mgtst_qiime_biom(biom_file, tree_file, 
                                                            refseq_file, sample_sheet_ns)
saveRDS(ps_qiime_denovo_chimera_filt, file = "../data/ps_qiime_denovo_chimera_filt.RDS")
```


```{r}
mrexp_qiime_denovo_chimera_filt <- load_mrexp_mgtst_qiime_biom(biom_file, sample_sheet_ns)
saveRDS(mrexp_qiime_denovo_chimera_filt, file = "../data/mrexp_qiime_denovo_chimera_filt.RDS")
```


## Mothur
### MRexperiment
```{r}
out_dir <- file.path(mothur_dir, "data","process")
biom_file <- file.path(out_dir,"mgtst.trim.contigs.good.unique.good.filter.unique.precluster.pick.an.unique_list.0.03.biom")
mrexp_mothur <- load_biom(biom_file)

## Metdata
sam_names <- colnames(mrexp_mothur@assayData$counts)
sample_id <- data_frame(id = sam_names)
sample_df <- left_join(sample_id, sample_sheet) %>%
      filter(seq_lab == "JHU", barcode_lab == "JHU", sampleID != "NTC") %>%
      mutate(pcr_16S_plate = as.character(pcr_16S_plate)) %>% 
      column_to_rownames(var = "id") %>% 
      as.data.frame()

mrexp_mothur@phenoData <-  sample_df %>% AnnotatedDataFrame()
saveRDS(mrexp_mothur, file = "../data/mrexp_mothur.RDS")
```


### Phyloseq
```{r}
ps_mothur <- import_biom(BIOMfilename = biom_file)
sample_id <- data_frame(id = sample_names(ps_mothur))
sample_df <- right_join(sample_id, sample_sheet) %>%
      filter(seq_lab == "JHU", barcode_lab == "JHU", sampleID != "NTC") %>% 
      mutate(pcr_16S_plate = as.character(pcr_16S_plate)) %>% 
      as.data.frame() %>% column_to_rownames(var = "id")
samdat <- sample_data(sample_df)
sample_data(ps_mothur) <- samdat
saveRDS(ps_mothur, file = "../data/ps_mothur.RDS")
```

