---
title: "Escherichia Features"
author: "Nate Olson"
date: '`r Sys.Date()`'
output:
  bookdown::pdf_document2:
    toc: FALSE
---

```{r include = FALSE}
library(tidyverse)
library(ggpubr)
```


These features either represent different copies of the 16S gene, different Escherichia species, or sequencing errors generated by the high abundance E. coli.

```{r eschData, include = FALSE}
## logFC
logFC_edgeR <- readRDS("../data/logFC_edgeR_df.RDS")
logFC_edgeR <- logFC_edgeR %>% rename(feature_id = OTUname)
logFC_esch <- logFC_edgeR %>%
      filter(Rank5 == "f__Enterobacteriaceae" | Rank6 == "Escherichia/Shigella")

## Escherichia Features (QIIME Enterobacteriaceae)
esch_features <- logFC_esch %>%
      filter(T1 == 0, T2 == 20) %>%
      group_by(pipe, biosample_id) %>%
      top_n(n = 3, logCPM) %>%
      ungroup() %>%
      select(pipe, feature_id) %>%
      unique()


## Raw Counts
raw_counts <- readRDS("../data/raw_counts.RDS")
raw_counts_esch <- raw_counts %>%
      right_join(esch_features) %>%
      filter(biosample_id != "NTC") %>%
      group_by(biosample_id, feature_id) %>%
      mutate(med_count = median(count)) %>%
      filter(med_count != 0)

## Enterobacteriaceae Features Updated
esch_features <- raw_counts_esch %>% ungroup() %>%
      select(pipe, feature_id) %>% unique()

## NB Counts
nb_counts <- readRDS("../data/nb_counts_titrations.RDS")
nb_counts_esch <- nb_counts %>%
      right_join(esch_features) %>%
      filter(biosample_id != "NTC")

## Count Data Frame
esch_counts <- left_join(raw_counts_esch, nb_counts_esch)

## updated logFC
logFC_esch <- left_join(esch_features, logFC_esch)

### Relative Abundance Plot
scaleFUN <- function(x) sprintf("%.3f", x)
nb_plot <- nb_counts_esch %>% ungroup() %>%
      mutate(pipe = factor(pipe, levels = c("unclustered","dada2","mothur","qiime"))) %>%
      mutate(t_fctr = as.numeric(as.character(t_fctr))) %>%
      ggplot() +
      geom_line(aes(x = t_fctr, y = nb_prop, color = feature_id)) +
      geom_point(aes(x = t_fctr, y = nb_prop, fill = feature_id), shape = 21) +
      facet_grid(biosample_id ~ pipe, scales = "free") + theme_bw() +
      scale_y_continuous(trans = "log2", labels = scaleFUN, name = "Relative Abundance") +
      labs(x = "Titration")

### LogFC plot
logFC_plot <- logFC_esch %>% filter(T1 == 1, T2 != 20) %>%
      ungroup() %>%
      mutate(pipe = factor(pipe, levels = c("unclustered","dada2","mothur","qiime"))) %>%
      mutate(T2 = as.numeric(as.character(T2))) %>%
      ggplot() +
      geom_line(aes(x = T2, y = logFC, color = feature_id)) +
      geom_point(aes(x = T2, y = logFC, fill = feature_id), shape = 21) +
      facet_grid(biosample_id~pipe) +
      theme_bw() + labs(x = "Titration")
```



```{r eschPlot, fig.cap = "Relative abundance (A) and log fold-change estimtates (B) for high abundance \\textit{Escherichia} features. Colors are used to differentiate features."}
ggarrange(nb_plot, logFC_plot,
          ncol = 1, nrow = 2, align = "hv",common.legend = TRUE,legend = "bottom",labels = "AUTO")
```

## Figures only for E01JH0011
```{r}
ec_11_counts <- esch_counts %>% filter(biosample_id == "E01JH0011")
```

```{r}
ec_11_counts %>% filter(pipe == "dada2") %>%
      mutate(pipe = factor(pipe, levels = c("unclustered","dada2","mothur","qiime"))) %>%
      mutate(t_fctr = as.numeric(as.character(t_fctr))) %>%
      ggplot() +
      geom_line(aes(x = t_fctr, y = count, color = feature_id)) +
      geom_point(aes(x = t_fctr, y = count, fill = feature_id), shape = 21) +
      # facet_grid(. ~ pipe, scales = "free") +
      theme_bw() +
      scale_y_continuous(trans = "log2", labels = scaleFUN, name = "Relative Abundance") +
      labs(x = "Titration") + theme(legend.position = "none")
```

```{r}
ec_11_counts %>% filter(pipe == "dada2") %>%
      mutate(pipe = factor(pipe, levels = c("unclustered","dada2","mothur","qiime"))) %>%
      mutate(t_fctr = as.numeric(as.character(t_fctr))) %>%
      ggplot() +
      geom_line(aes(x = t_fctr, y = nb_prop, color = feature_id)) +
      geom_point(aes(x = t_fctr, y = nb_prop, fill = feature_id), shape = 21) +
      # facet_grid(. ~ pipe, scales = "free") +
      theme_bw() +
      scale_y_log10() +
      labs(x = "Titration", y = "Relative Abundance") + theme(legend.position = "none")
ggsave("~/Desktop/ecoli_rel_abu.png", height = 3, width = 3)
```

```{r}
ec_11_prepost <- ec_11_counts %>% filter(t_fctr %in% c(0,20)) %>%
      select(pipe, biosample_id, feature_id, t_fctr, nb_prop) %>%
      distinct() %>%
      mutate(t_fctr = if_else(t_fctr == 0, "post","pre")) %>%
      spread(t_fctr, nb_prop)
ec_11_exp_counts <- ec_11_counts %>%
      filter(t_fctr %in% c(1:15)) %>%
      select(biosample_id, pipe, feature_id, t_fctr, nb_prop) %>%
      distinct() %>%
      left_join(ec_11_prepost) %>%
      mutate(t_fctr = as.numeric(as.character(t_fctr)),
            exp_prop = post * 2^-t_fctr + pre * (1 - 2^-t_fctr))
```
```{r}
ec_11_exp_counts %>%
      filter(pipe == "dada2") %>%
      ggplot() +
      geom_point(aes(x = exp_prop, y = nb_prop, fill = feature_id), shape = 21) +
      geom_abline(aes(intercept = 0, slope = 1), linetype = 2, color = "grey60") +
      # facet_wrap(~feature_id, nrow = 1) +
      theme_bw() +
      scale_x_log10() + scale_y_log10() +
      labs(x = "Expected Relative Abundance", y = "Relative Abundance") +
      theme(legend.position = "none")
ggsave("~/Desktop/ecoli_obs_v_exp.png", height = 3, width = 3)
```

```{r}
ec_11_feat_metrics <- ec_11_exp_counts %>%
      filter(pipe == "dada2", feature_id == "SV1") %>%
      mutate(error_rate = abs(exp_prop - nb_prop)/exp_prop) %>%
      group_by(feature_id) %>%
      summarise(med_error = median(error_rate),
                rcov_error = IQR(error_rate)/med_error)

ec_11_exp_counts %>% filter(pipe == "dada2", feature_id == "SV1") %>%
      mutate(error_rate = abs(exp_prop - nb_prop)/exp_prop) %>%
      mutate(pipe = factor(pipe, levels = c("unclustered","dada2","mothur","qiime"))) %>%
      mutate(t_fctr = as.numeric(as.character(t_fctr))) %>%
      ggplot() +
      geom_linerange(aes(x = t_fctr, ymin = nb_prop, ymax = exp_prop)) +
      geom_line(aes(x = t_fctr, y = nb_prop, color = feature_id)) +
      geom_point(aes(x = t_fctr, y = nb_prop, fill = feature_id), shape = 21) +

      geom_text(aes(x = t_fctr, y = nb_prop * 1.35, label = signif(error_rate, digits = 2))) +
      geom_text(data = ec_11_feat_metrics,
                aes(x = 12, y = 0.75,
                    label = paste0("Median Error: ", signif(med_error, digits = 2)))) +
            geom_text(data = ec_11_feat_metrics,
                aes(x = 12, y = 0.5,
                    label = paste0("RCOV Error: ", signif(rcov_error, digits = 2)))) +
      # facet_wrap(~ feature_id) +
      theme_bw() +
      scale_y_continuous(trans = "log2", labels = scaleFUN, name = "Relative Abundance") +
      labs(x = "Titration") + theme(legend.position = "none")
ggsave("~/Desktop/ecoli_metrics.png", height = 3, width = 5)
```
