---
title: "Introduction"
author: "Nate Olson"
date: '`r Sys.Date()`'
output:
  bookdown::pdf_document2:
    toc: FALSE
bibliography: [mgtst.bib]
---

Targeted sequencing of the 16S rRNA gene, commonly known as 16S rRNA marker-gene-surveys, is a commonly used method for characterizing microbial communities, microbiomes.
The 16S rRNA marker-gene-survey measurement process includes molecular (e.g. PCR and sequencing) and computational steps (e.g., sequence clustering) [@Goodrich2014].
Molecular steps are used to selectively target and sequence the 16S rRNA gene from prokaryotic organisms within a sample.
The computational steps convert the raw sequence data into a matrix with feature (e.g., operational taxonomic units) relative abundance values, feature abundance relative to all other features, for each sample [@Goodrich2014]. 
Both molecular and computational measurement process steps contribute to the overall measurement bias and dispersion [@Amore2016; @Goodrich2014; @brooks2015truth].
Proper measurement method evaluation allows for the characterization of how individual steps impact the measurement processes as a whole and determine where to focus efforts for improving the measurement process.
Appropriate datasets and methods are needed to evaluate the 16S rRNA marker-gene-survey measurement process.
A sample or dataset with "ground truth" is needed to characterize measurement process accuracy.
Numerous studies have evaluated quantitative and qualitative characteristics of the 16S rRNA measurement process using mock communities, simulated data, and environmental samples.


To assess the 16S rRNA sequencing measurement process qualitative characteristics of a mock communities are commonly used [@bokulich2016mockrobiota].
As the number of organisms in the mock community is known, the total number of features can be compared to the expected value.
The number of observed features in a mock community is often significantly higher than the expected number of organism [@Kopylova2014].
The higher than expected number of features is often attributed to sequencing and PCR artifacts as well as reagent contaminants [@brooks2015truth; @Huse2010].
A notable exception to this is mock community benchmarking studies evaluating sequencing inference method, such as DADA2 [@callahan2016dada2].
Sequence inference methods aim to reduce the number of sequence artifacts features.
While mock communities have a known value, they lack the feature diversity and relative abundance dynamic range of real samples [@bokulich2016mockrobiota].


The quantitative characteristics of 16S rRNA sequence data are normally assessed using mock communities and simulated data.
Mock communities of equimolar and staggered concentration are used to assess relative abundance estimate quantitative accuracy [@Kopylova2014].
Results from relative abundance estimates using mock communities generated from mixtures of DNA have shown taxonomic specific effects where individual taxa are under or over represented in a sample.
Taxonomic specific biases due to DNA extraction have been shown with Gram-negative bacteria having higher extraction efficiency compared to Gram-positive bacteria [@Costea2017; @Olson2012].
Mismatches in the primer binding sites are also responsible for taxonomic specific effects [@brooks2015truth; @klindworth2012evaluation; @Gohl2016].
Additionally, sequence template properties such as GC content, sequence secondary structure, and gene flanking regions have been attributed to taxon specific biases [@Pinto2012; @Hansen1998; @Gohl2016].
Simulated count tables have been used to assess differential abundance method, where specific taxa are artificially overrepresented in one set of samples compared to another [@McMurdie2014].
Using simulated data to assess log fold-change estimates only evaluates computational steps of the measurement process.


Quantitative and qualitative assessment can also be performed using sequence data generated from mixtures of environmental samples.
While simulated data and mock communities are useful in evaluating and benchmarking new methods one needs to consider that methods optimized for mock communities and simulated data are not necessarily optimized for the sequencing error profile and feature diversity of real samples.
Data from environmental samples, which are real samples, are often used to benchmark new molecular laboratory and computational methods.
However, without an expected to compare to, only measurement precision can be evaluated.
By mixing environmental samples, an expected value can be calculated using information from the unmixed samples and how they were mixed.
Mixtures of environmental samples have previously been used to evaluate gene expression measurements microarrays and RNAseq data [@parsons2015using; @pine2011adaptable; @thompson2005use].


In the present study, we developed a mixture dataset of extracted DNA from human stool samples for assessing 16S rRNA sequencing.
The mixture datasets were processed using three bioinformatic pipelines.
We developed metrics for qualitative and quantitative assessment of the bioinformatic pipeline results.
The quantitative results were similar across pipelines, but the qualitative results varied across pipelines.
We have made both the dataset and metrics developed in this study publicly available for evaluating bioinformatic pipelines.


<!-- A key step in the measurement process is clustering, the grouping of sequences into biologically relevant units, or operational taxonomic units (OTUs).
There are a number of different clustering methods.
The two most commonly used clustering methods are _de-novo_ clustering and open-reference clustering.
_de-novo_ clustering algorithms group sequences based on a defined similarity threshold [@westcott2015novo].
Open-reference clustering matches sequences to a set of previously clustered reference sequences (_de-novo_) then perform _de-novo_ clustering on the sequences in the dataset that do not match to sequences in the reference dataset with the desired similarity threshold [@he2015stability].
A third methods for clustering, sequence inference, uses statistical models or algorithms to differentiate true biological sequences within a dataset from sequencing errors [@callahan2016dada2;@amir2017deblur;@eren2015minimum].

Further challenging the measurement process is the compositional nature of the 16S data, that is the proportion of an organism within a sample is being measured and not the absolute abundance [@tsilimigras2016compositional].
Sequencing data only provide information regarding the relative abundance of organisms within a samples to other organisms within the same sample.
# When comparing the relative abundance of an organism across samples you are comparing organismal abundance relative to the rest of the organisms within the sample.
As a result an organism can have the same absolute abundance in two samples but due to differences in either the microbial community composition or for targeted assays such as 16S metagenomics differences in the proportion of human DNA in the sample.  -->



<!-- As with mock communities, simulated sequencing data has an expected value that can be used for benchmarking.  -->
<!-- However, as the sequencing error profile is not completely understood, simulated data does not recapitulate sequencing data generated from an environmental sample.  -->
