---
title: "Methods"
author: "Nate Olson"
date: '`r Sys.Date()`'
output:
  bookdown::pdf_document2:
    toc: FALSE
bibliography: [mgtst.bib]
---

### Titration Validation

qPCR was used to validate volumetric mixing and check for differences in the proportion of prokaryotic DNA across titrations.
To ensure that the two-sample titrations were volumetrically mixed according to the mixture design, independent ERCC plasmids were spiked into the unmixed PRE and POST samples [@baker2005external] (NIST SRM SRM 2374) (Table \@ref(tab:erccTable)).
The ERCC plasmids were resuspended in 100 $ng/\mu L$ tris-EDTA buffer and 2 $ng/\mu L$ was spiked into the appropriate unmixed sample.
Plasmids were spiked into unmixed samples after unmixed sample concentration was normalized to 12.5 $ng/\mu L$.
POST sample ERCC plasmid abundance was quantified using TaqMan gene expression assays (FAM-MGB) (Catalog # 4448892, ThermoFisher) specific to each ERCC plasmid using the TaqMan Universal MasterMix II (Catalog # 4440040, ThermoFisher Waltham, MA USA).
To check for differences in the proportion of bacterial DNA in the PRE and POST samples, titration bacterial DNA concentration was quantified using the Femto Bacterial DNA quantification kit (Zymo Research, Irvine CA).
All samples were run in triplicate along with an in-house _E. coli_ DNA $log_{10}$ dilution standard curve.
qPCR assays were performed using the QuantStudio Real-Time qPCR (ThermoFisher).
Amplification data and Ct values were exported as tsv files using QuantStudio™ Design and Analysis Software v1.4.1.
Statistical analysis was performed on the exported data using custom scripts in R [@R, https://github.com/nate-d-olson/mgtst_pub].

### Sequencing
The 45 samples (seven titrations and two unmixed samples for each of five subjects) were processed using the Illumina 16S library protocol (16S Metagenomic Sequencing Library Preparation, posted date 11/27/2013, downloaded from \url{https://support.illumina.com}).
This protocol specifies an initial 16S rRNA PCR followed by a sample indexing PCR, followed by normalization and sequencing.

A total of 192 16S rRNA PCR assays were run including four replicates per sample and 12 no-template controls, using Kapa HiFi HotStart ReadyMix reagents (KAPA Biosystems, Inc. Wilmington, MA).
The initial PCR assay targeted the V3-V5 region of the 16S rRNA gene, Bakt_341F and Bakt_806R [@klindworth2012evaluation].
The V3-V5 region is 464 base pairs (bp) long, with forward and reverse reads overlapping by 136 bp, using 2 X 300 bp paired-end sequencing [@yang2016sensitivity] ( \url{http://probebase.csb.univie.ac.at}).
Primer sequences include overhang adapter sequences for library preparation (forward primer 5'- TCG TCG GCA GCG TCA GAT GTG TAT AAG AGA CAG CCT ACG GGN GGC WGC AG - 3' and reverse primer 5'- GTC TCG TGG GCT CGG AGA TGT GTA TAA GAG ACA GGA CTA CHV GGG TAT CTA ATC C - 3').
For quality control, the PCR product was verified using agarose gel electrophoresis to check for appropriate size bands, and concentration measurements were made after the initial 16S rRNA PCR, the indexing PCR, and normalization steps.
DNA concentration was measured using the QuantIT Picogreen dsDNA Kit (Cat # P7589, ThermoFisher Scientific) and fluorescent measurements were made with a Synergy2 Multi-Detection MicroPlate Reader (BioTek Instruments, Inc, Winooski, VT).

<!-- Assay and instrument used at NIST
SpextraMax Accuclear Nano dsDNA Assay Bulk Kit (Part# R8357#, Lot 215737, Molecular Devices LLC. Sunnyvale CA, USA) and fluorescent measurements were made with a Molecular Devices SpectraMax M2 spectraflourometer (Molecular Devices LLC. Sunnyvale CA, USA).-->
<!-- Is that true for us as well?  @shao? We don't have such an instrument-->

Initial PCR products were purified using 0.8X AMPure XP beads (Beckman Coulter Genomics, Danvers, MA) following the manufacturer's protocol.
After purification, the 192 samples were indexed using the Illumina Nextera XT index kits A and D (Illumina Inc., San Diego CA) and then purified using 1.12X AMPure XP beads.
Prior to pooling purified sample concentration was normalized using SequalPrep Normalization Plate Kit (Catalog n. A10510-01, Invitrogen Corp., Carlsbad, CA), according to the manufacturer's protocol.
Pooled library concentration was checked using the Qubit dsDNA HS Assay Kit (Part# Q32851, Lot# 1735902, ThermoFisher, Waltham, MA USA).
Due to the low pooled amplicon library DNA concentration, a modified protocol for low concentration libraries was used.
The library was run on an Illumina MiSeq, and base calls were made using Illumina Real Time Analysis Software version 1.18.54.
Sequencing data quality control metrics for the 384 fastq sequence files (192 samples with forward and reverse reads) were computed using the Bioconductor `Rqc` package [@Rqc;@Bioconductor].

### Sequence Processing
Sequence data were processed using four bioinformatic pipelines: a _de-novo_ clustering method - Mothur [@schloss2009introducing], an open-reference clustering method - QIIME [@Caporaso2010], and a sequence inference method - DADA2 [@callahan2016dada2], and unclustered sequences as a control.
The code used to run the bioinformatic pipelines is available at \url{https://github.com/nate-d-olson/mgtst_pipelines}.


The Mothur pipeline follows the developer's MiSeq SOP [@schloss2009introducing;@kozich2013development].
The pipeline was run using Mothur version 1.37 (\url{http://www.mothur.org/})
We sequenced a larger 16S rRNA region, with smaller overlap between the forward and reverse reads, than the 16S rRNA region the SOP was designed.
Pipeline parameters were modified to account for the difference in overlap are noted for individual steps below.
The Makefile and scripts used to run the Mothur  pipeline are available https://github.com/nate-d-olson/mgtst_pipelines/blob/master/code/mothur.
The Mothur pipeline included an initial preprocessing step where the forward and reverse reads are trimmed and filtered using base quality scores and were merged into a single contig for each read pair.
The following parameters were used for the initial contig filtering, no ambiguous bases, max contig length of 500 bp, and max homopolymer length of 8 bases.
For the initial read filtering and merging step,
low-quality reads were identified and filtered from the dataset based on the presence of ambiguous bases, failure to align to the SILVA reference database (V119, \url{https://www.arb-silva.de/}) [@quast2012silva], and identification as chimeras.
Prior to alignment, the SILVA reference multiple sequence alignment was trimmed to the V3-V5 region, positions 6,388 and 25,316.
Chimera filtering was performed using UChime (version v4.2.40) without a reference database [@edgar2011uchime].
OTU clustering was performed using the OptiClust algorithm with a clustering threshold of 0.97 [@westcott2017opticlust].
The RDP classifier implemented in Mothur  was used for taxonomic classification against the Mothur  provided version of the RDP v9 training set [@wang2007naive].

The QIIME open-reference clustering pipeline for paired-end Illumina data was performed according to the online tutorial (Illumina Overview Tutorial (an IPython Notebook): open reference OTU picking and core diversity analyses, \url{http://qiime.org/tutorials/}) using QIIME version 1.9.1 [@Caporaso2010].
Briefly, the QIIME pipeline uses fastq-join (version 1.3.1) to merge paired-end reads [@aronesty2011ea] and the Usearch algorithm [@edgar2010search] with Greengenes database version 13.8 with a 97% similarity threshold [@desantis2006greengenes] was used for open-reference clustering.

DADA2, an R native pipeline was also used to process the sequencing data [@callahan2016dada2].
The pipeline includes a sequence inference step and taxonomic classification using the DADA2 implementation of the RDP naïve Bayesian classifier [@wang2007naive] and the SILVA database V123 provided by the DADA2 developers [@quast2012silva, \url{https://benjjneb.github.io/dada2/training.html}].

The unclustered pipeline was based on the Mothur  _de-novo_ clustering pipeline, where the paired-end reads were merged, filtered, and then dereplicated.
Reads were aligned to the reference Silva alignment (V119, \url{https://www.arb-silva.de/}), and reads failing alignment were excluded from the dataset.
Taxonomic classification of the unclustered sequences was performed using the same RDP classifier implemented in Mothur  used for the _de-novo_ pipeline.
To limit the size of the dataset the most abundant 40,000 OTUs (comparable to the Mothur  dataset), across all samples, were used as the unclustered dataset.

### Titration Proportion Estimates
The following linear model \@ref(eq:thetaInf) was used to infer the proportion of prokaryotic DNA in each titration, $\theta$.
Where $\textbf{Q}_{i}$ is a vector of titration $i$ feature relative abundance estimates and
$\textbf{Q}_{pre}$ and $\textbf{Q}_{post}$ are vectors of feature relative abundance estimates for the unmixed PRE and POST samples.
Average PCR replicate relative abundance values were calculated using a negative binomial model.

\begin{equation}
  \textbf{Q}_{i} = \theta_i (\textbf{Q}_{post} -\textbf{Q}_{pre}) + \textbf{Q}_{pre}
  (\#eq:thetaInf)
\end{equation}

To fit the model and prevent uninformative and low abundance features from biasing $\theta$ estimates only informative features meeting the following criteria were used.
Features included in the model were observed in at least 14 of the 28 total titration PCR replicates (4 replicates per 7 titrations),
demonstrated greater than 2-fold difference in relative abundance between the PRE and POST samples,
and were present in either all four or none of the PRE and POST PCR replicates.


16S rRNA sequencing count data is known to have a non-normal mean-variance relationship resulting in poor model fit for standard linear regression [@McMurdie2014].
Generalized linear models provide an alternative to standard least-squares regression.
The above model is additive and therefore unable to directly infer $\theta_i$ in log-space.
To address this issue, we fit the model using a standard least-squares regression then obtained non-parametric 95 \% confidence intervals for the $\theta$ estimates by bootstrapping with 1000 replicates.

### Qualitative Assessment
Our qualitative measurement assessment evaluated features only observed in unmixed samples (PRE or POST), _unmixed-specific_, or titrations, _titration-specific_.
_Unmixed-_ or _titration-specific_ features are due to differences in sampling depth (number of sequences) between the unmixed samples and titrations, artifacts of the feature inference process,
or PCR/sequencing artifacts.
Measurement process artifacts should be considered false positives or negatives.
Hypothesis tests were used to determine if differences in sampling depth could account for _unmixed-specific_ and _titration-specific_ features.
p-values were adjusted for multiple comparisons using the Benjamini & Hochberg method [@benjamini1995controlling].
For _unmixed-specific_ features, the binomial test was used to evaluate if true feature relative abundance is less than the expected relative abundance.
A binomial test could not be used to evaluate _titration-specific_ features, as the hypothesis would be formulated as such.
Given observed counts and the titration total feature abundance, the true feature relative abundance is equal to 0.
As non-zero counts were observed the true feature proportion is non-zero, and the test always fails.
Therefore, we formulated a Bayesian hypothesis test for _titration-specific_ features.


A Bayesian hypothesis test was used to evaluate if the true feature proportion is less than the minimum detected proportion.
The Bayesian hypothesis test was formulated using equation \@ref(eq:bht).
Which when assuming equal priors, $P(\pi < \pi_{min}) = P(\pi \geq \pi_{min})$, reduces to \@ref(eq:bht2).
For equations \@ref(eq:bht) and \@ref(eq:bht2) $\pi$ is the true feature proportion, $\pi_{min}$ is the minimum detected proportion, $C$ is the expected feature counts, and $C_{obs}$ is the observed feature counts.
Simulation was used to generate possible values of $C$, assuming $C$ has a binomial distribution given the observed sample total feature abundance, and a uniform probability distribution for $\pi$ between 0 and 1.
$\pi_{min}$ was calculated using the mixture equation \@ref(eq:mixEq) where $q_{pre,j}$ and $q_{post,j}$ are $min(\textbf{Q}_{pre})$ and $min(\textbf{Q}_{post})$ across all features for a subject and pipeline.
Our assumption is that $\pi$ is less than $\pi_{min}$ for features not observed in unmixed samples due to random sampling.


\begin{equation}
  \begin{split}
    p & = P(\pi < \pi_{min} | C \geq C_{obs}) \\
      & = \frac{P(C \geq C_{obs}| \pi < \pi_{min})P(\pi < \pi_{min})}{P(C \geq C_{obs}| \pi < \pi_{exp})P(\pi < \pi_{min}) + P(C \geq C_{obs}| \pi \geq \pi_{min})P(\pi \geq \pi_{min})} \\
  \end{split}
  (\#eq:bht)
\end{equation}

\begin{equation}
    p = \frac{P(C \geq C_{obs}| \pi < \pi_{min})}{P(C \geq C_{obs})}
  (\#eq:bht2)
\end{equation}


### Quantitative Assessment
For quantitative assessment, we compared observed relative abundance and log fold-changes to expected values derived from the titration experimental design.
Feature average relative abundance across PCR replicates was calculated using a negative binomial model, and used as observed relative abundance values ($obs$) for the relative abundance assessment.
Average relative abundance values were used to reduce PCR replicate outliers from biasing the assessment results.
Equation \@ref(eq:mixEq) and inferred $\theta$ values were used to calculate the expected relative abundance values ($exp$).
Relative abundance error rate is defined as $|exp - obs|/exp$.

We developed bias and variance metrics to assess feature performance.
The feature-level bias and variance metrics were defined as the median error rate and robust coefficient of variation ($RCOV=IQR/median$) respectively.
Mixed-effects models were used to compare feature-level error rate bias and variance metrics across pipelines with subject as a random effect.
Extreme feature-level error rate bias and variance metric outliers were observed,
these outliers were excluded from the mixed effects model to minimize biases due to poor model fit and were characterized independently.


Log fold-change between samples in the titration series including PRE and POST were compared to the expected log fold-change values to assess differential abundance log fold-change estimates.
Log fold-change estimates were calculated using EdgeR [@Robinson2010;@McCarthy2012].
Expected log fold-change for feature $j$ between titrations $l$ and $m$ is calculated using equation \@ref(eq:expLogFC), where $\theta$ is the proportion of POST bacterial DNA in a titration, and $q$ is feature relative abundance.
For features only present in PRE samples the expected log fold-change is independent of the observed counts for the unmixed samples and is calculated using \@ref(eq:expPreLogFC).
Due to a limited number of _PRE-specific_ features, both _PRE-specific_ and _PRE-dominant_ features were used in the differential abundance assessment.
_PRE-specific_ features were defined as features observed in all four PRE PCR replicates and not observed in any of the POST PCR replicates and _PRE-dominant_ features were also observed in all four PRE PCR replicates and observed in one or more of the POST PCR replicates with a log fold-change between PRE and POST samples greater than 5.


\begin{equation}
      logFC_{lm,j} = \log_2\left(\frac{\theta_l q_{post,j} + (1 - \theta_l) q_{pre,i}}{\theta_m q_{post,j} + (1 - \theta_m) q_{pre,j}}\right)
  (\#eq:expLogFC)
\end{equation}

\begin{equation}
      logFC_{lm,i} = log_2\left(\frac{1-\theta_l}{1-\theta_m}\right)
  (\#eq:expPreLogFC)
\end{equation}
