---
title: "Escherichia Assessment"
author: "Nate Olson"
date: "10/8/2017"
output: html_document
---

---
title: "Escherichia Features"
author: "Nate Olson"
date: '`r Sys.Date()`'
output:
  bookdown::pdf_document2: 
    toc: FALSE
---

```{r include = FALSE}
library(tidyverse)
library(ggpubr)
```


These features either represent different copies of the 16S gene, different Escherichia species, or sequencing errors generated by the high abundance E. coli. 

```{r eschData, include = FALSE}
## logFC 
logFC_edgeR <- readRDS("~/Desktop/logFC_edgeR_df.rds") 
logFC_edgeR <- logFC_edgeR %>% rename(feature_id = OTUname)
logFC_esch <- logFC_edgeR %>% filter(Rank5 == "f__Enterobacteriaceae" | Rank6 == "Escherichia/Shigella")

## Escherichia Features (QIIME Enterobacteriaceae)
esch_features <- logFC_esch %>% 
      filter(T1 == 0, T2 == 20) %>% 
      group_by(pipe, biosample_id) %>% 
      top_n(n = 3, logCPM) %>% 
      ungroup() %>% 
      select(pipe, feature_id) %>% 
      unique()


## Raw Counts
raw_counts <- readRDS("~/Desktop/raw_counts.RDS")
raw_counts_esch <- raw_counts %>% 
      right_join(esch_features) %>% 
      filter(biosample_id != "NTC") %>% 
      group_by(biosample_id, feature_id) %>% 
      mutate(med_count = median(count)) %>% 
      filter(med_count != 0)

## Enterobacteriaceae Features Updated
esch_features <- raw_counts_esch %>% ungroup() %>%
      select(pipe, feature_id) %>% unique()

## NB Counts  
nb_counts <- readRDS("~/Desktop/nb_counts_titrations.RDS")
nb_counts_esch <- nb_counts %>% 
      right_join(esch_features) %>% 
      filter(biosample_id != "NTC")

## Count Data Frame
esch_counts <- left_join(raw_counts_esch, nb_counts_esch)

## updated logFC
logFC_esch <- left_join(esch_features, logFC_esch)

### Relative Abundance Plot
scaleFUN <- function(x) sprintf("%.3f", x)
nb_plot <- nb_counts_esch %>% ungroup() %>% 
      mutate(pipe = factor(pipe, levels = c("unclustered","dada2","mothur","qiime"))) %>% 
      mutate(t_fctr = as.numeric(as.character(t_fctr))) %>% 
      ggplot() + 
      geom_line(aes(x = T2, y = nb_prop, color = feature_id)) + 
      geom_point(aes(x = T2, y = nb_prop, fill = feature_id), shape = 21) + 
      facet_grid(biosample_id ~ pipe, scales = "free") + theme_bw() +
      scale_y_continuous(trans = "log2", labels=scaleFUN,name = "Relative Abundance") +
      labs(x = "Titration")

### LogFC plot
logFC_plot <- logFC_esch %>% filter(T1 == 1, T2 != 20) %>% 
      ungroup() %>% 
      mutate(pipe = factor(pipe, levels = c("unclustered","dada2","mothur","qiime"))) %>% 
      mutate(T2 = as.numeric(as.character(T2))) %>% 
      ggplot() + 
      geom_line(aes(x = T2, y = logFC, color = feature_id)) + 
      geom_point(aes(x = T2, y = logFC, fill = feature_id), shape = 21) + 
      facet_grid(biosample_id~pipe) + 
      theme_bw() + labs(x = "Titration")
```



```{r eschPlot, fig.cap = "Relative abundance (A) and log fold-change estimtates (B) for high abundance \\textit{Escherichia} features. Colors are used to differentiate features."}
ggarrange(nb_plot, logFC_plot,
          ncol = 1, nrow = 2, align = "hv",common.legend = TRUE,legend = "bottom",labels = "AUTO")
```

