---
title: "Log Fold-Change Analysis"
author: "Nate Olson"
date: '`r Sys.Date()`'
output:
  pdf_document: default
  html_document: default
---

```{r message = FALSE}
library(ProjectTemplate)
cwd <- getwd()
setwd("../")
load.project()
setwd(cwd)
```

## Overview

## Code for Analysis
### Loading Pipeline Data
```{r}
mrexp_files <- list(
      dada2  = "../data/mrexp_dada2.RDS",
      mothur = "../data/mrexp_mothur.RDS",
      qiime  = "../data/mrexp_qiime_refclus_nochimera.RDS"
)
mrexp <- mrexp_files %>% map(readRDS) 

#Extracting metadata

meta_dat <- mrexp$mothur %>% pData()

##labeling PCR replicates
half1 <- paste(rep(c("A","B","C","D","E","F","G","H"), each = 6), 1:6, sep = "_")
sam_dat <- meta_dat %>% 
      mutate(pcr_half = if_else(pos %in% half1, "1","2"),
             pcr_rep = paste0(pcr_16S_plate,":",pcr_half)) %>% 
      select(sampleID, dilution,sam_names, pcr_rep) %>% 
      dplyr::rename(samID = sam_names)
```



#### Subsetting data to focus on one biological replicate
Only looking at biological replicate E01JH0004, to avoid overfitting the data.

```{r}
E01JH004_sams <- meta_dat %>% 
      filter(sampleID == "E01JH0004") %>% .$sam_names
mrexp_004 <- mrexp %>% 
      map(~.[,which(colnames(.) %in% E01JH004_sams)]) %>% 
      map(~.[which(rowSums(MRcounts(.)) > 0), ])
```

# Start of Log-Fold Change (Differential Abundance) Analysis  

## Log-Fold Change Variance
Looking at pre and post samples first

* Calculate variance for the observed log-fold change differences for pairwise combinations of PCR replicates  
      * Does it make sense to look at the distribution of the logFC values - to estiamte 95% CI  
      * Can also get quantiles (95% CI) from pairwise log-fold change values  

```{r}
E01JH004_pre_post_sams <- meta_dat %>% 
      filter(sampleID == "E01JH0004", dilution %in% c(0,-1)) %>% .$sam_names
mrexp_004_pre_post <- mrexp %>% 
      map(~.[,which(colnames(.) %in% E01JH004_pre_post_sams)]) %>% 
      map(~.[which(rowSums(MRcounts(.)) > 0), ])
```

```{r}
pre_post_meta <- meta_dat %>% filter(sampleID == "E01JH0004", dilution %in% c(0,-1))
pre_sams <- pre_post_meta %>% filter(dilution == 0) %>% .$sam_names
post_sams <- pre_post_meta %>% filter(dilution == -1) %>% .$sam_names
```


```{r}
pre_post_mat <- mrexp_004_pre_post$dada2 %>% metagenomeSeq::cumNormMat()

get_logFC <- function(pre, post){
      pre_post_mat[,pre]/pre_post_mat[,post] 
}

perm_logFC <- map2(rep(pre_sams,4), rep(post_sams, each = 4), get_logFC) %>% 
      set_names(paste0("X", 1:16)) %>% as_data_frame() %>% 
      add_column(feature_id = rownames(pre_post_mat)) %>% 
      gather("perm","FC", -feature_id)
```

```{r}
perm_logFC_summary <- perm_logFC %>% mutate(logFC = log2(FC+1)) %>% 
      group_by(feature_id) %>%
      summarise(logFC_mean = mean(logFC), logFC_median = median(logFC), 
                logFC_lq = quantile(logFC, 0.025, na.rm = TRUE), 
                logFC_uq = quantile(logFC, 0.975, na.rm = TRUE),
                logFC_var = var(logFC))
```

FC 0 - Post specific feature
FC NaN - Pre and Post 0
FC Inf - Pre specific feature


Filtering out features with mean premutation logFC values of 0, NaN, and Inf. 
Fold-change values of; 0 are post specific features, NaN are pre and post-specific features, and Inf are pre-specific features. 
See  `2017-01-29-logFC-permutation-feature-explore.Rmd` in reports for example features with 0, Inf, and NaN fold-change values.  

```{r}
perm_logFC_summary <- perm_logFC_summary %>% 
      filter(!is.nan(logFC_mean) | logFC_mean !=0 | !is.infinite(logFC_mean)) %>% 
      arrange(-logFC_mean)
```

## Feature logFC mean and variance

```{r}
ggplot(perm_logFC_summary) + geom_point(aes(x = logFC_mean, y = logFC_var))
```


Features with outlier variance values.
```{r}
perm_logFC_summary %>% filter(logFC_var > 0.8)
```

## Comparison of Permuted logFC value and variance with Diff Abu Method 

* Compare variance values to values estimated by differential abundance methods, metagenomeSeq, DESeq2, EdgeR, others?

## Log Fold-Change Bias 

### Treatment-Specific Features  
Use the following function to get treatment specific features
```{r}
#metagenomeSeq::uniqueFeatures()
```

* Expected logFC is the difference between titration factors for the titrations being compared   


* Compare expected values to those made by differential abundance methods, maybe permutation based method as well  

### Non-Treatment Specific Features  


## Log Fold-Change Bias-Variance Relationship

* Use EDA - scatter plot?


## Log Fold-Change Feature Exploration

* Correlating factors such as well position, primer matching, and GC content with observed variance and bias.
* Identify outliers for more detailed exploration
