---
title: "Log Fold-Change Analysis"
author: "Nate Olson"
date: '`r Sys.Date()`'
output:
  pdf_document: default
  html_document: default
---

```{r message = FALSE}
library(ProjectTemplate)
cwd <- getwd()
setwd("../")
load.project()
setwd(cwd)
```

## Overview

__NOTES__  
* Only use means for initial analysis  
* Evaluate the accuracy of fold change using replicates - defining confidence intervals can also use mixed effect with variance partition analysis   
* Looking for accuracy across biological replicates MA plots, use MA plots to make statements about methods

## Code for Analysis
### Loading Pipeline Data
```{r}
mrexp_files <- list(
      dada2  = "../data/mrexp_dada2.RDS",
      mothur = "../data/mrexp_mothur.RDS",
      qiime  = "../data/mrexp_qiime_refclus_nochimera.RDS"
)
mrexp <- mrexp_files %>% map(readRDS) 

#Extracting metadata

meta_dat <- mrexp$mothur %>% pData()

##labeling PCR replicates
half1 <- paste(rep(c("A","B","C","D","E","F","G","H"), each = 6), 1:6, sep = "_")
sam_dat <- meta_dat %>% 
      mutate(pcr_half = if_else(pos %in% half1, "1","2"),
             pcr_rep = paste0(pcr_16S_plate,":",pcr_half)) %>% 
      select(sampleID, dilution,sam_names, pcr_rep) %>% 
      dplyr::rename(samID = sam_names)
```



#### Subsetting data to focus on one biological replicate
Only looking at biological replicate E01JH0004, to avoid overfitting the data.

```{r}
E01JH004_sams <- meta_dat %>% 
      filter(sampleID == "E01JH0004") %>% .$sam_names
mrexp_004 <- mrexp %>% 
      map(~.[,which(colnames(.) %in% E01JH004_sams)]) %>% 
      map(~.[which(rowSums(MRcounts(.)) > 0), ])
```

# Start of Log-Fold Change (Differential Abundance) Analysis  

## Log-Fold Change Variance
Looking at pre and post samples first

* Calculate variance for the observed log-fold change differences for pairwise combinations of PCR replicates  
      * Does it make sense to look at the distribution of the logFC values - to estiamte 95% CI  
      * Can also get quantiles (95% CI) from pairwise log-fold change values  

```{r}
E01JH004_pre_post_sams <- meta_dat %>% 
      filter(sampleID == "E01JH0004", dilution %in% c(0,-1)) %>% .$sam_names
mrexp_004_pre_post <- mrexp %>% 
      map(~.[,which(colnames(.) %in% E01JH004_pre_post_sams)]) %>% 
      map(~.[which(rowSums(MRcounts(.)) > 0), ])
```

```{r}
pre_post_meta <- meta_dat %>% filter(sampleID == "E01JH0004", dilution %in% c(0,-1))
pre_sams <- pre_post_meta %>% filter(dilution == 0) %>% .$sam_names
post_sams <- pre_post_meta %>% filter(dilution == -1) %>% .$sam_names
```


```{r}
pre_post_mat <- mrexp_004_pre_post$mothur %>% metagenomeSeq::cumNormMat()

get_logFC <- function(pre, post){
      pre_post_mat[,pre]/pre_post_mat[,post] 
}

perm_logFC <- map2(rep(pre_sams,4), rep(post_sams, each = 4), get_logFC) %>% 
      set_names(paste0("X", 1:16)) %>% as_data_frame() %>% 
      add_column(feature_id = rownames(pre_post_mat)) %>% 
      gather("perm","FC", -feature_id)
```

```{r}
perm_logFC_summary <- perm_logFC %>% mutate(logFC = log2(FC+1)) %>% 
      group_by(feature_id) %>%
      summarise(logFC_mean = mean(logFC), logFC_median = median(logFC), 
                logFC_lq = quantile(logFC, 0.025, na.rm = TRUE), 
                logFC_uq = quantile(logFC, 0.975, na.rm = TRUE),
                logFC_var = var(logFC))
```

FC 0 - Post specific feature
FC NaN - Pre and Post 0
FC Inf - Pre specific feature


Filtering out features with mean premutation logFC values of 0, NaN, and Inf. 
Fold-change values of; 0 are post specific features, NaN are pre and post-specific features, and Inf are pre-specific features. 
See  `2017-01-29-logFC-permutation-feature-explore.Rmd` in reports for example features with 0, Inf, and NaN fold-change values.  

```{r}
perm_logFC_summary <- perm_logFC_summary %>% 
      filter(!is.nan(logFC_mean) | logFC_mean !=0 | !is.infinite(logFC_mean)) %>% 
      arrange(-logFC_mean)
```

## Feature logFC mean and variance

```{r}
ggplot(perm_logFC_summary) + geom_point(aes(x = logFC_mean, y = logFC_var))
```


Features with outlier variance values.
```{r}
perm_logFC_summary %>% filter(logFC_var > 0.8)
```

## Comparison of Permuted logFC value and variance with Diff Abu Method 

* Compare variance values to values estimated by differential abundance methods, metagenomeSeq, DESeq2, EdgeR, others?

## Log Fold-Change Bias 

### Treatment-Specific Features  

```{r}
dilution_samples <- meta_dat %>% filter(sampleID == "E01JH0004") %>% 
     .$dilution %>%  as.factor()

uniq_features <- metagenomeSeq::uniqueFeatures(mrexp_004$mothur,
                              cl = dilution_samples,
                              nsamples = 4, nreads = 1)
pre_specific_df <- uniq_features %>% as.data.frame() %>% 
      rownames_to_column(var = "featureNames") %>% 
      filter(`Samp. in -1` == 4, `Samp. in 0` == 0)

post_specific_df <- uniq_features %>% as.data.frame() %>% 
      rownames_to_column(var = "featureNames") %>% 
      filter(`Samp. in -1` == 0, `Samp. in 0` == 4)

## filter mrobj to get pre and post specific features
specific_features <- c(pre_specific_df$featureNames, post_specific_df$featureNames)
# want to normalize before subsetting features
#mrexp_specific <-mrexp$dada2 %>% {.[which(rownames(.) %in%  specific_features),]}
```




For the E01JH0004 sample DADA2 dataset there are `r nrow(pre_specific_df)` pre-treatment specific features, 
and `r nrow(post_specific_df)` pre-treatment specific features.

* Expected logFC is the difference between titration factors for the titrations being compared  

### Calculating logFC for differences between titrations
```{r}
meta_dat_004 <- meta_dat %>% filter(sampleID == "E01JH0004") %>% mutate(dilution = as.factor(dilution))

get_sams <- function(T1, T2){
      meta_dat_004 %>% filter(dilution %in% c(T1, T2)) %>% 
            .$sam_names
}

#### ASK HECTOR
## mrexp object defined globally
filter_mrexp <- function(sams, features){
      mrexp_004$mothur %>% {.[,which(colnames(.) %in% sams)]} %>% 
            # ?? Question-  better to filter normalize or filter features first
            ## Thinking is that you would want to normalize for all features so 
            ## that the normalization factor takes into account the total library
            ## size not just the features of interest
            filterData(present = 4, depth = 1) %>%
            ## not sure the best number to filter as present inclined to keep all
            ## features present in at least one sample - want to see where the model fails
            cumNorm(p = 0.75) %>%
            ## Filtering pre/ post specific features
            {.[which(rownames(.) %in%  features),]}
      
}

#### Ask Joe about numerator and denominator in logFC estimate
## Might be able to tell from model matrix
fit_model <- function(mrexp_obj, T1, T2){
      print(paste("Titrations:", T1, T2))
      pd <- pData(mrexp_obj)
      # defining levels to control the numerator and denominator in logFC estimate
      pd$dilution <- factor(pd$dilution, levels = c(T1, T2)) 
      mod <- model.matrix(~1 + dilution, data = pd) 
      print("fit model")
      fitFeatureModel(mrexp_obj, mod)
}
fit_safely <- safely(fit_model)
mrcoefs_safely <- safely(MRcoefs)

titrations <- dilution_samples %>% levels()
```






### Pre Specific Features
```{r}
## T1 and T2 represent the possible pairs of titrations
pre_fit_df <- data_frame(T1 = rep(titrations, each = length(titrations)) %>% as.numeric(), 
                         T2 = rep(titrations, length(titrations)) %>% as.numeric()) %>% 
      # excluding pre and post unmixed samples
      filter(!(T1 %in% c(-1,0)), !(T2 %in% c(-1,0)), T1 != T2) %>% 
      # Getting sample names for titrations
      mutate(sam_names = map2(T1, T2, get_sams)) %>% 
      mutate(features = rep(list(pre_specific_df$featureNames),42)) %>% 
      ## calculate expected logFC for pre specific features
      mutate(exp_logFC_post = - T1 + T2) %>% 
      mutate(exp_logFC_pre = log2((1-2^(-T1))/(1-2^(-T2)))) %>% 
      ## calculate logFC using fitFeature model for pre and post specific features
      mutate(mrexp_obj = map2(sam_names, features, filter_mrexp), # subset obj and normalize
             fit = pmap(list(mrexp_obj, T1, T2), fit_safely),
             fit_coefs = map(fit, ~mrcoefs_safely(.$result, number = Inf))) # get coefficient table
```

```{r}
mrexp_004$mothur %>% {.[,which(colnames(.) %in% pre_fit_df$sam_names[[2]])]} %>% 
            filterData(present = 4, depth = 1) %>% 
            ## not sure the best number to filter as present inclined to keep all
            ## features present in at least one sample - want to see where the model fails
            cumNorm(p = 0.75) %>% 
      {.[which(rownames(.) %in%  pre_fit_df$features[[2]]),]}

```

```{r}
pre_fit_df$fit
```

```{r}
pre_fit_df2 <- pre_fit_df %>% select(T1, T2, exp_logFC_pre, exp_logFC_post, fit_coefs) %>% 
      mutate(fit_coefs = map(fit_coefs, ~.$result),
             fit_coefs_class = map_chr(fit_coefs, class)) %>% 
      filter(fit_coefs_class == "data.frame") %>% 
      select(-fit_coefs_class) %>% 
      mutate(fit_coefs = map(fit_coefs, rownames_to_column, var = "featureNames")) %>% 
      unnest()
```

```{r}
pre_df <- pre_specific_df %>% select(featureNames, `Reads in -1`) %>% 
      filter(featureNames %in% paste0("Seq",c(11,129,156,229,269,325,351,500))) %>% 
      rename(`Reads in -1` = "reads") %>% 
      select(featureNames, reads)

pre_fit_df2 %>% ggplot(aes(x = exp_logFC_pre, y = logFC)) + 
      geom_text(aes(color = factor(T1), label = T2)) + facet_wrap(~featureNames) +
      geom_abline(aes(intercept = 0, slope = 1)) + 
      geom_text(data = pre_df, aes(x = -0.75, y = 1.1, label = reads))
```

### Post Specific Features
```{r}
## T1 and T2 represent the possible pairs of titrations
post_fit_df <- data_frame(T1 = rep(titrations, each = length(titrations)) %>% as.numeric(), 
                         T2 = rep(titrations, length(titrations)) %>% as.numeric()) %>% 
      # excluding pre and post unmixed samples
      filter(!(T1 %in% c(-1,0)), !(T2 %in% c(-1,0)), T1 != T2) %>% 
      # Getting sample names for titrations
      mutate(sam_names = map2(T1, T2, get_sams)) %>% 
      mutate(features = rep(list(post_specific_df$featureNames),42)) %>% 
      ## calculate expected logFC for pre specific features
      mutate(exp_logFC_pre = log2((2^(-T1)/(2^(-T2))))) %>% 
      mutate(exp_logFC_post = log2((1-2^(-T2))/(1-2^(-T1)))) %>% 
      ## calculate logFC using fitFeature model for pre and post specific features
      mutate(mrexp_obj = map2(sam_names, features, filter_mrexp), # subset obj and normalize
             fit = pmap(list(mrexp_obj, T1, T2), fit_safely),
             fit_coefs = map(fit, ~mrcoefs_safely(.$result, number = Inf))) # get coefficient table

post_fit_coefs <- post_fit_df %>% select(T1, T2, exp_logFC_pre, exp_logFC_post, fit_coefs) %>% 
      mutate(fit_coefs = map(fit_coefs, ~.$result),
             fit_coefs_class = map_chr(fit_coefs, class)) %>% 
      filter(fit_coefs_class == "data.frame") %>% 
      select(-fit_coefs_class) %>% 
      mutate(fit_coefs = map(fit_coefs, rownames_to_column, var = "featureNames")) %>% 
      unnest()
```

```{r}
post_df <- post_specific_df %>% select(featureNames, `Reads in 0`) %>% 
      filter(featureNames %in% (post_fit_coefs$featureNames %>% unique())) %>% 
      rename(`Reads in 0` = "reads") %>% 
      select(featureNames, reads)

post_fit_coefs %>% left_join(post_df) %>% #filter(reads %in% 500:1000) %>% #, T1 %in% 2:4, T2 %in% 2:4) %>% 
      ggplot(aes(x = exp_logFC_post, y = logFC)) + 
      geom_text(aes(color = factor(T1), label = T2)) + facet_wrap(~featureNames) +
      geom_abline(aes(intercept = 0, slope = 1)) + 
      geom_text(aes(x = -0.75, y = 1.1, label = reads))
```

```{r}
post_fit_coefs %>% left_join(post_df) %>% 
      mutate(residuals = exp_logFC_post - logFC,
             read_bins = cut_width(reads, width = 50)) %>% 
      ggplot() + geom_density(aes(x = residuals, fill = read_bins)) +
      facet_wrap(~read_bins)
```


* Compare expected values to those made by differential abundance methods, maybe permutation based method as well  

### Non-Treatment Specific Features  


## Log Fold-Change Bias-Variance Relationship

* Use EDA - scatter plot?


## Log Fold-Change Feature Exploration

* Correlating factors such as well position, primer matching, and GC content with observed variance and bias.
* Identify outliers for more detailed exploration
