---
title: "Pipeline Characterization"
author: "Nate Olson"
date: '`r Sys.Date()`'
always_allow_html: yes
output:
  pdf_document: default
  html_document: default
---

```{r setup, warning=FALSE, message=FALSE, echo = FALSE}
library(ProjectTemplate)
cwd <- getwd()
setwd("../")
load.project()
setwd(cwd)
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
```

```{r echo = FALSE}
pipeline_dir <- "../../mgtst_pipelines"
mrexp_files <- list(
      dada2 = file.path(pipeline_dir, "dada2/dada_mrexp.rds"),
      mothur =  file.path(pipeline_dir, "mothur/mothur_mrexp.rds"),
      qiime =  file.path(pipeline_dir, "qiime/qiime_mrexp.rds")
)
mrexp <- mrexp_files %>% map(readRDS)
```


The sequencing dataset was processed using three bioinformatic pipelines. 
The following analysis provides and overview of the resulting dataset; count table characteristics,  sample coverage, overall similarity between pre- and post-treatment samples.

## Count Table Characteristics
```{r}
## Loading seq count data
fastq_dir <- file.path(pipeline_dir, "data")
dat_files <- list.files(path = fastq_dir,pattern = "001.fastq.gz$",
                           recursive = TRUE, full.names = TRUE)
seq_ds_id <- dat_files %>% str_split("/") %>% flatten_chr() %>% 
    grep(pattern = "fastq.gz", .,value = TRUE) %>%
    str_replace(".fastq.gz","") %>% paste0("Fq_",.) %>% 
    str_replace("-",".")
names(dat_files) <- seq_ds_id 

# from script but not saved
read_groups <- rep(NA,length(dat_files))
read_groups[grepl("/1-.*_R1", dat_files)] <- "plate1_R1"
read_groups[grepl("/1-.*_R2", dat_files)] <- "plate1_R2"
read_groups[grepl("/2-.*_R1", dat_files)] <- "plate2_R1"
read_groups[grepl("/2-.*_R2", dat_files)] <- "plate2_R2"

qa_list <- readRDS("../data/rqcQA_list.rds") 

## Run 
grp_df <- data_frame(read_group = read_groups, 
                     seq_ds_id, 
                     filename = basename(dat_files)) %>%
    separate(read_group,c("plate","Read")) %>% 
    mutate(ill_id = str_replace(filename, "_.*",""))

## read count data 
qa_file_info <- perFileInformation(qa_list) %>% 
    select(-format,-path)

## study metadata
meta_df <- sampleSheet %>% 
    mutate(pos_ns = str_replace(pos, "_",""),
           ill_id = paste(pcr_16S_plate, pos_ns, sep = "-")) %>% 
    filter(seq_lab == "JHU", barcode_lab == "JHU") %>% 
    mutate(pcr_16S_plate = as.character(pcr_16S_plate)) %>% 
    left_join(grp_df) %>% left_join(qa_file_info)
```

```{r}
## number of OTUs
features <- mrexp %>% map_dbl(NROW)

## spareseness
extract_samples <- function(mrobj){
      sam_names <- pData(mrobj) %>% rownames_to_column() %>% 
            filter(biosample_id != "NTC") %>% .$rowname
      mrobj[,colnames(mrobj) %in% sam_names]
}

calc_sparsity <- function(mat){
      nentry <- length(mat)
      nzero <- sum(mat == 0)
      ## calculate sparsity 
      nzero/nentry
}

sparsity <- mrexp %>% map(extract_samples) %>% map(MRcounts, sl = 1) %>% map_dbl(calc_sparsity)
```


```{r}
## Read loss / Filter rate 
count_df <- mrexp %>% map(extract_samples) %>% 
      map(MRcounts,sl = 1) %>% 
      map(colSums) %>% map(as.data.frame) %>% 
      map_df(rownames_to_column, var = "id", .id = "pipe") %>% 
      dplyr::rename(counts = `.x[[i]]`)

count_summary <- count_df %>% group_by(pipe) %>% 
      summarise(count_med = median(counts) %>% round(0),
                count_min = min(counts) %>% round(0),
                count_max = max(counts) %>% round(0)) %>% 
      mutate(`Sample Coverage` = paste0(count_med, " (", count_min, "-", count_max,")")) %>% 
      select(pipe, `Sample Coverage`)

reads_df <- meta_df %>% filter(Read == "R1") %>% 
      select(ill_id, reads, group) %>% dplyr::rename(id = ill_id) 

filter_rate_df <- count_df %>% left_join(reads_df) %>%
      mutate(filter_rate = 1 - counts/reads) 

filter_rate_summary <- filter_rate_df %>% group_by(pipe) %>% 
      summarise(filt_med = median(filter_rate) %>% round(2),
                filt_min = min(filter_rate) %>% round(2),
                filt_max = max(filter_rate) %>% round(2)) %>% 
      mutate(`Filter Rate` = paste0(filt_med, " (", filt_min, "-", filt_max,")")) %>% 
      select(pipe, `Filter Rate`)
```


Total number of features for all samples and count table sparsity for the bioinformatic pipelines. 
The expectation is that this dataset will be more sparse relative to other datasets due to the redundant nature of the samples where 35 of the 45 samples are derived directly from the other 10 samples and that there are four PCR replicates for each sample. 
Sparsity lower for De-novo clustering (QIIME) than sequence inference (DADA2) even though DADA2 has fewer total features. 
Different pipelines have different approaches for handling low quality reads. See individual pipeline reports for which steps reads are excluded from the datasets. 
QIIME pipeline has the highest filter rate while the highest number of features per sample. 

```{r}
data_frame(pipe = names(mrexp), 
           Features = features, 
           Sparsity = round(sparsity,2)) %>% 
      left_join(count_summary) %>% left_join(filter_rate_summary) %>% 
      dplyr::rename(Pipelines = pipe) %>% 
      knitr::kable(caption = "Summary statistics for the different bioinformatic pipeliens. DADA2 is a denoising sequence inference pipeline, QIIME is a open-reference clustering pipeline, and mothur is a de-novo clustering pipeline. No template controls were excluded from summary statistics. Sparsity is the proportion of 0's in the count table. Features is the total number of OTUs (QIIME and mothur) or SVs (DADA2) in the count. Sample coverage is the median and range (minimum - maximum) per sample total feature abundance. Filter rate is the proportion of reads that were removed while processing the sequencing data for each bioinformatic pipeline.")
```


## Sample Coverage 
Rarefaction curves are commonly used to demonstrate how well a community has been sampled ( __REF__ ).
The slope of the curve decreases as sample coverage increases. 
Rarefaction curve for sequence inference has flattened out indicating the community has been fully sampled. 
Whereas the curves have not reached their asymptotes for _de novo_ and open reference clustering indicating that the community has not been fully sampled. 
Comparison of the individual sample rarefaction curves to the rarefaction curves after aggregating counts for the four replicates indicates that the rarefaction curves of the individual for the _de novo_ and open-reference clustering more representative of the diversity in the four replicates combined than the features generated using the sequence inference method. 

```{r}
get_curve_df <- function(mrobj){
      count_tbl <- mrobj[,pData(mrobj)$titration %in% c(20,0)]@assayData$counts %>% t() 
      dim(count_tbl)
      pData(mrobj)
      samID <- rownames(count_tbl)
      
      curve_dat <- rarecurve_dat(count_tbl, step = 500)
      
      curve_df <- curve_dat %>% set_names(samID) %>% map(data.frame) %>%
            map_df(rownames_to_column, var = "sample", .id = "samID") %>%
            mutate(sample = str_replace(sample, "N","") %>% as.numeric())
      
      colnames(curve_df)[3] <- "OTUs"
      
      pData(mrobj) %>% rownames_to_column(var = "samID") %>%
            # separate(id, into = c("pcr_16S_plate","well"),remove = FALSE, sep = "-") %>% 
            mutate(pcr_16S_plate = as.numeric(pcr_16S_plate)) %>% 
            select(biosample_id, samID, titration, pcr_16S_plate) %>%
            right_join(curve_df)
}
```

```{r results = 'hide'}
curve_df <- mrexp %>% map_df(get_curve_df, .id = "pipe")
```


```{r}
get_curve_agg_df <- function(mrobj){
      count_tbl <- mrobj@assayData$counts %>% as.matrix() %>% t() 
      samID <- rownames(count_tbl)
      count_tbl[1:10,1:10]
      curve_dat <- rarecurve_dat(count_tbl, step = 500)
      curve_df <- curve_dat %>% set_names(samID) %>% map(data.frame) %>%
            map_df(rownames_to_column, var = "sample", .id = "samID") %>%
            mutate(sample = str_replace(sample, "N","") %>% as.numeric())

      colnames(curve_df)[3] <- "OTUs"
      curve_df
}

agg_sample <- function(mrobj){
      mrobj <- mrobj[,pData(mrobj)$titration %in% c(20,0)]
      aggfct <- pData(mrobj) %>% unite(aggfct, biosample_id, titration, sep = ":") %>% 
            .$aggfct %>% factor()   
            
      metagenomeSeq::aggregateBySample(mrobj, fct = aggfct, aggfun = rowSums) 
}
agg_curve <- mrexp %>% map(extract_samples) %>% map(agg_sample) %>% 
      map_df(get_curve_agg_df, .id = "pipe")


agg_curve_df <- agg_curve %>% 
      separate(samID,c("biosample_id","titration"), sep = ":", remove = FALSE) 
```

```{r fig.cap = "Rarefaction curves for the unmixed pre- and post-treatment samples. The solid lines represent individual PCR replicates and the dashed lines the pooled replicates."}
agg_curve_df <- agg_curve_df %>% 
      mutate(Treatment = if_else(titration == 0, "Post", "Pre"))

curve_df %>% mutate(Treatment = if_else(titration == 0, "Post", "Pre")) %>% 
      ggplot() + 
      geom_path(aes(x = sample, y = OTUs, color = Treatment, group = samID)) +  
      geom_path(data = agg_curve_df, 
                aes(x = sample, y = OTUs, color = Treatment, group = samID), 
                alpha = 0.5, linetype = 2) + 
      facet_grid(biosample_id~pipe) + #, scale = "free") + 
      theme_bw() + theme(legend.position = "bottom") +
      labs(x = "Sampling Depth", y = "Features")
```


## Similarity Between Pre- and Post-Treatment Samples 
Replicates group together.  
Overall samples group together more by treatment status than biological sample \@ref(fig:ordPlots). 
__TODO__ Clean-up formatting  

```{r}
ps_files <- list(
      dada2 = file.path(pipeline_dir, "dada2/dada_ps.rds"),
      mothur =  file.path(pipeline_dir, "mothur/mothur_ps.rds"),
      qiime =  file.path(pipeline_dir, "qiime/qiime_ps.rds")
)
ps <- ps_files %>% map(readRDS) 
```

```{r}
sample_set <- sampleSheet %>% filter(barcode_lab == "JHU", seq_lab == "JHU", 
                       titration %in% c(20,0)) %>% 
      select(biosample_id, titration, pcr_16S_plate, pos) %>% 
      unite(sample_name, pcr_16S_plate, pos, sep = "-")
```


```{r}
subset_ps <- function(ps_obj){
      
      ## subset by titrations of interest
      ps_subset <- subset_samples(ps_obj, titration %in% c(20,0)) 
      ps_subset@sam_data$titration <- factor(ps_subset@sam_data$titration)
      
      ps_subset
}
```

```{r}
sample_ids <- sampleSheet$biosample_id %>% unique()
subset_obj <- ps %>% map(subset_ps)
subset_ord <- subset_obj %>% map(ordinate, "PCoA", "bray")
```

```{r}
pl_list <- map2(subset_obj, subset_ord, plot_ordination, 
                type = "sample",color = "biosample_id", shape = "titration")
```

```{r ordPlots, results='hold', fig.cap = "Pre- and Post-treatment PCoA for Bray-Curtis Beta diversity.", fig.width = 3, fig.height = 3}
pl_list$dada2 + ggtitle("DADA2 Ordination") + theme_bw() + theme(legend.position = "none")
pl_list$mothur + ggtitle("Mothur Ordination") + theme_bw() + theme(legend.position = "none")
pl_list$qiime + ggtitle("QIIME Ordination") + theme_bw() 
```


# Conclusion


# Session information
```{r}
s_info <- devtools::session_info()
print(s_info$platform)
s_info$packages %>% filter(`*` == "*") %>% select(-`*`) %>% 
      knitr::kable()
```