---
title: "Pop 16S pipeline"
author: "Nate Olson"
date: '`r Sys.Date()`'
output:
  pdf_document: default
  html_document: default
---

```{r}
library(tidyverse)
library(stringr)
```

1. Using the "mgtst.trim.contigs.good.unique.good.fasta" from mothur pipeline as input.  
2. Clustering - dnaclust on cbcb cluster using an interactive session
`dnaclust_linux_release3/dnaclust -d -l -s 0.99 -t 8  -e 1182016 -i mgtst.trim.contigs.good.unique.fasta > mgtst.trim.contigs.good.unique.clusters` 
3. Updating Table from mothur based on dnaclust clustering

```{r}
count_tbl <- readr::read_tsv("~/Projects/16S_etec_mix_study/analysis/pipelines/mothur/data/process/mgtst.trim.contigs.good.count_table")
count_df<- as.data.frame(count_tbl)
rownames(count_df) <- count_df$Representative_Sequence
count_mat <- as.matrix(count_df[,-c(1,2)])
```

__Group sequences by OTU for row subsetting__
Getting cluster output
```{r}
cluster_seqs <- readr::read_lines("~/Projects/16S_etec_mix_study/analysis/pipelines/pop/mgtst.trim.contigs.good.unique.clusters")
cluster_meta <- cluster_seqs[1]
cluster_list <- as.list(cluster_seqs[-1]) %>% set_names(paste0("OTU_",1:length(.)))
cluster_assignment <- cluster_list %>% map(gsub,pattern = "\t$",replacement = "")
cluster_assignment <- cluster_assignment %>% map(~str_split(.,pattern = "\t")[[1]])
```

```{r}
cluster_assign <- unlist(cluster_assignment)
row_assign <- names(cluster_assign)[match(rownames(count_mat), cluster_assign)]
cluster_mat <- count_mat %>% rowsum(group = row_assign)
```
Need Taxonomic Assignment for OTUs
```{r}
saveRDS(cluster_mat, "pop_counts.rds")
```


