---
title: "Sequencing Data Quality Assessment"
author: "Nate Olson"
date: '`r Sys.Date()`'
always_allow_html: yes
output:
  bookdown::pdf_document2: default
  html_document: default
---

```{r setup, warning=FALSE, message=FALSE, echo = FALSE}
library(ProjectTemplate)
cwd <- getwd()
setwd("../")
load.project()
setwd(cwd)
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE) 
project_dir <- "../../16S_etec_mix_study/"
pipeline_dir <- "../../mgtst_pipelines/"
```

Quality assessment of sequencing run summarizing the PhiX error rate, number of reads per sample, and quality score distributions over the length of the reads. 
Differences in the number of reads per sample are important for normalization and quality score distribution over read length is relevant to merging forward and reverse reads as well as the quality of the sequencing data in the middle of the amplicon.


```{r fq_data_files}
fastq_dir <- file.path(project_dir, "data/")
dat_files <- list.files(path = fastq_dir,pattern = "001.fastq.gz$",
                           recursive = TRUE, full.names = TRUE)
seq_ds_id <- dat_files %>% str_split("/") %>% flatten_chr() %>% 
    grep(pattern = "fastq.gz", .,value = TRUE) %>%
    str_replace(".fastq.gz","") %>% paste0("Fq_",.) %>% 
    str_replace("-",".")
names(dat_files) <- seq_ds_id
```

```{r loadQA}
# from script but not saved
read_groups <- rep(NA,length(dat_files))
read_groups[grepl("/1-.*_R1", dat_files)] <- "plate1_R1"
read_groups[grepl("/1-.*_R2", dat_files)] <- "plate1_R2"
read_groups[grepl("/2-.*_R1", dat_files)] <- "plate2_R1"
read_groups[grepl("/2-.*_R2", dat_files)] <- "plate2_R2"

qa_list <- readRDS("../data/rqcQA_list.rds")
```

```{r metadat}
# Tidy data

## Run 
grp_df <- data_frame(read_group = read_groups, 
                     seq_ds_id, 
                     filename = basename(dat_files)) %>%
    separate(read_group,c("plate","Read")) %>% 
    mutate(ill_id = str_replace(filename, "_.*",""))

## read count data 
qa_file_info <- perFileInformation(qa_list) %>% 
    select(-format,-path)

## study metadata
meta_df <- sampleSheet %>% 
    mutate(pos_ns = str_replace(pos, "_",""),
           ill_id = paste(pcr_16S_plate, pos_ns, sep = "-")) %>% 
    filter(seq_lab == "JHU", barcode_lab == "JHU") %>% 
    mutate(pcr_16S_plate = as.character(pcr_16S_plate)) %>% 
    left_join(grp_df) %>% left_join(qa_file_info)
```


#### Read Counts   
Two barcoded experimental samples have less than 50,000 reads \@ref(fig:readCount). The rest of the samples with less than 50,000 reads are negative PCR controls (NTC). Sample E01JH0016 titration 5 position F9 of plate 1 initial 16S PCR failed.  
Excluding the one failed reaction the total range in the observed number of sequences per sample is approximately 40,000 to 150,000 reads.  

__TODO__ Figure out why E01JH0011 titration 3 position D2 plate 2 is also low, look at picogreen post normalization data. 
__TODO__ Comparison to differences in number of reads per sample observed in other studies. 

```{r tableCount}
meta_df %>% mutate(exp_ntc = ifelse(biosample_id == "NTC", "NTC","EXP")) %>% 
    filter(Read == "R1", ill_id != "1-F9") %>% 
    group_by(exp_ntc, plate) %>% 
    summarise(mean_lib_size = mean(reads),
              min_lib_size = min(reads),
              median = median(reads),
              max_lib_size = max(reads)) %>% 
    kable(caption = "Summary statistics for experimental and no template control samples by PCR plate and read.")
```


```{r readCount, fig.cap="Number of reads per barcoded sample (Library Size), by read direction (X-facet) and replicate 16S PCR plate (Y-facet). Vertical line indicates 50,000 reads per barcoded sample.", fig.lp = "fig:"}
meta_df %>% # filter(Read == "R1") %>%
    ggplot() + 
        geom_histogram(aes(x = reads), position = "dodge") + 
        facet_grid(plate~.) + scale_x_log10() + 
        geom_rug(aes(x = reads, color = biosample_id)) +
        geom_vline(aes(xintercept = 50000), color = "grey80") +
        theme_bw() + labs(x = "Library Size (log10)", 
                          y = "Count", color = "Sample ID") +
        theme(legend.position = "bottom")
```

#### PhiX Error Rate  

```{r munge_error_dat} 

sav_dir <- file.path(project_dir,"data/basespace_sav/")
sav_list <- list(
    basespace_v3 = file.path(sav_dir,"run_3861867_sav"),
    april2016_v3 = file.path(sav_dir,"run_18527531_sav") 
)

error_df <- sav_list %>% map(savR) %>%
    map_df(errorMetrics,.id = "ds") %>%
    separate(ds, c("ds","chemistry"),sep = "_") %>%
    mutate(read_len = if_else(chemistry == "v3",301, 251),
           read = if_else(cycle < read_len, "R1","R2"))
```

The sequencing error rate data was obtained from the Basespace sequencing run report downloaded from Basespace (SAV file). 
Error rate is compared to a 16S public dataset on basespace ( 16S-Metagenomic-Library-Prep run id 3861867). 
The error rate for was higher at the ends of both R1 and R2 compared to the public dataset \@ref(fig:phixError). 
This higher error rate may impact the number of reads pairs successfully merged and the accuracy of the sequence in the middle of the amplicon. 


```{r phixError, fig.cap = "PhiX error rate for JHU barcoded samples compared to the public dataset. R1 and R2 are the forward and reverse reads and cycle is the position in the read.", fig.lp="fig:"}
ggplot(error_df) +
    geom_point(aes(x = cycle, y = errorrate, color = ds), alpha = 0.01) +
    geom_smooth(aes(x = cycle, y = errorrate, color = ds)) +
    theme_bw() +
    labs(x = "Cycle", y = "PhiX Error Rate", color = "Dataset") +
    facet_grid(.~read, scales = "free_x")
```

#### Base Quality Score 

```{r cycle_metrics}
## amplicon position __Cycle Level Metrics__
amp_pos_df <- data_frame(cycle = rep(1:300, 2), 
                         Read = rep(c("R1","R2"), each = 300),
                         amp_pos = c(1:300,c(460 - 1:300)))

qa_cycle_q_df <- qa_list %>% map_df(perCycleQuality) %>% 
    as_data_frame() %>% 
    mutate(cycle = as.numeric(as.character(cycle))) %>% 
    filter(count != 0) %>% # not sure if this impacts the smoothing function ...
    left_join(meta_df) %>% left_join(amp_pos_df)
```

Cycle base quality score is more homogeneous from PCR plate 2 samples than plate 1 \@ref(fig:lowcycleq). For the expected overlap region, based on primer positions and read lengths (16S PCR fig), the forward read has consistently higher base quality scores relative to the reverse read. 
The sample with a lower base quality score is titration 4 of biological replicate E01JH0017 (sample 1-F4).  

```{r lowcycleq, fig.cap = "Smoothing spline of the base quality score by sequencing cycle. Vertical lines indicate approximate overlap region between forward and reverse reads. This is not a read level analysis but average quality score for individual barcoded datasets.", fig.lp="fig:"}
qa_cycle_q_df %>% filter(biosample_id != "NTC", ill_id != "1-F9") %>% ggplot(aes(x = amp_pos, y = score)) + 
    geom_vline(aes(xintercept = 300), color = "grey60") +
    geom_vline(aes(xintercept = 160), color = "grey60") +
    geom_smooth(aes(weight = count, color = biosample_id, group = filename)) + 
    geom_vline(aes(xintercept = 280), color = "grey40", linetype = 2) +
    facet_grid(plate~.) + theme_bw() + 
    scale_x_continuous(breaks = c(0,150,300, 450)) +
    labs(x = "Amplicon Position", 
         y = "Base Quality Score", color = "Biological Replicate") + 
    theme(legend.position = "bottom")
```

## Conclusion
The low variabiltiy in number of reads per sample suggests that normalization methods are less likely to impact the results compared other datasets with larger variability in the number of reads per sample. 
The higher error rates at the ends of the reads has the potential to impact the quality of the reads in the overlap region and limit the number of successfully merged read pairs.  



# Session information
```{r}
s_info <- devtools::session_info()
print(s_info$platform)
s_info$packages %>% filter(`*` == "*") %>% select(-`*`) %>% 
      knitr::kable()
```