---
title: "Supplemental Material"
author: "Nate Olson"
date: "October 25, 2016"
output: html_document
---

# Supplemental
## Wet Lab
### Library Prep Supplemental 
__PCR reaction mix__  

| Reagent | Volume Per Rxn (ul)|
|:-|-:|
|2XKAPA HiFi HotStart Ready Mix|    12.5 |
|16S_F primer (1uM)|    5 |
|16S_R primer (1uM)|    5|
|DNA (5ng/ul in 10mM Tris pH8.5)|   2.5|
| Total   | 25|

__PCR Cocktail__  

| Reagent | Volume Per Rxn (ul)| Volume for 220X (ul) |
|:-|-:|-:|
|2XKAPA HiFi HotStart Ready Mix|    12.5 | 2750|
|16S_F primer (1uM)|    5 | 1100 |
|16S_R primer (1uM)|    5| 1100 |
|H20|      1.5| 330 |
| DNA (12.5ng/ulin 10mM Tris pH8.5)   | 1 ||

* PCR Prep layout \@ref(fig:plate_map)  

```{r plate_map, echo=FALSE, fig.cap="Image of tubes layout prior to 16S PCR."}
knitr::include_graphics("img/1st_PCR_PlateMap_20160224.jpg")
```


* Thermocycling conditions 

|Stage | Temp (deg C) | Time (min) |
|-:|-:|-:|
|Denaturation |95 | 3  |
|Amplification |95 | 0.5|
|25 X |55 | 0.5|
|   |72 | 0.5|
|Final Extension|72 |   5|
|Hold |4|     forever|

## Index Reaction
* Reaction information


## Pipelines 
Detailed description of the 16S bioinformatic pipelines used in the study.

### DADA2 Summary  
1. Quality filter read pairs, remove reads with ambiguous bases (Ns) or more than 4 expected errors, trim ends of reads with quality score of 2 or to 290 bp and 220 bp for forward and reverse reads respectively, and the first 10 bases are trimmed.  
      * Number of expected errors calculated based on quality scores, `EE = sum(10^(-Q/10))`.   
      * Quality score of 2 used by Illumina to indicate end of good sequencing data.   

2. The forward and reverse reads are dereplicated. A consensus (average) quality score is assigned for each position.  
3. Sequence inference, denoising - error correction step.  
4. Merging forward and reverse read pairs. Uses global ends-free alignment and requires exact overlap for merging.
5. Chimera removal filters sequences with complementary regions matching more abundant sequences. Chimeras are identified when  Needleman-Wunsch global alignments between a sequence and all more abundance sequences result in perfect matches for left and right partitions of the sequence to two different parent sequences. 

### Mothur Summary
- merging paired end reads using needleman and filtering merged contigs
- aligns sequences to reference alignment - SILVA
    - find closet match in reference multiple sequence alignment using a k-mer based method then aligns to closest match using needleman
- Removes duplicates and pre.clustering to reduce the number of sequences clustered
    - Pre.cluster - pseudo-single linkage algorithm
        - Ranks sequences in order of abundance - clusters less abundant sequences within the specified edit distance from the more abundant sequences
- Chimera filtering using UCHIME
    - looks for chimeras using more abundant sequences within a sample as reference
- Classify sequences using RDP with 80% threshold
- performs average neighbor clustering - distance threshold 0.03 after splitting sequences based on taxonomy
    - level 4- Order
- Classifies OTUs based on the consensus of the sequence classifications for the sequences assigned to the OTU

### QIIME Summary  
- Merging paired-end reads 
      - fastq-join - Erik Aronesty, 2011. ea-utils : “Command-line tools for processing biological sequencing data” (https://github.com/ExpressionAnalysis/ea-utils/blob/wiki/FastqJoin.md)
    - fastq-join uses sqr(distance)/len to anchor alignments
    - does not use Smith-Waterman alignment
    - method not suitable for sequencing technologies with high insertion and deletion rates
    - Reference describing methods http://benthamopen.com/contents/pdf/TOBIOIJ/TOBIOIJ-7-1.pdf
- Quality filtering - phred >= 3 default value, minimum base quality score
- Clustering: open reference using UCHIME and greengenes 13.8 database clustered at threshold of 0.97
      - reference based clustering of sequences
      - unclustered seqs: (1) random subset of sequences selected and clustered. (2) New cluster centers are used as references in reference based clustering. (3) Remaining sequences are de novo clustered.  
      - Clusters from four steps are combined.  
      - taxonomic assignments from UCLUST reference based clustering
- pyNAST refernece alignment - cluster centers and associated reads that fail to align to the reference database are removed from the analysis

* Seq budget - summarize fate of sequences, number successful merged read pairs, chimera filtered, alignment?, ect.
    * Table - pipeline sequence budget
        * number of reads filtered due to low quality
        * number of reads merged
        * number of chimeras

## Characterizing Sources of variability
### Response Variance
* Experimental replicate variance - how is the variance correlated with different types of technical replicates, how does the variance differ for pipelines
    * section objective - characterize count variance between PCR replicates  
        * is the variance correlated with experimental values e.g. biological sample, PCR plate, well, sequencing depth, or observed count value  
    * Figure - relationship between count and PCR replicate variance
    

### Response linearity 
* section objective - demonstrate how the dataset is used to characterize relative abundance estimates and identify potential sources of bias
* Figure observed vs expected plots
* Figure representative OTUs showing different types of response linearity
* Differentiating between high and low linearity OTUs